---
title: "Concatenated OSHetero2021 Code"
author: "Emily Franz"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 5
    number_sections: false
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  echo = TRUE,
  cache = TRUE,
  collapse = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  autodep = TRUE
)
```


```{r lib, cache=FALSE}
setwd("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021")

source("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/scSeurat.R")
source("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/Downstream.v2.R")
#loading libraries
library(Seurat)
library(future)
library(ggplot2)
# library(fgsea)
library(msigdbr)
library(clusterProfiler)
library(SingleR)
library(dplyr) 
library(pheatmap)
library(RColorBrewer)
library(viridis) # inferno color palette
library(grid)

set.seed(888)
```

# Background 

This Rmarkdown is a concatenation of Sanjana's R codes for OSHetero2021 (Git).

# Load Seurat objects 

Load Seurat objects here so only have to complete once (commented out elsewhere). 

```{r loadobj}
# Create Seurat objects and perform initial QC.  Label original source.
OB <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0031/outs/filtered_feature_bc_matrix/", spec = "human")
OB <- subset(OB, subset = nCount_RNA <20000 & percent.mt <10)
OB$src <- "OB"
OB$cond <- "OB"

#OS17
os17.cx.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0016/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
os17.cx.raw <- subset(os17.cx.raw, subset = nFeature_RNA >3500 & nCount_RNA <50000 & percent.mt <15)
os17.cx.raw$src <- "OS17_Culture"
os17.cx.raw$cond <- "OS17"

os17.tib.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
os17.tib.raw <- subset(os17.tib.raw, subset = nFeature_RNA >3000 & nCount_RNA <60000 & percent.mt <18)
os17.tib.raw$src <- "OS17_Tibia"
os17.tib.raw$cond <- "OS17"

## double check usage throughout rest of code - Emily (fixed issue where coming from wrong file)
os17.lung.raw <- tenXLoadQC(path10x = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/filtered_feature_bc_matrix", spec = "mixHuman")
os17.lung.raw <- subset(os17.lung.raw, subset = nFeature_RNA >1250 & nCount_RNA <60000 & percent.mt <25)
os17.lung.raw$src <- "OS17_Lung"
os17.lung.raw$cond <- "OS17"

#t143b
t143b.cx.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0017/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
t143b.cx.raw <- subset(t143b.cx.raw, subset = nFeature_RNA >4000 & nCount_RNA <74000 & percent.mt <17)
VlnPlot(t143b.cx.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
t143b.cx.raw$src <- "t143b_Culture"
t143b.cx.raw$cond <- "t143b"

t143b.tib.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0052/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
t143b.tib.raw <- subset(t143b.tib.raw, subset = nFeature_RNA >1000 & nCount_RNA <30000 & percent.mt <22)
VlnPlot(t143b.tib.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
t143b.tib.raw$src <- "t143b_Tibia"
t143b.tib.raw$cond <- "t143b"

t143b.lung.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0019-143B-lung/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
t143b.lung.raw <- subset(t143b.lung.raw, subset = nFeature_RNA >1000 & nCount_RNA <30000 & percent.mt <22)
VlnPlot(t143b.lung.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
t143b.lung.raw$src <- "t143b_Lung"
t143b.lung.raw$cond <- "t143b"

#NCHOS2
OS2.cx.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0076/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
OS2.cx.raw <- subset(OS2.cx.raw, subset = nCount_RNA <36000 & percent.mt <50)
VlnPlot(OS2.cx.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS2.cx.raw$src <- "NCHOS2_Flank"
OS2.cx.raw$cond <- "NCHOS2"

OS2.tib.raw<- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0042/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
OS2.tib.raw <- subset(OS2.tib.raw, subset = nFeature_RNA >2500 & nCount_RNA <40000 & percent.mt <20)
OS2.tib.raw$src <- "NCHOS2_Tibia"
VlnPlot(OS2.tib.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS2.tib.raw$cond <- "NCHOS2"

OS2.lung.raw<- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0041/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
OS2.lung.raw <- subset(OS2.lung.raw, subset = nFeature_RNA >2500 & nCount_RNA <60000 & percent.mt <30)
OS2.lung.raw$src <- "NCHOS2_Lung"
VlnPlot(OS2.lung.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS2.lung.raw$cond <- "NCHOS2"

#NCHOS7
OS7.cx.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0043/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
OS7.cx.raw <- subset(OS7.cx.raw, subset = nCount_RNA <50000 & percent.mt <25)
OS7.cx.raw$src <- "NCHOS7_Flank"
VlnPlot(OS7.cx.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS7.cx.raw$cond <- "NCHOS7"

OS7.tib.raw<- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0034/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
OS7.tib.raw <- subset(OS7.tib.raw, subset = nFeature_RNA >2000 & nCount_RNA <35000 & percent.mt <14)
OS7.tib.raw$src <- "NCHOS7_Tibia"
VlnPlot(OS7.tib.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS7.tib.raw$cond <- "NCHOS7"

OS7.lung.raw<- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0055/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
OS7.lung.raw <- subset(OS7.lung.raw, subset = nCount_RNA <42000 & percent.mt <20)
OS7.lung.raw$src <- "NCHOS7_Lung"
VlnPlot(OS7.lung.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS7.lung.raw$cond <- "NCHOS7"
```

# OS.list_all

**Objects generated:** (not shown since done above - OB, os17.cx.raw, os17.tib.raw, os17.lung.raw, t143b.cx.raw, t143b.tib.raw, os17.lung.raw, OS2.cx.raw, OS2.tib.raw, OS2.lung.raw, OS7.cx.raw, OS7.tib.raw, OS7.lung.raw), **OS_list** (list of above, cell cycle regressed)

```{r OS.list_all, cache.lazy=FALSE}
#Create List to compute ITH scores
OS.list <- list(OB = OB,
                OS17.cx = os17.cx.raw,
                OS17.Tibia = os17.tib.raw, 
                OS17.Lung = os17.lung.raw, 
                t143B.cx = t143b.cx.raw,
                t143B.Tibia = t143b.tib.raw, 
                t143B.Lung = t143b.lung.raw,
                OS2.Flank = OS2.cx.raw,
                OS2.Tibia = OS2.tib.raw, 
                OS2.Lung = OS2.lung.raw, 
                OS7.Flank= OS7.cx.raw,
                OS7.Tibia = OS7.tib.raw, 
                OS7.Lung = OS7.lung.raw) 

#load(file = "/gpfs0/home/gdrobertslab/rssxr002/Analysis/2020-OS/OS.list.RData")
for (i in 1:length(OS.list)) {
  OS.list[[i]] <- NormalizeData(OS.list[[i]]) %>%
    FindVariableFeatures(selection.method = "vst") %>%
    ScaleData() %>%
    RunPCA(pc.genes = OS.list[[i]]@var.genes, npcs = 20) %>%
    RunUMAP(reduction = "pca", dims = 1:20) %>%
    FindNeighbors(reduction = "pca", dims = 1:20) %>%
    FindClusters(resolution = 0.3) %>%
    subset(cells = sample(Cells(OS.list[[i]]), 1500))
}

# CCR
# Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

for (i in 1:length(OS.list)) {
  OS.list[[i]] <- CellCycleScoring(object = OS.list[[i]], s.features = s.genes,
                                   g2m.features = g2m.genes, set.ident = TRUE)
  OS.list[[i]] <- ScaleData(object = OS.list[[i]],
                            vars.to.regress = c("S.Score", "G2M.Score"),
                            features = rownames(x = OS.list[[i]]))
  OS.list[[i]] <- RunPCA(OS.list[[i]], pc.genes = OS.list[[i]]@var.genes, npcs = 20) %>%
    RunUMAP(reduction = "pca", dims = 1:20) %>%
    FindNeighbors(reduction = "pca", dims = 1:20) %>%
    FindClusters(resolution = 0.3)
}

save(OS.list, file = "Data/OS.listv3.CCR.RData")
```

# GSE152048

**Objects generated:** raw

```{r GSE152048, cache.lazy=FALSE}
library(rrrSingleCellUtils)
library(Seurat)
library(ggplot2)
library(tidyverse)
library(stringr)
# library(future)
library(harmony)

set.seed(888)
# plan("multisession", workers = 10)
# options(future.globals.maxSize= 10 * 24000 * 1024^2)

# Make a list of sample names
s <- c("BC5", "BC6", "BC10", "BC11", "BC16",
       "BC17", "BC20", "BC21", "BC22")
qc <- c(18000, 25000, 25000, 30000, 70000, 
        40000, 70000, 50000, 50000)
path <- c("Conventional", "Conventional", "Conventional",
          "Conventional", "Conventional", "Chondroblastic",
          "Chondroblastic", "Intraosseous", "Chondroblastic")
type <- c("Primary", "Primary", "Lung Met", "Primary", "Primary",
          "Lung Met", "Primary", "Primary", "Primary")

# Download, file, and extract the files from GEO
if(!dir.exists("GSE152048")) {
  tar_dir <- "PrimaryTumor/GSE152048"
  dir.create(tar_dir)
  geo_pre <- "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE152nnn/GSE152048/suppl/GSE152048_"
  for(i in 1:length(s)){
    gse_path <- str_c(geo_pre, s[i], ".matrix.tar.gz")
    tar_file <- str_c(tar_dir, "/", s[i], ".tar.gz ")
    download.file(gse_path, destfile = tar_file, method = "auto")
    untar(tar_file, exdir = tar_dir)
    file.remove(tar_file)
  }
}

# Create a vector that contains normalized Seurat objects for all samples
raw <- c()
for(i in 1:9) {
  x <- tenx_load_qc(path_10x = str_c("PrimaryTumor/GSE152048/",
                                     s[i], "/"))
  x <- subset(x, subset = nCount_RNA < qc[i] & percent.mt < 13)
  x$src <- s[i]
  x$type <- type[i]
  x$path <- path[i]
  x <- x %>%
    NormalizeData() %>%
    FindVariableFeatures() %>%
    ScaleData() %>%
    RunPCA(verbose = F) %>%
    FindNeighbors(dims = 1:20) %>%
    FindClusters(resolution = 0.3) %>%
    RunUMAP(dims = 1:20)
  print(DimPlot(x, pt.size = 1, label = T) +
          coord_fixed() +
          theme(legend.position = "none") +
          ggtitle(str_c(s[i], " basic clustering")))
  raw <- c(raw, x)
}
rm(x)

save(raw, file = "raw.RData")
```

# QC_ReadsPerCell

Uses OS.list and raw (patient data)

**Objects generated:** name, Cummulative_Var, ElbowPlot, vargenes, tumor_comb

```{r QC_ReadsPerCell, cache.lazy=FALSE}
library(gridExtra)
memory.limit(size=40000)
# load(file = "R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/OS.listv3.CCR.RData")
# for my local try:
# load(file = "OS.listv3.CCR.RData")

name <- names(OS.list)
for (i in 1:13){
  data <- OS.list[[i]]
  data <- Seurat::GetAssayData(data) 
  p <- ggplot(tibble(readsPerCell = colSums(data)), 
              aes(x = readsPerCell)) + 
    geom_histogram(bins = 200) +
    ggtitle(paste0(name[i]))
  pdf(paste0(name[i], " Reads.pdf", sep =" "))
  plot(p)
  dev.off()
}

######################Total variance explained by PCs

##load only scaled data
# load(file = "R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/OS.list_PC50.RData")
# Note from Emily that the above object comes from OS.list_all_local

# # Determine percent of variation associated with each PC
# pct <- data@reductions$pca@stdev / sum(data@reductions$pca@stdev) * 100
# 
# # Calculate cumulative percents for each PC
# cum <- cumsum(pct)
# 
# # Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC as less than 5
# co1 <- which(cum > 90 & pct < 5)[1]
# co1
# 
# # Determine the difference between variation of PC and subsequent PC
# co2 <- sort(which((pct[1:length(pct)-1] - pct[2:length(pct)]) > 0.1),  decreasing = T)[1] + 1 # last point where change of % of variation is more than 0.1%.
# co2


Cummulative_Var <- vector()
for (i in 1:length(OS.list)){
  data <- OS.list[[i]]
  pdf(paste0("ElbowPlot_", name[i], ".pdf"))
  p <- ElbowPlot(data, ndims = 50)
  print(p)
  dev.off()
  pct <- data@reductions$pca@stdev / sum(data@reductions$pca@stdev) * 100
  Cummulative_Var <- append(Cummulative_Var, sum(pct[1:20]))
}

min(Cummulative_Var)
max(Cummulative_Var)

# For each dataset, the first twenty principal components were selected based on the elbow plot 
# for percentage explained variances, representing ~55.5?60.3% of total variances. 

ElbowPlot <- list()
for(i in 1:length(OS.list)){
  data <- OS.list[[i]]
  ElbowPlot[[i]] <- ElbowPlot(data, ndims = 50)
}


pdf("ElbowPlot.pdf", width = 15, height = 15)
grid.arrange(ElbowPlot[[1]], ElbowPlot[[2]], ElbowPlot[[3]], ElbowPlot[[4]],
             ElbowPlot[[5]], ElbowPlot[[6]], ElbowPlot[[7]], ElbowPlot[[8]],
             ElbowPlot[[9]], ElbowPlot[[10]], ElbowPlot[[11]], ElbowPlot[[12]],
             ElbowPlot[[13]], nrow = 4)
#dev.off()

vargenes <- vector()
for (i in 1:length(OS.list)){
  data <- OS.list[[i]]
  vargenes <- append(vargenes, length(data@assays$RNA@var.features))
}

####Patient data
# library(qs)
# tumor_comb <- qread("R:/RESRoberts/Bioinformatics/Analysis/Sanjana/tar_dir/tumor_comb.qs")
# load(file = "raw.RData")

Cummulative_Var <- vector()
for (i in 1:length(raw)){
  data <- raw[[i]] %>% RunPCA()
  pct <- data@reductions$pca@stdev / sum(data@reductions$pca@stdev) * 100
  Cummulative_Var <- append(Cummulative_Var, sum(pct[1:20]))
}

min(Cummulative_Var)
max(Cummulative_Var)
```

# Merge_all

**Objects generated:** OB, os17.cx.raw, os17.tib.raw, os17.lung.raw, t143b.cx.raw, t143b.tib.raw, os17.lung.raw, OS2.cx.raw, OS2.tib.raw, OS2.lung.raw, OS7.cx.raw, OS7.tib.raw, OS7.lung.raw, **OS** (merge of above, cell cycle regressed)

Plotted Dimplots have different tweaks and show OS (normalized, cell cycle regressed, etc). 

```{r merge_all, cache.lazy=FALSE}
memory.limit(size=40000)
library("Seurat")
library(ggplot2)
library(future)

# Merge into a single Seurat object
OS <- merge(OB, y = c(os17.cx.raw, os17.tib.raw, os17.lung.raw, 
                      t143b.cx.raw, t143b.tib.raw, t143b.lung.raw, 
                      OS2.cx.raw, OS2.tib.raw, OS2.lung.raw, 
                      OS7.cx.raw, OS7.tib.raw, OS7.lung.raw),
            add.cell.ids = c("OB", "OS17_Culture", "OS17_Tibia", "OS17_Lung", 
                             "t143b_Culture", "t143b_Tibia", "t143b_Lung",
                             "NCHOS2_Flank", "NCHOS2_Tibia", "NCHOS2_Lung", 
                             "NCHOS7_Flank", "NCHOS7_Tibia", "NCHOS7_Lung"),
            project = "Heterogeneity")

# Process and cluster
OS <- NormalizeData(OS) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = os.17@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

save(OS, file = "OS.combined.OBpreCCR.RData")

# rm(OB, os17.cx.raw, os17.tib.raw, os17.lung.raw, 
#    t143b.cx.raw, t143b.tib.raw, t143b.lung.raw, 
#    OS2.cx.raw, OS2.tib.raw, OS2.lung.raw, 
#    OS7.cx.raw, OS7.tib.raw, OS7.lung.raw)

# CCR
# Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
OS <- CellCycleScoring(object = OS, s.features = s.genes,
                       g2m.features = g2m.genes, set.ident = TRUE)

OS <- ScaleData(object = OS, vars.to.regress = c("S.Score", "G2M.Score"),
                features = rownames(x = OS)) 
OS <- RunPCA(OS, pc.genes = OS@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

save(OS, file = "OS.combined.OBpostCCR.RData")

OS$cond <- as.factor(OS$cond)
OS$cond <- factor(as.factor(OS$cond),
                  levels = c("OS17", "t143b", "NCHOS2", "NCHOS7", "OB"))

# Plot the data (what do you think of these tweaks in terms of color, transparency, and labels?)
DimPlot(OS, reduction = "umap", group.by = "cond", pt.size = 1, label = T) + 
  coord_fixed() + 
  ggtitle("OS by Source") + 
  scale_color_npg(alpha = 1)
DimPlot(OS, reduction = "umap", group.by = "src", pt.size = 1, label = T) + 
  coord_fixed() + 
  ggtitle("OS by Source")
DimPlot(OS, reduction = "umap", pt.size = 1, label = T) + 
  coord_fixed() + 
  ggtitle("OS Clusters") + 
  scale_color_npg(alpha = 0.7)
```

# ComplexHeatmap

Uses cx.pd, Fig1.txt

**Objects generated:** c, cx.pd.log, cx.pd_transpose, df, subj, aka2, aka3, aka4, Lpercent, aka5, P1, P2

*Figure 1, Figure 3* generated here

```{r ComplexHeatmap}
library(tibble)
library(tidyr)
library(ggplot2)
library(tidyverse)
library(stringr)

#ComplexHeatmap
# Fig1.txv from above code (previously called cxpd.tsv in Sanjana's files)
cx.pd <- read.delim("Fig1.tsv", row.names = "rn")

#remove "HALLMARK_"
c <- rownames(cx.pd)
c <- gsub("HALLMARK_", "", c)
#remove "_" by removing special characters
c <- gsub("_", " ", c)
rownames(cx.pd) <- c
cx.pd.log <- -log10(cx.pd[,-1]) #log transform
  
cx.pd_transpose <- as.data.frame(t(cx.pd.log))
df <- as.matrix(cx.pd_transpose)
subj <- c(paste0("A", 0:3),
          paste0("B", 0:3),
          paste0("C", 0:2),
          paste0("D", 0:3))
#rownames(df) <- subj
aka2 = data.frame(ID = factor(c(rep("OS17", 4),
                                rep("t143B", 4),
                                rep("OS2", 3),
                                rep("OS7", 4)
)))
rownames(aka2)<-subj
aka3 = list(ID = c(OS17 = "#E64B35FF",
                   t143B = "#4DBBD5FF",
                   OS2 = "#00A087FF",
                   OS7 = "#3C5488FF"))

#Include percentage of cells in lung for each cluster
aka4 <- data.frame (Lpercent = c(46.577747,
                                 10.17901,
                                 39.69814,
                                 3.545104,
                                 2.2204908,
                                 92.306194,
                                 3.3502143,
                                 2.1231009,
                                 20.32767,
                                 71.480583,
                                 8.191748,
                                 71.01349,
                                 1.383604,
                                 1.176064,
                                 26.426842))
rownames(aka4) <- subj
aka5 = (c(rep("OS17", 4),
          rep("t143B", 4),
          rep("OS2", 3),
          rep("OS7", 4)))

############################## Figure 3 ################################
P1 <- df %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  full_join(aka4 %>%
              rownames_to_column(var = "Sample")) %>%
  pivot_longer(c(-Sample, -Lpercent),
               names_to = "Pathway",
               values_to = "pval",
               names_repair = "minimal") %>%
  mutate(Signif = pval >= (-1 * log10(0.05)),
         sample_order = str_remove(Sample, "[0-9]") %>%
           rank() * 1000 - Lpercent,
         Sample = reorder(Sample, sample_order)) %>%
  ggplot(., aes(x = Sample, y = Pathway, fill = Signif)) + 
  geom_tile() +
  geom_text(aes(label = sprintf("%0.2f", pval))) +
  scale_fill_manual(values = c("gray", "red")) +
  scale_y_discrete(limits = rev) +
  theme(legend.position = "none") +
  ylab("")

P2 <- aka4 %>%
  rownames_to_column(var = "Sample") %>%
  full_join(aka2 %>%
              rownames_to_column(var = "Sample")) %>%
  mutate(sample_order = str_remove(Sample, "[0-9]") %>%
           rank() * 1000 - Lpercent,
         Sample = reorder(Sample, sample_order),
         ID = reorder(ID, sample_order)) %>%
  ggplot(., aes(x = Sample, y = Lpercent, fill = ID)) +
  geom_bar(stat = "identity") +
  ylab("Percent\nin lung") +
  xlab("") +
  theme(legend.position = "top", ) +
  labs(fill = "")

cowplot::plot_grid(P2,
                   P1,
                   ncol = 1,
                   align = "v",
                   rel_heights = c(2, 10))

############################## Figure 1 ################################
Fig1 <- read_delim("Fig1.txt", 
                   delim = "\t", 
                   col_names = TRUE)

colnames(Fig1) <- c("Pathway", "A0", "A1", "A2", "A3", "B0", "B1", "B2", "B3")

Fig1 %>%
  column_to_rownames(var = "Pathway") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  left_join(aka4 %>% rownames_to_column(var = "Sample")) %>%
  pivot_longer(c(-Sample, -Lpercent),
               names_to = "Pathway",
               values_to = "pval",
               names_repair = "minimal") %>%
  mutate(Signif = pval >= (-1 * log10(0.05))) %>%
  ggplot(., aes(x = Sample, y = Pathway, fill = Signif)) + 
  geom_tile() +
  geom_text(aes(label = sprintf("%0.2f", pval))) +
  scale_fill_manual(values = c("gray", "red")) +
  scale_y_discrete(limits = rev) +
  theme(legend.position = "none",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks.x = element_blank()) +
  ylab("") 

```

# CulturePDXHetero

**Objects generated:** cx.raw (OS17), cx.raw.nC, pcx (from cx.raw, for plotting), cid (from cx.raw), S0043 (NCHOS7), S0043.CCR.nC, S0043.CCR, pcx (from S0043.CCR, for plotting), cid (from S0043.CCR), B.list (from cx.raw, S0043.CCR), cx.pd (for heatmap), cx.pd.log (for heatmap-red/grey)

Note-possible labeling error in:

Culture plots
pdf("NCHOS7.Culture.pdf", width = 5, height = 5)
pcx <- DimPlot(S0043.CCR, reduction = "umap", pt.size = 1, label = F) + 
  coord_fixed() + 
  ggtitle("OS17 in Culture") + NoAxes()
pcx +  scale_color_npg(alpha = 1)
dev.off()

```{r CulturePDXHetero, cache.lazy=FALSE}
source("scSeurat.R")
source("Downstream.v2.R")
#loading libraries
library(Seurat)
library(future)
library(ggplot2)
# library(fgsea)
library(msigdbr)
library(clusterProfiler)
library(SingleR)
library(dplyr) # %<%
library(pheatmap)
library(RColorBrewer)
library(viridis) # inferno color palette
library(grid)

# OS17
# Create Seurat objects and perform initial QC. Label original source.
cx.raw <- tenXLoadQC("/gpfs0/home/gdrobertslab/lab/Counts/S0016/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
cx.raw <- subset(cx.raw, subset = nFeature_RNA >3500 & nCount_RNA <50000 & percent.mt <15)
cx.raw$src <- "Culture"

cx.raw <- NormalizeData(cx.raw) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = cx.raw@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

DimPlot(cx.raw, reduction = "umap", pt.size = 1, label = T) + 
  coord_fixed() + 
  scale_color_npg(alpha = 0.7)

# Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
cx.raw <- CellCycleScoring(object = cx.raw, s.features = s.genes,
                           g2m.features = g2m.genes, set.ident = TRUE)

cx.raw <- ScaleData(object = cx.raw, vars.to.regress = c("S.Score", "G2M.Score"),
                    features = rownames(x = cx.raw)) 
cx.raw <- RunPCA(cx.raw, pc.genes = cx.raw@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

#Find optimal clustering resolution
cx.raw.nC <- nRes(cx.raw, 
                  res = seq(from = 0.1, to = 0.2, by = 0.05))
plot <- pSil(cx.raw.nC, 0.15)

#With res = 0.15, cells in cluster 3 do not have any significantly different genes that are deferentially regulated
#Therefore, we decided to proceed with res = 0.1
cx.raw <- FindClusters(cx.raw, resolution = 0.1)

#Culture plots
pdf("OS17.Culture.pdf", width = 5, height = 5)
pcx <- DimPlot(cx.raw, reduction = "umap", pt.size = 1, label = T) + 
  coord_fixed() + 
  ggtitle("OS17 in Culture") + NoAxes()
pcx +  scale_color_npg(alpha = 1)
dev.off()

#cell cycle distribution of clusters

cid <- sort(unique(cx.raw@active.ident))

vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
pdf("OS17_pie.pdf", width = 15, height = 15)
grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 4)))
for(i in 1:3){
  temp <- WhichCells(cx.raw, idents = cid[[i]]) 
  LT.cells <- subset(cx.raw, cells = temp)
  df <- table(LT.cells$Phase)
  bp <- ggplot(as.data.frame(df), aes(x="", y=Freq, fill=Var1))+
    geom_bar(width = 1, stat = "identity")
  pie <- bp + coord_polar("y", start=0) + scale_fill_brewer(palette="Blues")+
    theme_minimal() + NoAxes() +
    ggtitle(paste("Cluster", cid[[i]]))
  print(pie, vp=vplayout(ceiling(i/4), i))
}
dev.off()

#Total number of cells
# table(Idents(cx.raw)) %>%sum()
# [1] 3178

# NCHOS7 (flank)
# Create Seurat objects and perform initial QC. Label original source.

S0043 <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0043/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
S0043 <- subset(S0043, subset = nCount_RNA <50000 & percent.mt <25)
VlnPlot(S0043, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

S0043 <- NormalizeData(S0043) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = S0043@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

DimPlot(S0043, reduction = "umap", pt.size = 1, label = T) + 
  coord_fixed() + 
  scale_color_npg(alpha = 0.7)

# Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
S0043.CCR <- CellCycleScoring(object = S0043, s.features = s.genes,
                              g2m.features = g2m.genes, set.ident = TRUE)

S0043.CCR <- ScaleData(object = S0043.CCR, vars.to.regress = c("S.Score", "G2M.Score"),
                       features = rownames(x = S0043.CCR)) 
S0043.CCR <- RunPCA(S0043.CCR, pc.genes = S0043.CCR@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

#Find optimal clustering resolution
S0043.CCR.nC <- nRes(S0043.CCR, 
                     res = seq(from = 0.1, to = 0.15, by = 0.01))
plot <- pSil(S0043.CCR.nC, 0.15)

S0043.CCR <- FindClusters(S0043.CCR, resolution = 0.15)

# save(S0043.CCR, file ="R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/17.Harmony/Harmony in individual samples/Figure1/NCHOS7-S0043_CCR_v1.RData")
#save(S0043.CCR, file ="NCHOS7-S0043_CCR_v1.RData")

#Culture plots
pdf("NCHOS7.Flank.pdf", width = 5, height = 5)
pcx <- DimPlot(S0043.CCR, reduction = "umap", pt.size = 1, label = F) + 
  coord_fixed() + 
  ggtitle("NCHOS7 in Flank") + NoAxes()
pcx +  scale_color_npg(alpha = 1)
dev.off()

#cell cycle distribution of clusters

cid <- sort(unique(S0043.CCR@active.ident))

vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
pdf("NCHOS7_pie.pdf", width = 15, height = 15)
grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 4)))
for(i in 1:4){
  temp <- WhichCells(S0043.CCR, idents = cid[[i]]) 
  LT.cells <- subset(S0043.CCR, cells = temp)
  df <- table(LT.cells$Phase)
  bp <- ggplot(as.data.frame(df), aes(x="", y=Freq, fill=Var1))+
    geom_bar(width = 1, stat = "identity")
  pie <- bp + coord_polar("y", start=0) + scale_fill_brewer(palette="Blues")+
    theme_minimal() + NoAxes() + NoLegend()
  ggtitle(paste("Cluster", cid[[i]]))
  print(pie, vp=vplayout(ceiling(i/4), i))
}
dev.off()

#Total number of cells
# table(Idents(S0043.CCR)) %>%sum()
# [1] 1998

# library("scales")
# show_col(pal_nejm("default")(8))
# show_col(pal_nejm("default", alpha = 0.6)(8))

##########################Pathway enrichment analysis; display adjust p-values
B.list <- list(OS17 = cx.raw,
               OS7 = S0043.CCR)

em.hm.list <- list()
for (i in 1:length(B.list)) {
  em.hm.list[[i]] <- DGEA(B.list[[i]]) 
}

library(data.table)
for(i in 1:(length(em.hm.list))) {
  em.hm.list[[i]] <- setDT(em.hm.list[[i]], keep.rownames = TRUE)[]
}

temp1 <- em.hm.list[[1]]
temp2 <- em.hm.list[[2]]

cx.pd <- full_join(temp1, temp2, by = "rn")

##Write table to save for supplemental data and for generating Heatmap
# write.table(cx.pd, file=('R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2021/Fig1.tsv'), quote=FALSE, sep='\t')
# save.image(file ="R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2021/ProcessedData/CulturePDXHetero.RData")
#write.table(cx.pd, file=('Fig1.tsv'), quote=FALSE, sep='\t')
#save.image(file ="CulturePDXHetero.RData")


cx.pd <- column_to_rownames(cx.pd, var = "rn")
cx.pd[is.na(cx.pd)] <- 1

#remove "HALLMARK_"
c <- rownames(cx.pd)
c <- gsub("HALLMARK_", "", c)
#remove "_" by removing special characters
c <- gsub("_", " ", c)
rownames(cx.pd) <- c

cx.pd.log <- -log10(cx.pd)
colnames(cx.pd.log) <- c("A0", "A1", "A2", "B0", "B1", "B2", "B3")
##To prevent values that show up as -0.00, we take the absolute values
cx.pd.log <- abs(cx.pd.log)

cx.pd.log %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  left_join(aka4 %>% rownames_to_column(var = "Sample")) %>%
  pivot_longer(c(-Sample, -Lpercent),
               names_to = "Pathway",
               values_to = "pval",
               names_repair = "minimal") %>%
  mutate(Signif = pval >= (-1 * log10(0.05))) %>%
  ggplot(., aes(x = Sample, y = Pathway, fill = Signif)) + 
  geom_tile() +
  geom_text(aes(label = sprintf("%0.2f", pval))) +
  scale_fill_manual(values = c("gray", "red")) +
  scale_y_discrete(limits = rev) +
  theme(legend.position = "none",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks.x = element_blank()) +
  ylab("") 
```

# PatientData

Uses tar_dir

**Objects generated:** s, qc, path, type, raw, x, comb, primary, ms, mod_names, tumor, tumor_a (subset), tumor_b (subset), names_a, names_b, tumor_2, tumor_comb, s, raw, data, data.NC, gl, Glycolysis, Hypoxia, EMT, TNFA

```{r PatientData, cache.lazy=FALSE}
library(rrrSingleCellUtils)
library(Seurat)
library(ggplot2)
library(tidyverse)
library(stringr)
# library(future)
library(harmony)
library(devtools)

set.seed(888)
# plan("multisession", workers = 10)
# options(future.globals.maxSize= 10 * 24000 * 1024^2)

# Make a list of sample names
s <- c("BC5", "BC6", "BC10", "BC11", "BC16",
       "BC17", "BC20", "BC21", "BC22")
qc <- c(18000, 25000, 25000, 30000, 70000, 
        40000, 70000, 50000, 50000)
path <- c("Conventional", "Conventional", "Conventional",
          "Conventional", "Conventional", "Chondroblastic",
          "Chondroblastic", "Intraosseous", "Chondroblastic")
type <- c("Primary", "Primary", "Lung Met", "Primary", "Primary",
          "Lung Met", "Primary", "Primary", "Primary")

# Download, file, and extract the files from GEO
# 
tar_dir <- "/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/tar_dir"

  geo_pre <- "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE152nnn/GSE152048/suppl/GSE152048_"
  for(i in 1:length(s)){
    gse_path <- str_c(geo_pre, s[i], ".matrix.tar.gz")
    tar_file <- str_c(tar_dir, "/", s[i], ".tar.gz ")
    download.file(gse_path, destfile = tar_file, method = "auto")
    untar(tar_file, exdir = tar_dir)
    file.remove(tar_file)
  }

# Create a vector that contains normalized Seurat objects for all samples
raw <- c()
for(i in 1:9) {
  x <- tenx_load_qc(str_c("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/tar_dir/",
                          s[i], "/"))
  x <- subset(x, subset = nCount_RNA < qc[i] & percent.mt <13)
  x$src <- s[i]
  x$type <- type[i]
  x$path <- path[i]
  x <- x %>%
    NormalizeData() %>%
    FindVariableFeatures() %>%
    ScaleData() %>%
    RunPCA(verbose = F) %>%
    FindNeighbors(dims = 1:20) %>%
    FindClusters(resolution = 0.3) %>%
    RunUMAP(dims = 1:20)
  print(DimPlot(x, pt.size = 1, label = T) +
          coord_fixed() +
          theme(legend.position = "none") +
          ggtitle(str_c(s[i], " basic clustering")))
  raw <- c(raw, x)
}
rm(x)

# Save a stopping point - individual objects
# save(raw, file = "R:/RESRoberts/Bioinformatics/Analysis/Sanjana/tar_dir/raw.RData")
save(raw, file = "tar_dir/raw.RData")

#Apply Glycolysis module score to each of the primary tumor cells
gl <- list()
Glycolysis <- read.table(file = "/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/pathways/Glycolysis.txt", header = FALSE, sep = ",")
Glycolysis <- Glycolysis[,1]

# OxPhos <- read.table(file = "/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/pathways/OxPhos.txt", header = FALSE, sep = ",")
# OxPhos <- OxPhos[,1]

Hypoxia <- read.table(file = "/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/pathways/Hypoxia.txt", header = FALSE, sep = ",")
Hypoxia <- Hypoxia[,1]

# SASP <- read.table(file = "/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/pathways/SASP.txt", header = FALSE, sep = ",")
# SASP <- SASP[,1]

EMT <- read.table(file = "/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/pathways/EMT.txt", header = FALSE, sep = ",")
EMT <- EMT[,1]

TNFA <- read.table(file = "/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/pathways/TNFA.txt", header = FALSE, sep = ",")
TNFA <- TNFA[,1]

# growth <- read.table(file = "/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/pathways/growth.txt", header = FALSE, sep = ",")
# growth <- growth[,1]

gl$Glycolysis <- Glycolysis
gl$Hypoxia <- Hypoxia
gl$EMT <- EMT
gl$TNFA <- TNFA
# gl$growth <- growth

mod_names <- names(gl)

#Glycolysis - Added min.cutoff to isolate for cells most module like
for(i in 1:length(raw)) {
  raw[[i]] <- AddModuleScore(raw[[i]], gl[1], name = names(gl[1]))
  p <- FeaturePlot(raw[[i]], features = str_c(names(gl[1]), "1"), min.cutoff = 0.1, pt.size = 1,
                   order = T, cols = c("lightgoldenrod", "darkred"))
  pdf(paste0(names(gl[1]), " FeaturePlot ", names(raw[i]), ".pdf"))
  print(p)
  dev.off()
}

#Hypoxia - Added min.cutoff to isolate for cells most module like
for(i in 1:length(raw)) {
  raw[[i]] <- AddModuleScore(raw[[i]], gl[2], name = names(gl[2]))
  p <- FeaturePlot(raw[[i]], features = str_c(names(gl[2]), "1"), min.cutoff = 0.1, pt.size = 1,
                   order = T, cols = c("lightgoldenrod", "darkred"))
  pdf(paste0(names(gl[2]), " FeaturePlot ", names(raw[i]), ".pdf"))
  print(p)
  dev.off()
}

#EMT - Added min.cutoff to isolate for cells most module like
for(i in 1:length(raw)) {
  raw[[i]] <- AddModuleScore(raw[[i]], gl[3], name = names(gl[3]))
  p <- FeaturePlot(raw[[i]], features = str_c(names(gl[3]), "1"), min.cutoff = 0.1, pt.size = 1,
                   order = T, cols = c("lightgoldenrod", "darkred"))
  pdf(paste0(names(gl[3]), " FeaturePlot ", names(raw[i]), ".pdf"))
  print(p)
  dev.off()
}

#TNFA - Added min.cutoff to isolate for cells most module like
for(i in 1:length(raw)) {
  raw[[i]] <- AddModuleScore(raw[[i]], gl[4], name = names(gl[4]))
  p <- FeaturePlot(raw[[i]], features = str_c(names(gl[4]), "1"), min.cutoff = 0.1, pt.size = 1,
                   order = T, cols = c("lightgoldenrod", "darkred"))
  pdf(paste0(names(gl[4]), " FeaturePlot ", names(raw[i]), ".pdf"))
  print(p)
  dev.off()
}
# 
# #Growth - Added min.cutoff to isolate for cells most module like
# for(i in 1:length(raw)) {
#   raw[[i]] <- AddModuleScore(raw[[i]], gl[5], name = names(gl[5]))
#   p <- FeaturePlot(raw[[i]], features = str_c(names(gl[5]), "1"), min.cutoff = 0.1, pt.size = 1,
#                    order = T, cols = c("lightgoldenrod", "darkred"))
#   pdf(paste0(names(gl[5]), " FeaturePlot ", names(raw[i]), ".pdf"))
#   print(p)
#   dev.off()
# }

# ##Perform cell cycle regression on each of the samples
# for(i in 1:length(raw)) {
#   raw[[i]] <- killCC(raw[[i]], cc.regress = "Y", use.pcs = 20)
# }

##Glycolysis and Hypoxia seem to have very high correlation. 
#Output numbe rof cells in each primary tumor sample

# sum(table(Idents(raw[[1]])))
# 8872
# sum(table(Idents(raw[[2]])))
# 825
# sum(table(Idents(raw[[3]])))
# 6357
# sum(table(Idents(raw[[4]])))
# 3115
# sum(table(Idents(raw[[5]])))
# 6886
# sum(table(Idents(raw[[6]])))
# 986
# sum(table(Idents(raw[[7]])))
# 2812

```

# Ob_ITH_Score

*code not on Github*

*Located at "R:\RESRoberts\Bioinformatics\Analysis\Sanjana\2020\17.Harmony\Harmony in individual samples\Figure3"*

Generates "OB-S0031_CCR_v2.RData"

```{r Ob_ITH_Score, cache.lazy=FALSE}
#ITH in Osteoblasts

#source("R:/RESRoberts/Bioinformatics/Analysis/scSeurat.R")
#Culture
S0031 <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0031/outs/filtered_feature_bc_matrix/", spec = "human")
S0031 <- subset(S0031, subset = nCount_RNA <20000 & percent.mt <10)
VlnPlot(S0031, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
S0031 <- NormalizeData(S0031)
S0031 <- FindVariableFeatures(S0031, selection.method = "vst")
S0031 <- ScaleData(S0031)
S0031 <- RunPCA(S0031, pc.genes = S0031@var.genes, npcs = 20)
S0031 <- RunUMAP(S0031, reduction = "pca", dims = 1:20)
S0031 <- FindNeighbors(S0031, reduction = "pca", dims = 1:20)
S0031 <- FindClusters(S0031, resolution = 0.3)
DimPlot(S0031, reduction = "umap", pt.size = 1) +coord_fixed()

# Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
# s.genes <- paste0('hg19-', s.genes)
# g2m.genes <- paste0('hg19-', g2m.genes)
S0031<- CellCycleScoring(object = S0031, s.features = s.genes,
                         g2m.features = g2m.genes, set.ident = TRUE)


#Before CCR
DimPlot(S0031, reduction = "umap", group.by = "Phase", pt.size = 1) +coord_fixed()

#Do we need cell cyle regression to estimate heterogeneity - I would think yes
#Before CCR
z <- (Embeddings(S0031, reduction = "pca")[, 1:20])
mat <- dist(z, diag = TRUE, upper = FALSE)
mat2 <- as.matrix(mat)
mat2[upper.tri(mat2, diag = TRUE)] <- NA
S0031_mean <- colMeans(mat2, na.rm = TRUE)
summary(S0031_mean)

# Box Plots
x1 <- S0031_mean

boxplot(x1, main="ITH Score",
        xlab="Osteoblasts", ylab="ITH Score",
        names=c("Ob-culture"),
        col="gold")

# save(S0031, file ="R:/RESRoberts/Bioinformatics/Analysis/Sanjana/17.Harmony/Harmony in individual samples/Figure3/OB-S0031")

#CCR
S0031.CCR <- ScaleData(object = S0031, vars.to.regress = c("S.Score", "G2M.Score"),
                       features = rownames(x = S0031))

# save(S0031.CCR, file ="OB-S0031_CCR_v1")

#Post-CCR Clustering
S0031.CCR <- RunPCA(S0031.CCR, pc.genes = S0031.CCR@var.genes, npcs = 20)
S0031.CCR <- RunUMAP(S0031.CCR, reduction = "pca", dims = 1:20)
S0031.CCR <- FindNeighbors(S0031.CCR, reduction = "pca", dims = 1:20)
S0031.CCR <- FindClusters(S0031.CCR, resolution = 0.3)
DimPlot(S0031.CCR, reduction = "umap", pt.size = 1) +coord_fixed()
DimPlot(S0031.CCR, reduction = "umap", group.by = "Phase", pt.size = 1) +coord_fixed()

save(S0031.CCR, file ="OB-S0031_CCR_v2.RData")

#After CCR
z <- (Embeddings(S0031.CCR, reduction = "pca")[, 1:20])
mat <- dist(z, diag = TRUE, upper = FALSE)
mat2 <- as.matrix(mat)
mat2[upper.tri(mat2, diag = TRUE)] <- NA
S0031.CCR_mean <- colMeans(mat2, na.rm = TRUE)
summary(S0031.CCR_mean)

# Box Plots
x2 <- S0031.CCR_mean

boxplot(x1, x2,data=mtcars, main="ITH Score",
        xlab="Osteoblasts", ylab="ITH Score",
        names=c("Ob-pre_CCR", "Ob-post_CCR"),
        col="gold")

boxplot(x2,data=mtcars, main="ITH Score",
        xlab="Osteoblasts", ylab="ITH Score",
        names="Osteoblast",
        col="gold")
```


# OsteoblastAnalysis

Uses S0031.CCR (OB-S0031_CCR_v2.RData), SanjanaPlots2 (.txt), OS.list (OS.listv3.CCR.RData)

**Objects generated:** data (from S0031.CCR), data.nC (from S0031.CCR), pcx, plot (from S0031.CCR), cid, vplayout, ob.markers, B.list, em.hm.list, temp1, heatmap, cluster.0, cluster.1, cluster.2, cluster0, cluster1, cluster2, data (from OS.list), data.nC (from OS.list), plot (from OS.list)

```{r OsteoblastAnalysis, cache.lazy=FALSE, eval=F}
############################Overlay osteoblast markers on Culture/Flank cells 
load("OB-S0031_CCR_v2.RData")

library(tibble)
library(tidyr)
library(ggplot2)
library(tidyverse)
library(stringr)

data <- S0031.CCR
DimPlot(data, reduction = "umap", pt.size = 1, label = T) + 
  coord_fixed() + 
  scale_color_npg(alpha = 0.7)

#Set optimum resolution for clustering
data.nC <- nRes(data, 
                res = seq(from = 0.1, to = 0.9, by = 0.1))
#Optimal resolution identified as 0.1
plot <- pSil(data.nC, 0.9)
data <- data %>% FindClusters(resolution = 0.05)

#Culture plots
pdf("Osteoblast_Cx.pdf", width = 5, height = 5)
pcx <- DimPlot(data, reduction = "umap", pt.size = 1, label = F) + 
  coord_fixed() + 
  ggtitle("OS17 in Culture") + NoAxes()
pcx +  scale_color_npg(alpha = 1)
dev.off()

#cell cycle distribution of clusters

cid <- sort(unique(data@active.ident))

vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
pdf("OB_pie.pdf", width = 15, height = 15)
grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 2)))
for(i in 1:2){
  temp <- WhichCells(data, idents = cid[[i]]) 
  LT.cells <- subset(data, cells = temp)
  df <- table(LT.cells$Phase)
  bp <- ggplot(as.data.frame(df), aes(x="", y=Freq, fill=Var1))+
    geom_bar(width = 1, stat = "identity")
  pie <- bp + coord_polar("y", start=0) + scale_fill_brewer(palette="Blues")+
    theme_minimal() + NoAxes() +
    ggtitle(paste("Cluster", cid[[i]]))
  print(pie, vp=vplayout(ceiling(i/2), i))
}
dev.off()


###################Heatmap of OB markers
ob.markers <- FindAllMarkers(data, only.pos = FALSE, min.pct = 0.25)

ob.markers %>%
  group_by(cluster) %>%
  top_n(n = 10, wt = avg_log2FC) -> top10

pdf("OBHeatmap.pdf", width = 7, height = 6)
p <- DoHeatmap(data, features = top10$gene, size = 5) + NoLegend()
p + theme(text = element_text(size=16, color = "black"))
dev.off()

####################Pathways associated with OB clusters
B.list <- list(OB = data)

em.hm.list <- list()
for (i in 1:length(B.list)) {
  em.hm.list[[i]] <- DGEA(B.list[[i]]) 
}

temp1 <- em.hm.list[[1]]
temp1 <- -log10(temp1)
#remove "_" by removing special characters
c <- rownames(temp1)
c <- gsub("SIG ", "", c)
c <- gsub("NABA ", "", c)
c <- gsub("SA ", "", c)
c <- gsub("_", " ", c)
rownames(temp1) <- c
#remove negative 0 values due to number of decimal points displayed
temp1 <- abs(temp1)

heatmap <- temp1 %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  left_join(aka4 %>% rownames_to_column(var = "Sample")) %>%
  pivot_longer(c(-Sample, -Lpercent),
               names_to = "Pathway",
               values_to = "pval",
               names_repair = "minimal") %>%
  mutate(Signif = pval >= (-1 * log10(0.05))) %>%
  ggplot(., aes(x = Sample, y = Pathway, fill = Signif)) + 
  geom_tile() +
  geom_text(aes(label = sprintf("%0.2f", pval))) +
  scale_fill_manual(values = c("gray", "red")) +
  scale_y_discrete(limits = rev) +
  theme(legend.position = "none",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks.x = element_blank()) +
  ylab("") 

pdf("OB_Pathway.pdf", height = 8, width = 8)
plot(heatmap)
dev.off()

#############################Generate phentoype module
cluster.0 <- FindMarkers(data, ident.1 = "0", only.pos = TRUE, min.pct = 0.25)
cluster.1 <- FindMarkers(data, ident.1 = "1", only.pos = TRUE, min.pct = 0.25)
cluster.2 <- FindMarkers(data, ident.1 = "2", only.pos = TRUE, min.pct = 0.25)

#Extract gene names
cluster0 <- list(rownames(cluster.0))
cluster1 <- list(rownames(cluster.1))
cluster2 <- list(rownames(cluster.2))

write.xlsx(cluster.0, file = "OBAnalysis/cluster.0.xlsx")
write.xlsx(cluster.1, file = "OBAnalysis/cluster.1.xlsx")
write.xlsx(cluster.2, file = "OBAnalysis/cluster.2.xlsx")

##############Annotation of phenotype module
#Ran msigdb gene set enrichment on http://www.gsea-msigdb.org/gsea/index.jsp
#Visualization of p values usign barplot

#editted out since cannot find txt file _ Emily 
#SanjanaPlots2 <- read.delim("C:/Users/rssxr002/Downloads/SanjanaPlots2.txt")
 SanjanaPlots2 <- read.delim("SanjanaPlots.txt")
 
 ggplot(SanjanaPlots2, aes(x = -1 * Order,
                           y = FDR.log,
                           fill = FDR.log > 0)) +
   geom_bar(stat = "identity") +
   coord_flip() +
   facet_wrap(~ Cluster, ncol = 1) +
   scale_fill_brewer(type = "qual", palette = "Set1") +
   geom_text(aes(y = Label_y * 15, label = Pathway)) +
   theme_bw() +
   ylab("") 

#Addmodule score to OS cell culture/flank models
#load("OS.listv3.CCR.RData")

# OB <- OS.list[[1]]
# OS17.cx <- OS.list[[2]]
# OS17.Tibia <-  OS.list[[3]]
# OS17.Lung <-OS.list[[4]]
# t143B.cx <- OS.list[[5]]
# t143B.Tibia <-OS.list[[6]]
# t143B.Lung <- OS.list[[7]]
# OS2.Flank <- OS.list[[8]]
# OS2.Tibia <- OS.list[[9]]
# OS2.Lung <- OS.list[[10]]
# OS7.Flank <-  OS.list[[11]]
# OS7.Tibia <- OS.list[[12]]
# OS7.Lung <- OS.list[[13]]

for (i in 1:13) {
  OS.list[[i]] <- AddModuleScore(OS.list[[i]], features = cluster0, ctrl = 5, name = 'cluster0')
  OS.list[[i]] <- AddModuleScore(OS.list[[i]], features = cluster1, ctrl = 5, name = 'cluster1')
  OS.list[[i]] <- AddModuleScore(OS.list[[i]], features = cluster2, ctrl = 5, name = 'cluster2')
}

plot1 <- DotPlot(OS.list[[2]], features = c("cluster01", "cluster11", "cluster21")) + coord_flip() 
plot2 <- DotPlot(OS.list[[3]], features = c("cluster01", "cluster11", "cluster21")) + coord_flip()
plot3 <- DotPlot(OS.list[[4]], features = c("cluster01", "cluster11", "cluster21")) + coord_flip()

# DotPlot(OS.list[[1]], features = c("cluster01", "cluster11", "cluster21")) + coord_flip() 

# plot1 <- RidgePlot(OS.list[[11]], features = c("cluster01", "cluster11", "cluster21"))
# plot2 <- RidgePlot(OS.list[[12]], features = c("cluster01", "cluster11", "cluster21"))
# plot3 <- RidgePlot(OS.list[[13]], features = c("cluster01", "cluster11", "cluster21"))

  CombinePlots(
    plots = list(plot1, plot2, plot3),
    legend = 'none', nrow = 1)
  
  #####################Optimize cluster resolution 
  data <- OS.list[[2]] 
  DimPlot(data, reduction = "umap", pt.size = 1, label = T) + 
    coord_fixed() + 
    scale_color_npg(alpha = 0.7)
  
  #Set optimum resolution for clustering
  data.nC <- nRes(data, 
                  res = seq(from = 0.1, to = 0.3, by = 0.05))
  plot <- pSil(data.nC, 0.2)
  OS.list[[4]] <- OS.list[[2]] %>% FindClusters(resolution = 0.2)
  
  # OS.list[[1]] res 0.15
  # OS.list[[2]] res 0.15
  # OS.list[[3]] res 0.25
  # OS.list[[4]] res 0.2
  
  # OS.list[[5]] res 0.25
  # OS.list[[6]] res 0.25
  # OS.list[[7]] res 0.25
  # OS.list[[8]] res 0.25
  # OS.list[[9]] res 0.25
  # OS.list[[10]] res 0.25
  # OS.list[[11]] res 0.25
  # OS.list[[12]] res 0.25
  # OS.list[[13]] res 0.25
  
  
  OS.list[[1]] <- OS.list[[1]] %>% FindClusters(resolution = 0.15)
  OS.list[[2]] <- OS.list[[2]] %>% FindClusters(resolution = 0.15)
  OS.list[[3]] <- OS.list[[3]] %>% FindClusters(resolution = 0.25)
  OS.list[[4]] <- OS.list[[4]] %>% FindClusters(resolution = 0.2)
  # OS.list[[5]] <- OS.list[[5]] %>% FindClusters(resolution = 0.2)
  # OS.list[[6]] <- OS.list[[6]] %>% FindClusters(resolution = 0.2)
  # OS.list[[7]] <- OS.list[[7]] %>% FindClusters(resolution = 0.2)
  # OS.list[[8]] <- OS.list[[8+]] %>% FindClusters(resolution = 0.2)
  # OS.list[[9]] <- OS.list[[9]] %>% FindClusters(resolution = 0.2)
  # OS.list[[10]] <- OS.list[[10]] %>% FindClusters(resolution = 0.2)
  # OS.list[[11]] <- OS.list[[11]] %>% FindClusters(resolution = 0.2)  
  # OS.list[[12]] <- OS.list[[12]] %>% FindClusters(resolution = 0.2)
  # OS.list[[13]] <- OS.list[[13]] %>% FindClusters(resolution = 0.2)
```


# AllpseudoBulk_Venn

Uses OS

**Objects generated:** os17, t143B, NCHOS2, NCHOS7, data (different than previously used; multiple definitions), Lung.up, Tibia.up, Lung.down, Tibia.down, genes, x, "Lung.down.pdf", intersect, tmp, em, pathways, "pathways.xlsx", "AllpseudoBulk_Venn.RData", SanjanaPlots, 

```{r AllpseudoBulk_Venn, cache.lazy=FALSE}
# Psudobulk to determine gene differentially regulated on tissue colonization and 
# Venn diagram to visualize shared genes

# # Merge into a single Seurat object
# OS <- merge(OB, y = c(os17.cx.raw, os17.tib.raw, os17.lung.raw, 
#                       t143b.cx.raw, t143b.tib.raw, t143b.lung.raw, 
#                       OS2.cx.raw, OS2.tib.raw, OS2.lung.raw, 
#                       OS7.cx.raw, OS7.tib.raw, OS7.lung.raw),
#             add.cell.ids = c("OB", "OS17_Culture", "OS17_Tibia", "OS17_Lung", 
#                              "t143b_Culture", "t143b_Tibia", "t143b_Lung",
#                              "NCHOS2_Flank", "NCHOS2_Tibia", "NCHOS2_Lung", 
#                              "NCHOS7_Flank", "NCHOS7_Tibia", "NCHOS7_Lung"),
#             project = "Heterogeneity")

# Emily added/edited
DimPlot(OS)
Idents(OS) <- OS$src
DimPlot(OS)

OS <- RenameIdents(OS,
             "OS17_Culture" = "Culture",
             "OS17_Lung" = "Lung",
             "OS17_Tibia" = "Tibia",
             "t143b_Culture" = "Culture",
             "t143b_Lung" = "Lung",
             "t143b_Tibia" = "Tibia",
             "NCHOS2_Flank" = "Culture",
             "NCHOS2_Lung" = "Lung",
             "NCHOS2_Tibia" = "Tibia",
             "NCHOS7_Flank" = "Culture",
             "NCHOS7_Lung" = "Lung",
             "NCHOS7_Tibia" = "Tibia")
OS$tissue <- Idents(OS)
DimPlot(OS)


os17 <- subset(OS, cells = WhichCells(OS, expression = cond == "OS17"))
t143B <- subset(OS, cells = WhichCells(OS, expression = cond == "t143b"))
NCHOS2 <- subset(OS, cells = WhichCells(OS, expression = cond == "NCHOS2"))
NCHOS7 <- subset(OS, cells = WhichCells(OS, expression = cond == "NCHOS7"))

#extract markers
#data <- os17
#DimPlot(data, group.by = "src")

#rename Idents to src type before DGE analysis
#Idents(data, cells = WhichCells(data, expression = src == "OS17_Culture")) <- "Culture"
#Idents(data, cells = WhichCells(data, expression = src == "OS17_Lung")) <- "Lung"
#Idents(data, cells = WhichCells(data, expression = src == "OS17_Tibia")) <- "Tibia"

#data[["tissue"]] <- Idents(data)

#os17 <- data
#DimPlot(os17)

data <- list(
  os17 = os17,
  t143B = t143B,
  NCHOS2 = NCHOS2, 
  NCHOS7 = NCHOS7
)

#Lung marker - UP
Lung.up <- list(NULL)
for (i in 1:length(data)) {
  tmp <- data[[i]]
  x <- FindMarkers(tmp,
                   ident.1 = "Lung",
                   ident.2 = "Culture",
                   only.pos = TRUE,
                   min.pct = 0.1)
  Lung.up[[i]] <- rownames(x)
  }

#Tibia markers - UP 
Tibia.up <- list(NULL)
for (i in 1:length(data)) {
  tmp <- data[[i]]
  x <- FindMarkers(tmp,
                   ident.1 = "Tibia",
                   ident.2 = "Culture",
                   only.pos = TRUE,
                   min.pct = 0.1)
  Tibia.up[[i]] <- rownames(x)
}

#Lung markers - DOWN
Lung.down <- list(NULL)
for (i in 1:length(data)) {
  tmp <- data[[i]]
  x <- FindMarkers(tmp,
                   ident.1 = "Lung",
                   ident.2 = "Culture",
                   only.pos = FALSE,
                   min.pct = 0.1)
  x <- x[x$avg_log2FC<0,]
  Lung.down[[i]] <- rownames(x)
}

#Tibia markers - DOWN 
Tibia.down <- list(NULL)
for (i in 1:length(data)) {
  tmp <- data[[i]]
  x <- FindMarkers(tmp, ident.1 = "Tibia",
                   ident.2 = "Culture",
                   only.pos = FALSE,
                   min.pct = 0.1)
  x <- x[x$avg_log2FC<0,]
  Tibia.down[[i]] <- rownames(x)
}

############################Example Venn Diagram
# if (!require(devtools)) install.packages("devtools")
# devtools::install_github("yanlinlin82/ggvenn")
library(ggvenn)

# if (!require(devtools)) install.packages("devtools")
# devtools::install_github("gaospecial/ggVennDiagram")
# library("ggVennDiagram")
genes <- Lung.down
x <- list(
  A = genes[[1]],
  B = genes[[2]],
  C = genes[[3]],
  D = genes[[4]]
)

pdf("Lung.down.pdf", width = 7, height = 7)
plot <- ggvenn(
  x,
  fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF"),
  stroke_size = 0.9, set_name_size = 10,  show_percentage = TRUE
)
plot
dev.off()

# ggVennDiagram(x[1:4], label_alpha = 0.7, label = "count",
#               category.names = c("OS17",
#                                  "143B",
#                                  "NCHOS2",
#                                  "NCHOS7")) + 
#   scale_fill_gradient(low = "#F4FAFE", high = "#4981BF")+
#   theme(legend.title = element_text(color = "black"),
#         legend.position = "right")

###Extract genes that are shared in these datasets
genes <- Lung.down
A = genes[[1]]
B = genes[[2]]
C = genes[[3]]
D = genes[[4]]

# Preparing clusterProfiler to perform hypergeometric test on msigdb signatures
# m_t2g.c2 <- msigdbr(species = "Homo sapiens", category = "C2") %>%
#   dplyr::select(gs_name, human_gene_symbol)
# m_t2g.c6 <- msigdbr(species = "Homo sapiens", category = "C6") %>%
#   dplyr::select(gs_name, human_gene_symbol)
 m_t2g.h <- msigdbr(species = "Homo sapiens", category = "H") %>%
   dplyr::select(gs_name, human_gene_symbol)
 m_t2n.h <- msigdbr(species = "Homo sapiens", category = "H") %>%
   dplyr::select(gs_id, gs_name)
 #m_t2g=rbind(m_t2g.c2,m_t2g.c6)
 
# msigdb signature to use
 msig.gene.set = m_t2g.h
 msig.name = m_t2n.h

intersect <- c((intersect(intersect(intersect(A,B),C),D)), 
               (intersect(intersect(A,B),C)),
               (intersect(intersect(A,B),D)),
               (intersect(intersect(A,D),C)),
               (intersect(intersect(B,D),C)))

tmp <- enricher(intersect, TERM2GENE=msig.gene.set, TERM2NAME = msig.name)
em = tmp@result[,c("ID", "p.adjust")] #pvalue/p.adjust
rownames(em) <- NULL
em <- em[1:10,]
em[,2] <- -log10(em[,2])
# barplot(em)

pathways <- list(NULL)
pathways[[4]] <- em

# write.xlsx(pathways, file ="R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/pathways.xlsx", col.names = TRUE, row.names = TRUE, append = FALSE)
# save.image(file ="R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/AllpseudoBulk_Venn.RData")
#write.xlsx(pathways, file ="pathways.xlsx", col.names = TRUE, row.names = TRUE, append = FALSE)
#save.image(file ="AllpseudoBulk_Venn.RData")

####################Intersection of deferentially regulated genes between Bone colonization and Lung colonization
# total <- list(
#   Tibia.up = Tibia.up,
#   Tibia.down = Tibia.down,
#   Lung.up = Lung.up,
#   Lung.down = Lung.down
# )
# all.genes <- list(NULL)
# for (i in 1:4) {
#   genes <- total[[i]]
#   A = genes[[1]]
#   B = genes[[2]]
#   C = genes[[3]]
#   D = genes[[4]]
#   intersect <- intersect(intersect(intersect(A,B),C),D)
#   all.genes[[i]] <- intersect
# }
# 
# x <- list(
#   A = all.genes[[1]],
#   B = all.genes[[2]],
#   C = all.genes[[3]],
#   D = all.genes[[4]]
# )
# 
# pdf("TibiaUpdown_Lungupdown.pdf", width = 5, height = 5)
# plot <- ggvenn(
#   x,
#   fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF"),
#   stroke_size = 0.9, set_name_size = 4,  show_percentage = TRUE
# )
# plot
# dev.off()

################ Two sided Barplot
SanjanaPlots <- read.delim("SanjanaPlots.txt")

ggplot(SanjanaPlots, aes(x = -1 * Order,
                         y = p.adjust,
                         fill = p.adjust > 0)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  facet_wrap(~ Tissue) +
  scale_fill_brewer(type = "qual", palette = "Set1") +
  geom_text(aes(y = Label_y * 3, label = Pathway)) +
  theme_bw() +
  ylab("") +
  ylim(-5, 5)
```

# OS17CTL_Heterogeneity (Lineage Tagging)

```{r OS17CTL_Heterogeneity, cache.lazy=FALSE, message=TRUE}
#source("R:/RESRoberts/Bioinformatics/Analysis/Sanjana/scSeurat.R")
#source("R:/RESRoberts/Bioinformatics/Analysis/Sanjana/Downstream.v2.R")
#loading libraries
library(Seurat)
library(future)
library(ggplot2)
# library(fgsea)
library(msigdbr)
library(clusterProfiler)
library(SingleR)
library(dplyr)
library(pheatmap)
library(RColorBrewer)
library(viridis) # inferno color palette
library(grid)
source("/home/gdrobertslab/rsexf001/OSHetero2021/scSeurat_updated.R")

# Note that below is all OS-17 - Emily
# Create Seurat objects and perform initial QC.  Label original source.
# below loading is not correct and does not match the paper's figure -Emily
#cx.raw <- tenXLoadQC("/gpfs0/home2/gdrobertslab/lab/Counts/S0016xS0027/filtered_feature_bc_matrix/", spec = "mixHuman")
#cx.raw
#cx.raw <- subset(cx.raw, subset = nFeature_RNA >3500 & nCount_RNA <50000 & percent.mt <15)
#cx.raw$src <- "Culture"

#Zymo copied from RESRoberts (orginal code referenced this)
# Create Seurat objects and perform initial QC. Label original source.
cx.raw <- tenXLoadQC("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/S0016-zymo/filtered_feature_bc_matrix/", spec = "mixHuman")
cx.raw
cx.raw <- subset(cx.raw, subset = nFeature_RNA >3500 & nCount_RNA <50000 & percent.mt <15)
cx.raw$src <- "Culture"

tib.raw <- tenXLoadQC("/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
tib.raw <- subset(tib.raw, subset = nFeature_RNA >3000 & nCount_RNA <60000 & percent.mt <18)
tib.raw$src <- "Tibia"

lung.raw <- tenXLoadQC("/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
lung.raw <- subset(lung.raw, subset = nFeature_RNA >1250 & nCount_RNA <60000 & percent.mt <25)
lung.raw$src <- "Lung"

# Add lineage tracing tags to the Seurat objects
cx.raw <- processLTBC(cx.raw,
                      lt.loc = "/gpfs0/home/gdrobertslab/rsexf001/OSHetero2021/S0016-zymo/lt.fq",
                      cid.loc = "/gpfs0/home/gdrobertslab/rsexf001/OSHetero2021/S0016-zymo/cid.fq",
                      histogram = T,
                      title = "Culture, Top 40 Clones",
                      ymax = 1.75,
                      col.fill = "#E64B35FF",
                      relative = T)
tib.raw <- processLTBC(tib.raw,
                       lt.loc = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/lt.fq",
                       cid.loc = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/cid.fq",
                       histogram = T,
                       title = "Tibia, Top 40 Clones",
                       ymax = 1.75,
                       col.fill = "#00A087FF",
                       relative = T)
lung.raw <- processLTBC(lung.raw,
                        lt.loc = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/lt.fq",
                        cid.loc = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/cid.fq",
                        histogram = T,
                        title = "Lineage Barcode Enrichment by Microenvironment",
                        ymax = 1.75,
                        relative = T)


#######Export Lineage Tag enrichment plots
# Unclear where cx.raw$fig comes from -Emily 
#pdf("Culture_LT.pdf", width = 7, height = 2)
#cx.raw$fig
#dev.off()

#pdf("Tibia_LT.pdf", width = 7, height = 2)
#tib.raw$fig
#dev.off()

#pdf("Lung_LT.pdf", width = 7, height = 2)
#lung.raw$fig
#dev.off()

# Unclear what the below functions are doing, no $sobject in these objects -Emily
#cx.raw <- cx.raw$sobject
#tib.raw <- tib.raw$sobject
#lung.raw <- lung.raw$sobject
# Figure out what type of object these are now -Emily
cx.raw
tib.raw
lung.raw
os17.cx.raw
```

```{r split, cache.lazy=FALSE, dependson="OS17CTL_Heterogeneity"}
#Subset all to #2800 cells in each condition 
cx.raw <- subset(cx.raw, cells = sample(Cells(cx.raw), 2800))
tib.raw <- subset(tib.raw, cells = sample(Cells(tib.raw), 2800))
lung.raw <- subset(lung.raw, cells = sample(Cells(lung.raw), 2800))

# Merge into a single Seurat object
os17 <- merge(cx.raw, y = c(tib.raw, lung.raw),
              add.cell.ids = c("Culture", "Tibia", "Lung"),
              project = "LineageTracing")

# Process and cluster
os17 <- NormalizeData(os17) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = os.17@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

# CCR
# Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
os17 <- CellCycleScoring(object = os17, s.features = s.genes,
                         g2m.features = g2m.genes, set.ident = TRUE)

os17 <- ScaleData(object = os17, vars.to.regress = c("S.Score", "G2M.Score"),
                  features = rownames(x = os17)) 
os17 <- RunPCA(os17, pc.genes = os.17@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

# os17 <- RunPCA(os17, pc.genes = os17@var.genes, npcs = 20)
# os17 <- RunHarmony(os17, group.by.vars = "src", plot_convergence = T)
# os17 <- RunUMAP(os17, reduction = "harmony", dims = 1:20)
# os17 <- FindNeighbors(os17, reduction = "harmony", dims = 1:20)
# os17 <- FindClusters(os17, resolution = 0.3)

# pdf("Harmony.pdf", width = 7, height = 7)
# DimPlot(os17, reduction = "umap", group.by = "src", pt.size = 1, label = F, order = cell.ids) + 
#   coord_fixed() + 
#   ggtitle("OS17 by Source") + 
#   scale_color_npg()
# dev.off()

# Plot the data 
set.seed(100)
cell.ids <- sample(colnames(os17))
DimPlot(os17, reduction = "umap", group.by = "src", pt.size = 1, label = F, order = cell.ids) + 
  coord_fixed() + 
  ggtitle("OS17 by Source") + 
  scale_color_npg()
DimPlot(os17, reduction = "umap", pt.size = 1, label = T) + 
  coord_fixed() + 
  ggtitle("OS17 Clusters") + 
  scale_color_npg(alpha = 0.7)

##################Find the optimum resolution for clustering
os17.nC <- nRes(os17, 
                res = seq(from = 0.2, to = 0.3, by = 0.01))
plot <- pSil(os17.nC, 0.3)
os17.3 <- os17 %>% FindClusters(resolution = 0.3)
os17.2 <- os17 %>% FindClusters(resolution = 0.2)

# find markers for every cluster compared to all remaining cells, report only the positive ones
os17.markers <- FindAllMarkers(os17, only.pos = FALSE, min.pct = 0.25, logfc.threshold = 0.25)

#Heatmap
os17.markers %>%
  group_by(cluster) %>%
  top_n(n = 5, wt = avg_log2FC) -> top10

pdf("All.pdf", width = 15, height = 10)
p <- DoHeatmap(os17, features = top10$gene, size = 5) + NoLegend()
p + theme(text = element_text(size=20, color = "black"))
dev.off()

# Copied above to visualize in Rm html output file -Emily
p <- DoHeatmap(os17, features = top10$gene, size = 5) + NoLegend()
p + theme(text = element_text(size=20, color = "black"))
```

```{r os17, dependson="split"}
save(os17, file = "/home/gdrobertslab/rsexf001/OSHetero2021/os17.RData")
```

```{r split2_DGEA, cache.lazy=FALSE, fig.width = 15,  fig.height = 10, dependson="split"}
source("/home/gdrobertslab/rsexf001/OSHetero2021/Downstream.v2.R")
load("/home/gdrobertslab/rsexf001/OSHetero2021/os17.RData")
data <- os17
#Run through function
em.hm <- DGEA(data)

c <- rownames(em.hm)
#remove "HALLMARK_"
c <- gsub("HALLMARK_", "", c)
#remove "_" by removing special characters
c <- gsub("_", " ", c)
rownames(em.hm) <- c
em.hm

#Below line is defined in the sourced Downstream.v2.R, but the object is not created in this chunk -Emily
clust.ids <- sort(unique(data@active.ident))

col.breaks=seq(-log10(1),min(max(-log10(em.hm))+1,20),by=0.5)
col=inferno(length(col.breaks)) # library(viridis)
col=c("white",colorRampPalette(brewer.pal(n = 7, name ="Reds"))(50))

dev.cur()
#pdf("GSEA_OS17CTL.pdf", width = 15, height = 10)
#dev.cur()
pheatmap(-log10(em.hm[,1:length(clust.ids)]),
              cluster_rows = TRUE,
              cluster_cols = FALSE,
              cellwidth = 20,
              cellheight = 10,
              treeheight_row = 0,
              treeheight_col=0,
              color = col,
              scale='none',
              breaks=col.breaks,
              fontsize = 11)
#dev.off()
dev.cur()

#########################Add module score for known markers 
# Osteoblastic cells 	    RUNX2, COL1A1, CDH11, IBSP 
# Chondroblastic cells 	  SOX9, ACAN, PTH1R 
# Osteoclasts 	          ACP5, CTSK, MMP9 
# Myeloid cells 	        CD74, CD14, FCGR3A 
# Mesenchymal stem cells 	MME, THY1, CXCL12, SFRP2 
# Proliferating cells 	  MKI67, TOP2A, PCNA 


Ob_features <- list(c('RUNX2', 'COL1A1', 'CDH11', 'IBSP'))
Cb_features <- list(c('SOX9', 'ACAN', 'PTH1R'))
Oc_features <- list(c('ACP5', 'CTSK', 'MMP9'))
MSC_features <- list(c('MME', 'THY1', 'CXCL12', 'SFRP2'))

#data <- os17

data <- AddModuleScore(data, features = Ob_features, ctrl = 5, name = 'Ob_features')
data <- AddModuleScore(data, features = Cb_features, ctrl = 5, name = 'Cb_features')
data <- AddModuleScore(data, features = Oc_features, ctrl = 5, name = 'Oc_features')
data <- AddModuleScore(data, features = MSC_features, ctrl = 5, name = 'MSC_features')

# FeaturePlot(data, features = "Ob_features1", min.cutoff = 0)
# FeaturePlot(data, features = "Cb_features1")
# FeaturePlot(data, features = "Oc_features1")
# FeaturePlot(data, features = "MSC_features1")

RidgePlot(data, features = c("Ob_features1", "Cb_features1", "Oc_features1", "MSC_features1"))
# FeatureScatter(seurat, "Ob_features", "nFeature_RNA")
```

# OS17CTL_Heterogeneity_LT (Lineage Tracing in DimPlot)

```{r OS17CTL_Heterogeneity_LT, cache.lazy=FALSE, dependson="split"}
source("/home/gdrobertslab/rsexf001/OSHetero2021/scSeurat_updated.R")
load("/home/gdrobertslab/rsexf001/OSHetero2021/os17.RData")
# Extract the barcode frequency lists
cx.lt.list <- processLTBC(cx.raw,
                          lt.loc = "/gpfs0/home/gdrobertslab/rsexf001/OSHetero2021/S0016-zymo/lt.fq",
                          cid.loc = "/gpfs0/home/gdrobertslab/rsexf001/OSHetero2021/S0016-zymo/cid.fq",
                          ret.list = TRUE)
tib.lt.list <- processLTBC(tib.raw,
                           lt.loc = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/lt.fq",
                           cid.loc = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/cid.fq",
                           ret.list = TRUE)
lung.lt.list <- processLTBC(lung.raw,
                            lt.loc = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/lt.fq",
                            cid.loc = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/cid.fq",
                            ret.list = TRUE)

# Subset the three samples, but retain the current umap data
cx.sub <- subset(os17, subset = src == "Culture")
tib.sub <- subset(os17, subset = src == "Tibia")
lung.sub <- subset(os17, subset = src == "Lung")
```

```{r ostestEm, cache.lazy=FALSE}
Idents(os17) <- os17$RNA_snn_res.0.3
DimPlot(os17, label = T, split.by="src") + coord_fixed()

# Subset the three samples, but retain the current umap data
cx.sub <- subset(os17, subset = src == "Culture")
tib.sub <- subset(os17, subset = src == "Tibia")
lung.sub <- subset(os17, subset = src == "Lung")

DimPlot(lung.sub, label = T) + coord_fixed()
```

# OS17CTL_Heterogeneity_LTpathway
Had to place this code midway through the previous chunk (OS17CTL_Heterogeneity_LT) in order to define certain objects used later in the chunk.

```{r OS17CTL_Heterogeneity_LTpathway, cache.lazy=FALSE, dependson="OS17CTL_Heterogeneity_LT"}
####################subset to cells that have a lineage tag
#memory.limit(size = 40000)
source("/home/gdrobertslab/rsexf001/OSHetero2021/Downstream.v2.R")
#library(xlsx)
#Culture
Culture.LTcells <- vector(mode = "character")
for(i in 1:nrow(cx.lt.list)){
  temp <- WhichCells(cx.sub, expression = lt == cx.lt.list[[i,1]]) 
  Culture.LTcells <- append(Culture.LTcells, temp)
}
cx.tagged <- subset(cx.sub, cells = Culture.LTcells)

#Tibia
tib.LTcells <- vector(mode = "character")
for(i in 1:nrow(tib.lt.list)){
  temp <- WhichCells(tib.sub, expression = lt == tib.lt.list[[i,1]]) 
  tib.LTcells <- append(tib.LTcells, temp)
}
tib.tagged <- subset(tib.sub, cells = tib.LTcells)

#Lung
lung.LTcells <- vector(mode = "character")
for(i in 1:nrow(lung.lt.list)){  #originally said 1:515, unclear why and received a "subscript out of bounds" error, so changed to match the previous code for cx and tib -Emily
  temp <- WhichCells(lung.sub, expression = lt == lung.lt.list[[i,1]]) 
  lung.LTcells <- append(lung.LTcells, temp)
}
lung.tagged <- subset(lung.sub, cells = lung.LTcells)

lung.sub <- subset(lung.sub, idents = c("0", "1", "3", "6")) ##Remove outlier cells in cluster 2 and 5 # Had to edit some of these cluster names (likely due to random seed changes in plotting) -Emily

pdf("LungOutline.pdf", height = 5, width = 5)
DimPlot(lung.sub, reduction = "umap", pt.size = 1, label = TRUE) + 
  coord_fixed() + 
  ggtitle("Lung-derived") +
  #xlim(0,9) +
  NoAxes() + 
  scale_color_npg(alpha = 1)
dev.off()

# Show in Rm html file -Emily
DimPlot(lung.sub, reduction = "umap", pt.size = 1, label = TRUE) + 
  coord_fixed() + 
  ggtitle("Lung-derived") +
  #xlim(0,9) +
  NoAxes() + 
  scale_color_npg(alpha = 1)

#Lung
vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
pdf("grid_lung_v4.pdf", width = 10, height = 10)
grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 4)))
for (i in 1:4) {
  p <- DimPlot(lung.tagged, 
               reduction = "umap",
               pt.size = 1,
               cells.highlight = WhichCells(lung.sub, expression = lt == lung.lt.list[[i,1]]), 
               cols.highlight = "#4DBBD5FF",
               sizes.highlight = 3) + 
    coord_fixed() +
    theme(legend.position = "none") +
    ggtitle(paste("Clone", lung.lt.list[[i,1]])) +
    #xlim(0,9) +
    NoLegend() +
    NoAxes()
  print(p, vp=vplayout(ceiling(i/4), i))
}
dev.off()
dev.cur() # Emily added

#cell cycle distribution of top 4 lineages - change to Pie chart and percentage

vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
pdf("pie.pdf", width = 15, height = 15)
grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 4)))
for(i in 1:4){
  temp <- WhichCells(lung.sub, expression = lt == lung.lt.list[[i,1]]) 
  LT.cells <- subset(lung.sub, cells = temp)
  df <- table(LT.cells$Phase)
  bp <- ggplot(as.data.frame(df), aes(x="", y=Freq, fill=Var1))+
    geom_bar(width = 1, stat = "identity")
  pie <- bp + coord_polar("y", start=0) + scale_fill_brewer(palette="Blues")+
    theme_minimal() + NoAxes() +
    ggtitle(paste("Clone", lung.lt.list[[i,1]]))
  print(pie, vp=vplayout(ceiling(i/4), i))
}
dev.off()
dev.cur() # Emily added 

#barplot
bp <-  ggplot(as.data.frame(df), aes(x = Var1, y = Freq)) +
  geom_bar(fill = "#4DBBD5FF", stat = "identity") +
  ggtitle("Cell cycle distribution") +
  ylab("Count") +
  xlab("Phase")
bp
#######################Pathways activated in top 10 lineages

lung.lt.list[1:10,1]
# 052-082154 
# 039-074146 
# 037-078792 
# 039-052640 
# 095-044696 
# 039-098388 
# 064-065939 
# 013-044371 
# 063-007191 
# 018-082025

data <- lung.tagged

#Subset cells of each lineage into a separate cluster in lung tagged
for(i in 1:10) {
  cells <- WhichCells(data, expression = lt == lung.lt.list[[i,1]])
  Idents(object = data, cells = cells) <- paste("Enriched Lineage", i, sep = " ")
}

library(forcats)
Idents(data) <- fct_rev(Idents(data))

# Idents(data) <- data$seurat_clusters
# Stash cell identity classes
data[["Lineage"]] <- Idents(data)

DimPlot(data)
#reorganize Lineage levels


#Top markers for each lineage

# find markers for every cluster compared to all remaining cells, report positive and negative ones 
# Changed obj names to indicate pos vs neg markers -Emily
data.markers.neg <- FindAllMarkers(data, only.pos = FALSE, min.pct = 0.25, logfc.threshold = 0.25)
data.markers.pos <- FindAllMarkers(data, only.pos = TRUE)

#Run through function
em.hm <- DGEA(data)

#Remove non-LT clusters
em.hm.lt <- em.hm[-c(1:4)]
clust.ids = sort(unique(data@active.ident))
clust.ids <- clust.ids[-c(1:4)]
c <- rownames(em.hm.lt)
#remove "HALLMARK_"
c <- gsub("HALLMARK_", "", c)
#remove "_" by removing special characters
c <- gsub("_", " ", c)
rownames(em.hm.lt) <- c
#remove rows with low pvalues
row_names_df_to_remove<-c("FATTY ACID METABOLISM",
                          "ADIPOGENESIS",
                          "PEROXISOME",
                          "SPERMATOGENESIS",
                          "XENOBIOTIC METABOLISM",
                          "PI3K AKT MTOR SIGNALING",
                          "HEDGEHOG SIGNALING")
em.hm.lt <- em.hm.lt[!(row.names(em.hm.lt) %in% row_names_df_to_remove),]

col.breaks=seq(-log10(1),min(max(-log10(em.hm.lt))+1,20),by=0.5)
col=inferno(length(col.breaks)) # library(viridis)
col=c("white",colorRampPalette(brewer.pal(n = 7, name ="Reds"))(50))
p <- pheatmap(-log10(em.hm.lt[,1:length(clust.ids)]),
         cluster_rows = TRUE,
         cluster_cols = FALSE,
         cellwidth = 20,
         cellheight = 10,
         treeheight_row = 0,
         treeheight_col=0,
         color = col,
         scale='none',
         breaks=col.breaks,
         fontsize = 11)

pdf("EnrichedLTHeatmap.pdf", width = 9, height = 10)
print(p)
dev.off()
p #Emily

########################Run IPA on the DEGs from top ten lineages
#NOTE: After Enriched lineage 5, not more than 25 DEGs with P_adj_val < 0.01

# L1.markers <- FindMarkers(data, ident.1 = "Enriched Lineage 1", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
# L2.markers <- FindMarkers(data, ident.1 = "Enriched Lineage 2", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
# L3.markers <- FindMarkers(data, ident.1 = "Enriched Lineage 3", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
# L4.markers <- FindMarkers(data, ident.1 = "Enriched Lineage 4", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
# L5.markers <- FindMarkers(data, ident.1 = "Enriched Lineage 5", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
# L6.markers <- FindMarkers(data, ident.1 = "Enriched Lineage 6", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
# L7.markers <- FindMarkers(data, ident.1 = "Enriched Lineage 7", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
# L8.markers <- FindMarkers(data, ident.1 = "Enriched Lineage 8", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
# L9.markers <- FindMarkers(data, ident.1 = "Enriched Lineage 9", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
# L10.markers <- FindMarkers(data, ident.1 = "Enriched Lineage 10", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
# 
# #remove unnecessary columns
# myList <- list(L1.markers, L2.markers, L3.markers, L4.markers, L5.markers, L6.markers, L7.markers, L8.markers, L9.markers, L10.markers)
# myList <- lapply(myList, function(x) { x["p_val"] <- NULL; x[, 2:3] <- NULL; x })
# 
# #write markers to excel sheet
# names <- c("L1.markers", "L2.markers", "L3.markers", "L4.markers", "L5.markers", "L6.markers", "L7.markers", "L8.markers", "L9.markers", "L10.markers")
# library(xlsx)
# for (i in 1:10){
#   location <- paste0("R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/L",i,".markers.xlsx", sep = "")
#   write.xlsx(myList[[i]], file = location, 
#              col.names = TRUE, row.names = TRUE, append = FALSE)
# } 

#######################Gllycolysis and Hypoxia in LT enriched cells
#load("R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/Preprocessed_Seurat_objects/OS17_Zymo_CTL.RData")
#load("R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/OS17CTL.sub.RData")

DimPlot(lung.tagged, reduction = "umap", pt.size = 1, label = TRUE) + 
  coord_fixed() + 
  ggtitle("Lung-derived") +
  xlim(0,9) + NoAxes()+ 
  scale_color_npg(alpha = 1)

data <- lung.tagged

#Apply Glycolysis module score to each of the primary tumor cells
gl <- list()
Glycolysis <- read.table(file = "/home/gdrobertslab/rsexf001/OSHetero2021/pathways/Glycolysis.txt", header = FALSE, sep = ",")
Glycolysis <- Glycolysis[,1]
OxPhos <- read.table(file = "/home/gdrobertslab/rsexf001/OSHetero2021/pathways/OxPhos.txt", header = FALSE, sep = ",")
OxPhos <- OxPhos[,1]
Hypoxia <- read.table(file = "/home/gdrobertslab/rsexf001/OSHetero2021/pathways/Hypoxia.txt", header = FALSE, sep = ",")
Hypoxia <- Hypoxia[,1]
SASP <- read.table(file = "/home/gdrobertslab/rsexf001/OSHetero2021/pathways/SASP.txt", header = FALSE, sep = ",")
SASP <- SASP[,1]
EMT <- read.table(file = "/home/gdrobertslab/rsexf001/OSHetero2021/pathways/EMT.txt", header = FALSE, sep = ",")
EMT <- EMT[,1]
TNFA <- read.table(file = "/home/gdrobertslab/rsexf001/OSHetero2021/pathways/TNFA.txt", header = FALSE, sep = ",")
TNFA <- TNFA[,1]
  
gl$Glycolysis <- Glycolysis
gl$OxPhos <- OxPhos
gl$Hypoxia <- Hypoxia
gl$SASP <- SASP
gl$EMT <- EMT
gl$TNFA <- TNFA
mod_names <- names(gl)

#Added min.cutoff to isolate for cells most module like
for(i in 1:length(mod_names)) {
  data <- AddModuleScore(data, gl[i], name = names(gl[i]))
  p <- FeaturePlot(data, features = str_c(names(gl[i]), "1"), min.cutoff = 0.1, pt.size = 1,
                   order = T, cols = c("lightgoldenrod", "darkred"))
  pdf(paste0("FeaturePlot ",names(gl[i]), ".pdf"))
  print(p)
  dev.off()
}
p #Emily

#Compare Enriched vs non-enriched
#Glycolysis and Hypoxia seem to be trending together

#Group first 12 against the tagged Lung
data <- lung.tagged
LT <- vector(mode = "character")
for(i in 1:10){
  temp <- WhichCells(lung.sub, expression = lt == lung.lt.list[[i,1]]) 
  LT <- append(LT, temp)
}
Idents(data, cells = LT) <- 'Enriched'

#Non-enriched cells are clones with a frequency of 1
# lung.lt.list[134,] 
# 2
# lung.lt.list[134,] 
# 1
LT <- vector(mode = "character")
for(i in 135:377){
  temp <- WhichCells(lung.sub, expression = lt == lung.lt.list[[i,1]])
  LT <- append(LT, temp)
}
Idents(data, cells = LT) <- 'Non-enriched'

cell.ids <- sample(colnames(data))

#Culture plots
# pdf("OS17_Enr.pdf", width = 5, height = 5)
DimPlot(data, 
             reduction = "umap",
             pt.size = 1,
             cells.highlight = WhichCells(data, idents = "Enriched"), 
             cols.highlight = "#E64B35FF",
             sizes.highlight = 2, order = cell.ids) + 
  coord_fixed() 
# dev.off()
####################Pathways associated with Enriched clonal families
library(tibble)
library(tidyr)
library(ggplot2)
library(tidyverse)
library(stringr)

B.list <- list(OS17 = data)

em.hm.list <- list()
for (i in 1:length(B.list)) {
  em.hm.list[[i]] <- DGEA(B.list[[i]]) 
}

temp1 <- em.hm.list[[1]]
temp1 <- -log10(temp1)
#remove "_" by removing special characters
c <- rownames(temp1)
c <- gsub("HALLMARK ", "", c)
c <- gsub("_", " ", c)
rownames(temp1) <- c
#remove negative 0 values due to number of decimal points displayed
temp1 <- abs(temp1)
temp1 <- temp1["Enriched"]

#Subset to only pathways with siginificant pvalue (-1 * log10(0.05)) = 1.30103
temp1["Enriched"] >= 1.30103 -> temp1$logical
colnames(temp1) <- c("pvalue", "significant")
logical <- temp1$significant
temp1 <- temp1[logical,]
temp1 <- rownames_to_column(temp1, var = "Pathway")
position <- order(temp1$pvalue, decreasing = TRUE)
temp1 <- temp1[position,]
plot <- ggplot(temp1, aes(x = pvalue,
                         y = Pathway)) +
  geom_bar(stat = "identity") +
  scale_fill_brewer(type = "qual", palette = "Set1") +
  theme_bw() +
  ylab("") 

pdf("Enr_Pathway.pdf", height = 5, width = 10)
plot(plot)
dev.off()
plot #Emily

# heatmap <- temp1 %>%
#   t() %>%
#   as.data.frame() %>%
#   rownames_to_column(var = "Sample") %>%
#   left_join(aka4 %>% rownames_to_column(var = "Sample")) %>%
#   pivot_longer(c(-Sample, -Lpercent),
#                names_to = "Pathway",
#                values_to = "pval",
#                names_repair = "minimal") %>%
#   mutate(Signif = pval >= (-1 * log10(0.05))) %>%
#   ggplot(., aes(x = Sample, y = Pathway, fill = Signif)) + 
#   geom_tile() +
#   geom_text(aes(label = sprintf("%0.2f", pval))) +
#   scale_fill_manual(values = c("gray", "red")) +
#   scale_y_discrete(limits = rev) +
#   theme(legend.position = "none",
#         panel.grid.major = element_blank(),
#         panel.grid.minor = element_blank(),
#         axis.ticks.x = element_blank()) +
#   ylab("") 
# 
# pdf("Enr_Pathway.pdf", height = 8, width = 8)
# plot(heatmap)
# dev.off()

# Find differentially expressed features between Enriched and remaining Lung
enr.markers <- FindMarkers(data, 
                           ident.1 = "Enriched",
                           logfc.threshold = 0.25, 
                           only.pos = TRUE, 
                           test.type= "DESeq2")

lung.markers <- FindMarkers(data, 
                            ident.1 = "Non-enriched", 
                            logfc.threshold = 0.25, 
                            only.pos = TRUE, 
                            test.type= "DESeq2")


# view results
#write.xlsx(enr.markers, file = "enr.10_markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(lung.markers, file = "lung_markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)

VlnPlot(object = data, idents = c('Enriched', 'Non-enriched'), features = c('ENO2','PGK1','ENO1'), combine = TRUE, ncol = 3)
VlnPlot(object = data, idents = c('Enriched', 'Non-enriched'), features = c('IFITM3','ISG15','IFI6'), combine = TRUE, ncol = 3)
```


# OS17CTL_Heterogeneity_LT (Lineage Tracing in DimPlot) continued

```{r OS17CTL_Heterogeneity_LT_cont}
# Generate cluster identification plots for the top 4 enriched clones from each sample
# Use colors from the npg palatte in ggsci
library("grid")
#Culture
vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
pdf("OS17CTL_Heterogeneity_LT_cont/grid_cx_v1.pdf", width = 10, height = 10)
grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 4)))
for (i in 1:4) {
  p <- DimPlot(cx.tagged, 
               reduction = "umap",
               pt.size = 1,
               cells.highlight = WhichCells(cx.sub, expression = lt == cx.lt.list[[i,1]]), 
               cols.highlight = "#E64B35FF",
               sizes.highlight = 3, order = cell.ids) + 
    coord_fixed() +
    theme(legend.position = "none") +
    ggtitle(paste("Clone", cx.lt.list[[i,1]])) +
    #xlim(-12,-6) +
    NoLegend() +
    NoAxes()
  print(p, vp=vplayout(ceiling(i/4), i))
}
dev.off()

#Tibia
vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
pdf("OS17CTL_Heterogeneity_LT_cont/grid_tib_v1.pdf", width = 10, height = 10)
grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 4)))
for (i in 1:4) {
  p <- DimPlot(tib.tagged, 
               reduction = "umap",
               pt.size = 1,
               cells.highlight = WhichCells(tib.sub, expression = lt == tib.lt.list[[i,1]]), 
               cols.highlight = "#00A087FF",
               sizes.highlight = 3) + 
    coord_fixed() +
    theme(legend.position = "none") +
    ggtitle(paste("Clone", tib.lt.list[[i,1]])) +
    #xlim(0,9) +
    NoLegend() +
    NoAxes()
  print(p, vp=vplayout(ceiling(i/4), i))
}
dev.off()

#Lung
vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
pdf("OS17CTL_Heterogeneity_LT_cont/grid_lung_v4.pdf", width = 10, height = 10)
grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 4)))
for (i in 1:4) {
  p <- DimPlot(lung.tagged, 
               reduction = "umap",
               pt.size = 1,
               cells.highlight = WhichCells(lung.sub, expression = lt == lung.lt.list[[i,1]]), 
               cols.highlight = "#4DBBD5FF",
               sizes.highlight = 3) + 
    coord_fixed() +
    theme(legend.position = "none") +
    ggtitle(paste("Clone", lung.lt.list[[i,1]])) +
    #xlim(0,9) +
    NoLegend() +
    NoAxes()
  print(p, vp=vplayout(ceiling(i/4), i))
}
dev.off()

#Culture plots
pdf("OS17CTL_Heterogeneity_LT_cont/Culture.pdf", width = 2.3, height = 2.3)
pcx <- DimPlot(cx.sub, reduction = "umap", pt.size = 1, label = F, group.by = "src") + 
  coord_fixed() + 
  ggtitle("Plate-derived") +
  #xlim(-12,-6) +
  NoLegend() +
  NoAxes()
pcx + scale_color_manual(values = c("#E64B35FF"))
dev.off()
#Emily for html:
pcx + scale_color_manual(values = c("#E64B35FF"))

pdf("OS17CTL_Heterogeneity_LT_cont/Tibia.pdf", width = 2.3, height = 2.3)
ptib <- DimPlot(tib.sub, reduction = "umap", pt.size = 1, label = F, group.by = "src") + 
  coord_fixed() + 
  ggtitle("Bone-derived")+
  #xlim(0,9) +
  NoLegend() +
  NoAxes() 
ptib + scale_color_manual(values = c("#00A087FF"))
dev.off()
#Emily for html:
ptib + scale_color_manual(values = c("#00A087FF"))

pdf("OS17CTL_Heterogeneity_LT_cont/Lung.pdf", width = 2.3, height = 2.3)
plung <- DimPlot(lung.sub, reduction = "umap", pt.size = 1, label = F, group.by = "src") + 
  coord_fixed() + 
  ggtitle("Lung-derived") +
  #xlim(0,9) +
  NoLegend() +
  NoAxes()
plung + scale_color_manual(values = c("#4DBBD5FF"))
dev.off()
#Emily for html:
plung + scale_color_manual(values = c("#4DBBD5FF"))

#Together plot
pdf("OS17CTL_Heterogeneity_LT_cont/All.pdf", width = 7, height = 7)
DimPlot(os17, reduction = "umap", group.by = "src", pt.size = 1, label = T) + 
  coord_fixed() +  
  scale_color_npg(alpha = 1)
dev.off()
#Emily for html:
DimPlot(os17, reduction = "umap", group.by = "src", pt.size = 1, label = T) + 
  coord_fixed() +  
  scale_color_npg(alpha = 1)

DimPlot(os17, reduction = "umap", group.by = "src", pt.size = 1, label = T) + 
  coord_fixed() + 
  ggtitle("OS17 Clusters") + 
  scale_color_npg(alpha = 0.7)
```

# TL_OS17_143B_OS2_OS7 (Longitudinal Study)

from Longitudinal Study directory 

```{r TL_OS17_143B_OS2_OS7, cache.lazy=FALSE}
#source("R:/RESRoberts/Bioinformatics/Analysis/scSeurat.R")
#source("R:/RESRoberts/Bioinformatics/Analysis/Sanjana/Downstream.v2.R")
#loading libraries
#library(Seurat)
#library(future)
#library(ggplot2)
## library(fgsea)
#library(msigdbr)
#library(clusterProfiler)
#library(SingleR)
#library(dplyr) # %<%
#library(pheatmap)
#library(RColorBrewer)
#library(viridis) # inferno color palette
#library(grid)

#Merge by condition
#OS17
#os17.tib.raw <- tenXLoadQC("R:/RESRoberts/Bioinformatics/scRNAOuts/S0018xS0028/filtered_feature_bc_matrix/", spec = "mixHuman")
#os17.tib.raw <- subset(os17.tib.raw, subset = nFeature_RNA >3000 & nCount_RNA <60000 & percent.mt <18)
os17.tib.raw$src <- "OS17_Tibia"
os17.tib.raw$cond <- "OS17"

#os17.lung.raw <- tenXLoadQC("R:/RESRoberts/Bioinformatics/scRNAOuts/S0024xS0029/filtered_feature_bc_matrix/", spec = "mixHuman")
#os17.lung.raw <- subset(os17.lung.raw, subset = nFeature_RNA >1250 & nCount_RNA <60000 & percent.mt <25)
os17.lung.raw$src <- "OS17_Lung"
os17.lung.raw$cond <- "OS17"

  #Subset number of cells in each Seurat object to the least in the group
  a <- table(os17.tib.raw$orig.ident)
  b <- table(os17.lung.raw$orig.ident)
  
  list <- c(a, b)
  min <- min(list)
  
  os17.tib.raw <- subset(os17.tib.raw, cells = sample(Cells(os17.tib.raw), min))
  os17.lung.raw <- subset(os17.lung.raw, cells = sample(Cells(os17.lung.raw), min))

#t143b
#t143b.cx.raw <- tenXLoadQC(path10x = "R:/RESRoberts/Bioinformatics/scRNAOuts/S0017-143b/filtered_feature_bc_matrix/", spec = "mixHuman")
#t143b.cx.raw <- subset(t143b.cx.raw, subset = nFeature_RNA >4000 & nCount_RNA <74000 & percent.mt <17)
#VlnPlot(t143b.cx.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
t143b.cx.raw$src <- "t143b_Cx"
t143b.cx.raw$cond <- "t143b"

#t143b.tib.raw <- tenXLoadQC(path10x = "R:/RESRoberts/Bioinformatics/scRNAOuts/S0052-143B-Tibia/filtered_feature_bc_matrix/", spec = "mixHuman")
#t143b.tib.raw <- subset(t143b.tib.raw, subset = nFeature_RNA >1000 & nCount_RNA <30000 & percent.mt <22)
#VlnPlot(t143b.tib.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
t143b.tib.raw$src <- "t143b_Tibia"
t143b.tib.raw$cond <- "t143b"

#t143b.lung.raw <- tenXLoadQC(path10x = "R:/RESRoberts/Bioinformatics/scRNAOuts/S0019-143B-lung/filtered_feature_bc_matrix/", spec = "mixHuman")
#t143b.lung.raw <- subset(t143b.lung.raw, subset = nFeature_RNA >1000 & nCount_RNA <30000 & percent.mt <22)
#VlnPlot(t143b.lung.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
t143b.lung.raw$src <- "t143b_Lung"
t143b.lung.raw$cond <- "t143b"

  #Subset number of cells in each Seurat object to the least in the group
  a <- table(t143b.tib.raw$orig.ident)
  b <- table(t143b.lung.raw$orig.ident)
  
  list <- c(a, b)
  min <- min(list)
  
  t143b.tib.raw <- subset(t143b.tib.raw, cells = sample(Cells(t143b.tib.raw), min))
  t143b.lung.raw <- subset(t143b.lung.raw, cells = sample(Cells(t143b.lung.raw), min))

#NCHOS2
#OS2.tib.raw<- tenXLoadQC(path10x = "R:/RESRoberts/Bioinformatics/scRNAOuts/S0042-NCHOS2-tibia/filtered_feature_bc_matrix/", spec = "mixHuman")
#OS2.tib.raw <- subset(OS2.tib.raw, subset = nFeature_RNA >2500 & nCount_RNA <40000 & percent.mt <20)
OS2.tib.raw$src <- "NCHOS2_Tibia"
#VlnPlot(OS2.tib.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS2.tib.raw$cond <- "NCHOS2"

#OS2.lung.raw<- tenXLoadQC(path10x = "R:/RESRoberts/Bioinformatics/scRNAOuts/S0041-NCHOS2-lung/filtered_feature_bc_matrix/", spec = "mixHuman")
#OS2.lung.raw <- subset(OS2.lung.raw, subset = nFeature_RNA >2500 & nCount_RNA <60000 & percent.mt <30)
OS2.lung.raw$src <- "NCHOS2_Lung"
#VlnPlot(OS2.lung.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS2.lung.raw$cond <- "NCHOS2"

  #Subset number of cells in each Seurat object to the least in the group
  a <- table(OS2.tib.raw$orig.ident)
  b <- table(OS2.lung.raw$orig.ident)
  
  list <- c(a, b)
  min <- min(list)
  
  OS2.tib.raw <- subset(OS2.tib.raw, cells = sample(Cells(OS2.tib.raw), min))
  OS2.lung.raw <- subset(OS2.lung.raw, cells = sample(Cells(OS2.lung.raw), min))

#NCHOS7
#OS7.tib.raw<- tenXLoadQC(path10x = "R:/RESRoberts/Bioinformatics/scRNAOuts/S0034-NCHOS7-tib/filtered_feature_bc_matrix/", spec = "mixHuman")
#OS7.tib.raw <- subset(OS7.tib.raw, subset = nFeature_RNA >2000 & nCount_RNA <35000 & percent.mt <14)
OS7.tib.raw$src <- "NCHOS7_Tibia"
#VlnPlot(OS7.tib.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS7.tib.raw$cond <- "NCHOS7"

#OS7.lung.raw<- tenXLoadQC(path10x = "R:/RESRoberts/Bioinformatics/scRNAOuts/S0055-NCHOS7-lung/filtered_feature_bc_matrix/", spec = "mixHuman")
#OS7.lung.raw <- subset(OS7.lung.raw, subset = nCount_RNA <42000 & percent.mt <20)
OS7.lung.raw$src <- "NCHOS7_Lung"
#VlnPlot(OS7.lung.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS7.lung.raw$cond <- "NCHOS7"

  #Subset number of cells in each Seurat object to the least in the group
  a <- table(OS7.tib.raw$orig.ident)
  b <- table(OS7.lung.raw$orig.ident)
  
  list <- c(a, b)
  min <- min(list)
  
  OS7.tib.raw <- subset(OS7.tib.raw, cells = sample(Cells(OS7.tib.raw), min))
  OS7.lung.raw <- subset(OS7.lung.raw, cells = sample(Cells(OS7.lung.raw), min))


# Merge into a single Seurat object
OS17.TL <- merge(os17.tib.raw, y = c(os17.lung.raw),
                   add.cell.ids = c("OS17_Tibia", "OS17_Lung"),
                   project = "Heterogeneity")

t143b <- merge(t143b.tib.raw, y = c(t143b.lung.raw),
               add.cell.ids = c("t143b_Tibia", "t143b_Lung"),
               project = "Heterogeneity")
  
OS2 <- merge(OS2.tib.raw, y = c(OS2.lung.raw),
            add.cell.ids = c("OS2_Tibia", "OS2_Lung"),
            project = "Heterogeneity")

OS7 <- merge(OS7.tib.raw, y = c(OS7.lung.raw),
             add.cell.ids = c("OS7_Tibia", "OS7_Lung"),
             project = "Heterogeneity")

# Process and cluster
OS17.TL <- NormalizeData(OS17.TL) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = OS17.TL@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

t143b <- NormalizeData(t143b) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = t143b@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

OS2 <- NormalizeData(OS2) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = OS2@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

OS7 <- NormalizeData(OS7) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = OS7@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)
```

```{r TL_OS17_143B_OS2_OS7_ccr, cache.lazy=FALSE}
# CCR
# Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
#OS17
OS17.TL <- CellCycleScoring(object = OS17.TL, s.features = s.genes,
                            g2m.features = g2m.genes, set.ident = TRUE)

OS17.TL <- ScaleData(object = OS17.TL, vars.to.regress = c("S.Score", "G2M.Score"),
                     features = rownames(x = OS17.TL)) 
OS17.TL <- RunPCA(OS17.TL, pc.genes = OS17.TL@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

#t143b
t143b <- CellCycleScoring(object = t143b, s.features = s.genes,
                            g2m.features = g2m.genes, set.ident = TRUE)

t143b <- ScaleData(object = t143b, vars.to.regress = c("S.Score", "G2M.Score"),
                     features = rownames(x = t143b)) 
t143b <- RunPCA(t143b, pc.genes = t143b@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

#OS2
OS2 <- CellCycleScoring(object = OS2, s.features = s.genes,
                       g2m.features = g2m.genes, set.ident = TRUE)

OS2 <- ScaleData(object = OS2, vars.to.regress = c("S.Score", "G2M.Score"),
                features = rownames(x = OS2)) 
OS2 <- RunPCA(OS2, pc.genes = OS2@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

#OS7
OS7 <- CellCycleScoring(object = OS7, s.features = s.genes,
                       g2m.features = g2m.genes, set.ident = TRUE)

OS7 <- ScaleData(object = OS7, vars.to.regress = c("S.Score", "G2M.Score"),
                features = rownames(x = OS7)) 
OS7 <- RunPCA(OS7, pc.genes = OS7@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

# Plot the data 
DimPlot(OS17.TL, reduction = "umap", group.by = "src", pt.size = 1, label = T) + 
  coord_fixed() + 
  ggtitle("OS17.TL by Source") + 
  scale_color_npg(alpha = 0.7)
DimPlot(OS17.TL, reduction = "umap", pt.size = 1, label = T, split.by = "src") + 
  coord_fixed() + 
  ggtitle("OS17 Clusters") + 
  scale_color_npg(alpha = 0.7)

DimPlot(t143b, reduction = "umap", group.by = "src", pt.size = 1, label = T) + 
  coord_fixed() + 
  ggtitle("t143b by Source") + 
  scale_color_npg(alpha = 0.7)
DimPlot(t143b, reduction = "umap", pt.size = 1, label = T, split.by = "src") + 
  coord_fixed() + 
  ggtitle("t143b Clusters") + 
  scale_color_npg(alpha = 1)

DimPlot(OS2, reduction = "umap", group.by = "src", pt.size = 1, label = T) + 
  coord_fixed() + 
  ggtitle("OS by Source") + 
  scale_color_npg(alpha = 0.7)
DimPlot(OS2, reduction = "umap", pt.size = 1, label = T, split.by = "src") + 
  coord_fixed() + 
  ggtitle("OS Clusters") + 
  scale_color_npg(alpha = 0.7)

DimPlot(OS7, reduction = "umap", group.by = "src", pt.size = 1, label = T) + 
  coord_fixed() + 
  ggtitle("OS by Source") + 
  scale_color_npg(alpha = 0.7)
DimPlot(OS7, reduction = "umap", pt.size = 1, label = T, split.by = "src") + 
  coord_fixed() + 
  ggtitle("OS Clusters") + 
  scale_color_npg(alpha = 0.7)

#save.image(file = "OS_All.RData")
```

```{r TL_OS17_143B_OS2_OS7_2, cache.lazy=FALSE}
#Plot the cluster distribution in each sample
#temp <- colSums(table(Idents(OS2), OS2$src))
#min <- min(temp)
#cell.ids <- sample(colnames(OS2[["OS2"]]))

table(Idents(OS17.TL), OS17.TL$src)
prop <- prop.table(table(Idents(OS17.TL), OS17.TL$src), margin = 2)*100
colSums(prop)
prop
#     Tibia       Lung
# 0   50.263250   46.577747
# 1   43.453843   10.179010
# 2   1.053001   39.698140
# 3   5.229905   3.545104

table(Idents(t143b), t143b$src)
prop <- prop.table(table(Idents(t143b), t143b$src), margin = 2)*100
colSums(prop)
prop
#     Tibia        Lung
# 0   97.8963771    2.2204908
# 1   1.1492014   92.3061940
# 2   0.4479938   3.3502143
# 3   0.5064277   2.1231009

table(Idents(OS2), OS2$src)
prop <- prop.table(table(Idents(OS2), OS2$src), margin = 2)*100
colSums(prop)
prop
#     Tibia       Lung
# 0   82.281553   20.327670
# 1   5.339806    71.480583
# 2   12.378641   8.191748

table(Idents(OS7), OS7$src)
prop <- prop.table(table(Idents(OS7), OS7$src), margin = 2)*100
colSums(prop)
prop
#     Tibia       Lung
# 0    0.000000   71.013490
# 1   59.425804    1.383604
# 2   40.574196    1.176064
# 3    0.000000   26.426842

# Commented out by Emily
#save(list = c(cx.sub, lung.sub, tib.sub, os17), 
#           file = "TL_OS17_143B_OS2_OS7_2/OS17CTL.sub.RData")
#save(list = c(os17.tib.raw, os17.lung.raw, OS17.TL), 
#     file = "TL_OS17_143B_OS2_OS7_2/Fig3_OS17TL.RData")
#save(list = c(t143b.cx.raw, t143b.lung.raw, t143b.tib.raw, t143b), 
#     file = "TL_OS17_143B_OS2_OS7_2/Fig3_t143bTL.RData")
#save(list = c(OS2.tib.raw, OS2.lung.raw, OS2), 
#     file = "TL_OS17_143B_OS2_OS7_2/Fig3_OS2TL.RData")
#save(list = c(OS7.tib.raw, OS7.lung.raw, OS7), 
#     file = "TL_OS17_143B_OS2_OS7_2/Fig3_OS7TL.RData")

#remove these objects and then run CCR on OS and save that file separately; create OS.listonce cleared and re-started
#save(list = c(OS, OS.CCR), 
#     file = "TL_OS17_143B_OS2_OS7_2/Fig3_OS7TL.RData")

#Modify data to account for optimal number of clusters using nRes and pSil functions
OS17.TL.nRes <- nRes(OS17.TL, 
                     res = seq(from = 0.1, to = 0.2, by = 0.05))

plot <- pSil(OS17.TL.nRes , 0.15)

OS17.TL <- FindClusters(OS17.TL, resolution = 0.20)

DimPlot(OS17.TL, reduction = "umap", pt.size = 1, label = T, split.by = "src") +
  coord_fixed() +
  NoLegend() + NoAxes()

#Modify data to account for optimal number of clusters using nRes and pSil functions
t143b.nRes <- nRes(t143b, 
                     res = seq(from = 0.1, to = 0.4, by = 0.05))

plot <- pSil(t143b.nRes , 0.15)

t143b <- FindClusters(t143b, resolution = 0.15)

DimPlot(t143b, reduction = "umap", pt.size = 1, label = T, split.by = "src") +
  coord_fixed()

#Modify data to account for optimal number of clusters using nRes and pSil functions
OS2.nRes <- nRes(OS2, 
                     res = seq(from = 0.1, to = 0.4, by = 0.05))

plot <- pSil(OS2.nRes, 0.15)

OS2 <- FindClusters(OS2, resolution = 0.15)

DimPlot(OS2, reduction = "umap", pt.size = 1, label = T, split.by = "src") +
  coord_fixed() +
  NoLegend() + NoAxes()

#Modify data to account for optimal number of clusters using nRes and pSil functions
OS7.nRes <- nRes(OS7, 
                     res = seq(from = 0.1, to = 0.2, by = 0.05))

plot <- pSil(OS7.nRes , 0.1)

OS7 <- FindClusters(OS7, resolution = 0.1)

DimPlot(OS7, reduction = "umap", pt.size = 1, label = T, split.by = "src") +
  coord_fixed() +
  NoLegend() + NoAxes()


#For lab meeting
#cx.raw.nC <- nRes(cx.raw, 
#                  res = seq(from = 0.1, to = 0.5, by = 0.1))
#plot <- pSil(cx.raw.nC, 0.5)#
#
#cx.raw <- FindClusters(cx.raw, resolution = 0.2)

#plot <- DimPlot(cx.raw, reduction = "umap", pt.size = 1, label = T, split.by = "src") +
#  coord_fixed()
# 
# #Discussion from lab meeting - manually assign cluster IDs for OS17-culture
# cx.raw.new <- cx.raw
# library(grDevices)
# library(grDevices, lib.loc = "C:/Program Files/R/R-3.6.2/library")
# windows()
# cx.raw.new <- CellSelector(plot = plot, object = cx.raw.new, ident = "4")
# 
# Idents(cx.raw) <- factor(x = Idents(cx.raw), levels = sort(levels(cx.raw)))
# Clusters <- Idents(object = cx.raw)
# names(Clusters) <- colnames(x = cx.raw)
# cx.raw.new <- AddMetaData(
#   object = cx.raw.new,
#   metadata = Clusters,
#   col.name = 'cluster.idents'
# )
# head(x = cx.raw.new[[]])
# 
# clusters <- cx.raw.new@meta.data[["cluster.idents"]]
# dist.matrix <- dist(x = Embeddings(object = cx.raw.new[["pca"]])[, 1:20])
# sil <- silhouette(x = as.numeric(x = as.factor(x = clusters)), dist = dist.matrix)
# # cx.raw.new$sil <- sil[, 3]
# 
# windows()
# # col= pal_npg("nrc")(n)
# plot(sil,
#      col = pal_npg("nrc")(4)) # with cluster-wise coloring
# 
# pdf("temp.pdf")
# DimPlot(cx.raw.new, reduction = "umap", pt.size = 1, label = T) +
#   coord_fixed()
# dev.off()

#subset after finding clusters
# re-arrange and re-name levels 

# temp <- OS17.TL
# temp$src <- factor(as.factor(temp$src), levels = c("OS17_Tibia", "OS17_Lung"))
library(plyr)
# temp$src <- revalue(temp$src, c("OS17_Tibia"="Tibia", "OS17_Lung"="Lung"))
OS17.TL$src <- as.factor(OS17.TL$src)
OS17.TL$src <- factor(as.factor(OS17.TL$src), levels = c("OS17_Tibia", "OS17_Lung"))
OS17.TL$src <- revalue(OS17.TL$src, c("OS17_Tibia"="Tibia", "OS17_Lung"="Lung"))

OS2$src <- as.factor(OS2$src)
OS2$src <- factor(as.factor(OS2$src), levels = c("NCHOS2_Tibia", "NCHOS2_Lung"))
OS2$src <- revalue(OS2$src, c("NCHOS2_Tibia"="Tibia", "NCHOS2_Lung"="Lung"))

OS7$src <- as.factor(OS7$src)
OS7$src <- factor(as.factor(OS7$src), levels = c("NCHOS7_Tibia", "NCHOS7_Lung"))
OS7$src <- revalue(OS7$src, c("NCHOS7_Tibia"="Tibia", "NCHOS7_Lung"="Lung"))

OS17.TL <- OS17.TL %>% FindClusters(resolution = 0.2)
t143b <- t143b %>% FindClusters(resolution = 0.15)
OS2 <- OS2 %>% FindClusters(resolution = 0.15)
OS7 <- OS7 %>% FindClusters(resolution = 0.2)

pdf("TL_OS17_143B_OS2_OS7_2/OS17.TL_fig3.pdf", width = 5, height = 5)
OS17.TL %>% FindClusters(resolution = 0.2) %>% 
  subset(cells = sample(Cells(OS17.TL), 3000)) %>% 
  DimPlot(reduction = "umap", pt.size = 1, label = T, split.by = "src") + 
  coord_fixed() + NoLegend() + NoAxes()+
  scale_color_npg(alpha = 1)
dev.off()

pdf("TL_OS17_143B_OS2_OS7_2/t143b_fig3.pdf", width = 5, height = 5)
t143b %>% FindClusters(resolution = 0.15) %>% 
  subset(cells = sample(Cells(t143b), 3000)) %>% 
  DimPlot(reduction = "umap", pt.size = 1, label = T, split.by = "src") + 
  coord_fixed() + NoLegend() + NoAxes()+
  scale_color_npg(alpha = 1)
dev.off()

pdf("TL_OS17_143B_OS2_OS7_2/OS2_fig3.pdf", width = 5, height = 5)
OS2 %>% FindClusters(resolution = 0.15) %>% 
  subset(cells = sample(Cells(OS2), 3000)) %>% 
  DimPlot(reduction = "umap", pt.size = 1, label = T, split.by = "src") + 
  coord_fixed() + NoLegend() + NoAxes()+
  scale_color_npg(alpha = 1)
dev.off()

pdf("TL_OS17_143B_OS2_OS7_2/OS7_fig3.pdf", width = 10, height = 10)
OS7 %>% FindClusters(resolution = 0.2) %>% 
  subset(cells = sample(Cells(OS7), 3000)) %>% 
  DimPlot(reduction = "umap", pt.size = 1, label = T, split.by = "src") +
  ggtitle("OS7")+
  coord_fixed() + NoLegend() + NoAxes()+
  scale_color_npg(alpha = 1)
dev.off()

#create heatmap 
B.list <- list(OS17 = OS17.TL,
               t143b = t143b,
               OS2 = OS2,
               OS7 = OS7)

#Heatmap and em.hm files
em.hm.list <- list()
for (i in 1:length(B.list)) {
  em.hm.list[[i]] <- DGEA(B.list[[i]]) 
}

library(data.table)
for(i in 1:(length(em.hm.list))) {
  em.hm.list[[i]] <- setDT(em.hm.list[[i]], keep.rownames = TRUE)[]
}

temp1 <- em.hm.list[[1]]
temp2 <- em.hm.list[[2]]
temp3 <- em.hm.list[[3]]
temp4 <- em.hm.list[[4]]

cx <- inner_join(temp1, temp2, by = "rn")
pd <- inner_join(temp3, temp4, by = "rn")
cx.pd <- inner_join(cx, pd, by = "rn")

head(cx.pd)

##Write table to edit rownames to easily convert to dataframe
write.table(cx.pd, file=('TL_OS17_143B_OS2_OS7_2/cxpd.tsv'), quote=FALSE, sep='\t')
cx.pd <- read.table(file = "TL_OS17_143B_OS2_OS7_2/cxpd.tsv", header=T)

# Make the "rn" column the rownames -Emily
cx.pd <- remove_rownames(cx.pd) %>%
        column_to_rownames(var = "rn")

#remove "HALLMARK_"
c <- rownames(cx.pd)
c <- gsub("HALLMARK_", "", c)
#remove "_" by removing special characters
c <- gsub("_", " ", c)
rownames(cx.pd) <- c

cx.pd.log <- -log10(cx.pd) #log transform
library(pheatmap)
cx.pd_transpose <- as.data.frame(t(cx.pd.log))
df <- as.matrix(cx.pd_transpose)
#Write table to print supplemental information
write.table(df, file=('TL_OS17_143B_OS2_OS7_2/cxpd_plot.tsv'), quote=FALSE, sep='\t')

subj<-c(paste0("A", 0:3),
        paste0("B", 0:3),
        paste0("C", 0:2),
        paste0("D", 0:3))
rownames(df)<-subj
aka2 = data.frame(ID = factor(c(rep("OS17", 4),
                                rep("t143B", 4),
                                rep("OS2", 3),
                                rep("OS7", 4)
                                )))
rownames(aka2)<-subj
aka3 = list(ID = c(OS17 = "#E64B35FF", t143B = "#4DBBD5FF", OS2 = "#00A087FF", OS7 = "#3C5488FF"))
# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# 
# BiocManager::install("ComplexHeatmap")
library(ComplexHeatmap)
color = colorRampPalette(rev(brewer.pal(n = 7, name =
                                          "RdYlBu")))(100)
#calculate percentage of cells in lung for each cluster
aka4 <- data.frame (Lpercent = c(46.577747,
                                 10.17901,
                                 39.69814,
                                 3.545104,
                                 2.2204908,
                                 92.306194,
                                 3.3502143,
                                 2.1231009,
                                 20.32767,
                                 71.480583,
                                 8.191748,
                                 71.01349,
                                 1.383604,
                                 1.176064,
                                 26.426842))
rownames(aka4)<-subj
aka5 = (c(rep("OS17", 4), rep("t143B", 4), rep("OS2", 3), rep("OS7", 4)))

column_ha = HeatmapAnnotation(model = aka5, bar1 = anno_barplot(aka4))
#pdf("TL_OS17_143B_OS2_OS7_2/temp.pdf", height = 7, width = 8)
#Heatmap(t(df), 
#        col = col, 
#        name = "-log10(p)", 
##        cluster_columns = FALSE,
#       top_annotation = column_ha)+ geom_text(aes(label = pvalue))
#dev.off()

col.breaks=seq(-log10(1),min(max(-log10(cx.pd))+1,18),by=0.5)
col=inferno(length(col.breaks)) # library(viridis)
col=c("white",colorRampPalette(brewer.pal(n = 7, name ="Reds"))(50))
pheatmap(-log10(cx.pd),cluster_rows = TRUE,cluster_cols = TRUE,
         cellwidth = 5,cellheight = 7,treeheight_row = 0,treeheight_col=0,
         color = col,scale='none',breaks=col.breaks,fontsize = 8)
```

```{r TL_OS17_143B_OS2_OS7_2_cont, cache.lazy=F}
library(dichromat)
#IPA analysis
# Decided to proceed with A2, B1, (B2, B3), C1, D0 and D3
# Find differentially expressed features selected cluster and remaining cells
#OS17
A0.markers <- FindMarkers(OS17.TL, ident.1 = "0", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
A1.markers <- FindMarkers(OS17.TL, ident.1 = "1", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
A2.markers <- FindMarkers(OS17.TL, ident.1 = "2", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
A3.markers <- FindMarkers(OS17.TL, ident.1 = "3", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)

#remove unnecessary columns
myList <- list(A0.markers, A1.markers, A2.markers, A3.markers)
myList <- lapply(myList, function(x) { x["p_val"] <- NULL; x[, 2:3] <- NULL; x })
A0.markers <- myList[[1]]
A1.markers <- myList[[2]]
A2.markers <- myList[[3]]
A3.markers <- myList[[4]]
  
# view results
# head(A2.markers)
# FeaturePlot(object = OS17.TL, features = 'NFKBIA')
B0.markers <- FindMarkers(t143b, ident.1 = "0", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
B1.markers <- FindMarkers(t143b, ident.1 = "1", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
B2.markers <- FindMarkers(t143b, ident.1 = "2", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
B3.markers <- FindMarkers(t143b, ident.1 = "3", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)

myList <- list(B0.markers, B1.markers, B2.markers, B3.markers)
myList <- lapply(myList, function(x) { x["p_val"] <- NULL; x[, 2:3] <- NULL; x })
B0.markers <- myList[[1]]
B1.markers <- myList[[2]]
B2.markers <- myList[[3]]
B3.markers <- myList[[4]]

C0.markers <- FindMarkers(OS2, ident.1 = "0", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
C1.markers <- FindMarkers(OS2, ident.1 = "1", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
C2.markers <- FindMarkers(OS2, ident.1 = "2", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)

myList <- list(C0.markers, C1.markers, C2.markers)
myList <- lapply(myList, function(x) { x["p_val"] <- NULL; x[, 2:3] <- NULL; x })
C0.markers <- myList[[1]]
C1.markers <- myList[[2]]
C2.markers <- myList[[3]]

D0.markers <- FindMarkers(OS7, ident.1 = "0", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
D1.markers <- FindMarkers(OS7, ident.1 = "1", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
D2.markers <- FindMarkers(OS7, ident.1 = "2", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
D3.markers <- FindMarkers(OS7, ident.1 = "3", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)

myList <- list(D0.markers, D1.markers, D2.markers, D3.markers)
myList <- lapply(myList, function(x) { x["p_val"] <- NULL; x[, 2:3] <- NULL; x })
D0.markers <- myList[[1]]
D1.markers <- myList[[2]]
D2.markers <- myList[[3]]
D3.markers <- myList[[4]]

# commented out due to inability to install xlsx -Emily
#library(xlsx)
#write.xlsx(A0.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/A0.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(A1.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/A1.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(A2.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/A2.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(A3.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/A3.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)


#write.xlsx(B0.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/B0.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(B1.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/B1.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(B2.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/B2.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(B3.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/B3.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)

#write.xlsx(C0.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/C0.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(C1.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/C1.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(C2.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/C2.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)

#write.xlsx(D0.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/D0.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(D1.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/D1.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(D2.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/D2.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(D3.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/D3.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)

#Uploaded individual datasets on IPA; ran Expression analysis with adj p value cutoff at 0.01
#Emily commented out below
#df <- read.delim("TL_OS17_143B_OS2_OS7_2/IPA_updown.txt", header = TRUE)
#rownames(df) <- df[,1]
#df[,1] <- NULL
```

```{r TL_OS17_143B_OS2_OS7_2_cont2}
# cx.pd <- cx.pd[-(23:30),] #remove rows with names NA - figure out why we have NAs!!
cx.pd.log <- -log10(cx.pd) #log transform
library(pheatmap)   

cx.pd_transpose <- as.data.frame(t(cx.pd.log))
df <- as.matrix(cx.pd_transpose)

subj<-c(paste0("A", 0:3),
        paste0("B", 0:3),
        paste0("C", 0:2),
        paste0("D", 0:3))
rownames(df)<-subj
aka2 = data.frame(ID = factor(c(rep("OS17", 4),
                                rep("t143B", 4),
                                rep("OS2", 3),
                                rep("OS7", 4)
)))
rownames(aka2)<-subj
aka3 = list(ID = c(OS17 = "#E64B35FF", t143B = "#4DBBD5FF", OS2 = "#00A087FF", OS7 = "#3C5488FF"))
# 
# df[] <- lapply(df, gsub, pattern='HALLMARK_', replacement='')
# df

cols <- colorRampPalette(c("snow1", "red"), # distances 3 to max(distmat) colored from green to black
                             100)

# not the heatmap code used in final version -Emily commented out
#pheatmap(t(scale(df)),
 #        color = cols, 
 #        annotation_col = aka2, 
 #        annotation_colors = aka3,
 #        annotation_legend = TRUE,
         #gaps_col =  4,
 #        show_colnames = T, show_rownames = T, cluster_rows = T, 
 #        cluster_cols = T, legend = TRUE, 
#         clustering_distance_rows = "euclidean", border_color = FALSE)

# unsure what package this function is from -Emily
#Heatmap(t(scale(df)), 
#        col = col, 
#        name = "-log10(p)", 
#        top_annotation = column_ha)

#not used elsewhere here -Emily
#color = colorRampPalette(rev(brewer.pal(n = 7, name =
#                                          "RdYlBu")))(100)

##Apply glycolysis, Hypoxia, tnfa, emt modules and compare between clusters
data <- OS17.TL

for(i in 1:length(mod_names)) {
  data <- AddModuleScore(data, gl[i], name = names(gl[i]))
}
RidgePlot(data, features = str_c(names(gl[6]), "1"))
# VlnPlot(data, features = str_c(names(gl[3]), "1"))
DotPlot(data, features = TNFA, cluster.idents = TRUE)

#noted a clear differnece in the TNFA_VIA_NFkB module
#Identify genes that are most differntially expressed in this module

data <- OS17.TL
pdf("TL_OS17_143B_OS2_OS7_2/Dotplot_OS17.pdf", height = 30, width = 7)
DotPlot(data, features = TNFA, cluster.idents = TRUE)+coord_flip()
dev.off()

data <- t143b
pdf("TL_OS17_143B_OS2_OS7_2/Dotplot_143B.pdf", height = 30, width = 7)
DotPlot(data, features = TNFA, cluster.idents = TRUE)+coord_flip()
dev.off()

data <- OS2
pdf("TL_OS17_143B_OS2_OS7_2/Dotplot_OS2.pdf", height = 30, width = 7)
DotPlot(data, features = TNFA, cluster.idents = TRUE)+coord_flip()
dev.off()

data <- OS7
pdf("TL_OS17_143B_OS2_OS7_2/Dotplot_OS7.pdf", height = 30, width = 7)
DotPlot(data, features = TNFA, cluster.idents = TRUE)+coord_flip()
dev.off()

################################Pathway enrichment analysis 

P1 <- df %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  full_join(aka4 %>% rownames_to_column(var = "Sample")) %>%
  pivot_longer(c(-Sample, -Lpercent),
               names_to = "Pathway",
               values_to = "pval",
               names_repair = "minimal") %>%
  mutate(Signif = pval >= (-1 * log10(0.05)),
         sample_order = str_remove(Sample, "[0-9]") %>%
           rank() * 1000 - Lpercent,
         Sample = reorder(Sample, sample_order)) %>%
  ggplot(., aes(x = Sample, y = Pathway, fill = Signif)) + 
  geom_tile() +
  geom_text(aes(label = sprintf("%0.2f", pval))) +
  scale_fill_manual(values = c("gray", "#CC3333")) +
  scale_y_discrete(limits = rev) +
  theme(legend.position = "none") +
  ylab("") 
P2 <- aka4 %>%
  rownames_to_column(var = "Sample") %>%
  full_join(aka2 %>%
              rownames_to_column(var = "Sample")) %>%
  mutate(sample_order = str_remove(Sample, "[0-9]") %>%
           rank() * 1000 - Lpercent,
         Sample = reorder(Sample, sample_order),
         ID = reorder(ID, sample_order)) %>%
  ggplot(., aes(x = Sample, y = Lpercent, fill = ID)) +
  geom_bar(stat = "identity") +
  ylab("Percent\nin lung") +
  xlab("") + theme_bw()+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  theme(legend.position = "top", ) +
  labs(fill = "") 

pdf("TL_OS17_143B_OS2_OS7_2/Pvalueplot.pdf", height = 12, width = 12)
cowplot::plot_grid(P2, P1, ncol = 1, align = "v",rel_heights = c(2, 10))
dev.off()

```

```{r heatmap_reviewer_add, cache.cache.lazy=F, eval = F}
objlist <- c(cx.raw, S0043.CCR, OS17.TL, t143b, OS2, OS7)
objname <- c("OS17 Culture", "NCH-OS-7 Flank", "OS17 Tibia and Flank", "143B", "NCH-OS-2", "NCH-OS-7")

for(i in 1:length(objlist)) {
  heat <- DoHeatmap(objlist[[i]], group.bar = T) 
  p <- heat + scale_color_npg(alpha = 1)
  pdf(paste0("heatmap_add_",names(objname[i]),".pdf"), height = 12, width = 12)
  print(p)
}
dev.off()

```









# BatchEffectCorrection_LongitudinalData

```{r BatchEffectCorrection_LongitudinalData, cache.lazy=F, eval = F}
#Harmony on different samples
source("R:/RESRoberts/Bioinformatics/Analysis/Sanjana/scSeurat.R")
set.seed(108)
library("harmony")

load("R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/17.Harmony/Harmony in individual samples/Figure2/OS.combined.postCCR.RData")

DimPlot(OS, reduction = "umap", group.by = "Phase", pt.size = 1, label = F) + 
  coord_fixed() + 
  ggtitle("OS by Source") + 
  scale_color_npg(alpha = 1)

DimPlot(OS, reduction = "umap", pt.size = 1, label = F, group.by = "cond") +
  coord_fixed() + 
  ggtitle("OS Clusters") +
scale_color_npg(alpha = 0.9) 

#Run Harmony
OS <- OS %>%
  RunPCA(pc.genes = OS@var.genes, npcs = 20) %>%
  RunHarmony(group.by.vars = "src", plot_convergence = T) %>%
  RunUMAP(reduction = "harmony", dims = 1:20) %>%
  FindNeighbors(reduction = "harmony", dims = 1:20) %>%
  FindClusters(resolution = 0.2) ##Note change in resolution

#re-order levels
#Feature Plot for INPP4B, FOSL1, EGR1, FOS
OS$cond <- as.factor(OS$cond)
OS$cond <- factor(as.factor(OS$cond), levels = c("OB", "OS17", "t143b", "NCHOS2", "NCHOS7"))
OS$cond <- revalue(OS$cond, c("OB"="Osteoblast", 
                              "OS17"="OS17", 
                              "t143b" = "143B", 
                              "NCHOS2" = "NCH-OS-2", 
                              "NCHOS7" = "NCH-OS-7"))

#Use FastMNN for batch effect correction; cells still cluster based on cell cycle post-correction
#Note: Running FastMNN after running Harmony gives the same result as without

OS <- RunFastMNN(object.list = SplitObject(OS, split.by = "cond"))
OS <- RunUMAP(OS, reduction = "mnn", dims = 1:20)
OS <- FindNeighbors(OS, reduction = "mnn", dims = 1:20)
OS <- FindClusters(OS, resolution = 0.2)
DimPlot(OS, group.by = c("cond", "src"), ncol = 3)

DimPlot(OS, reduction = "umap", group.by = "Phase", pt.size = 1, label = T) + 
  coord_fixed() + 
  ggtitle("OS by Source") + 
  scale_color_npg(alpha = 1)
```

# BatchEffectCorrection_OS17replicates

```{r BatchEffectCorrection_OS17replicates, cache.lazy=F, eval=F}
#Determine overlay of replicates with or without the use batch-effect correction methods such as Harmony, FastMNN

#source("R:/RESRoberts/Bioinformatics/Analysis/Sanjana/scSeurat.R")
set.seed(108)
library(harmony)
library(Seurat)
library(SeuratData)
library(SeuratWrappers)

#Cellecta - Replicate 1
#Load cell culture data
cellecta.cx.raw <- tenXLoadQC("R:/RESRoberts/Bioinformatics/scRNAOuts/S0015-cellecta/filtered_feature_bc_matrix/", spec = "mixHuman")
cellecta.cx.raw <- subset(cellecta.cx.raw, subset = nFeature_RNA >2500 & nCount_RNA <40000 & percent.mt <14)
cellecta.cx.raw$src <- "Cellecta_Cx"
cellecta.cx.raw$cond <- "Plate"
cellecta.cx.raw$rep <- "Rep1"

#Load orthotopic lung data
cellecta.lung.raw <- tenXLoadQC("R:/RESRoberts/Bioinformatics/scRNAOuts/S0023-cellecta-lung/filtered_feature_bc_matrix/", spec = "mixHuman")
cellecta.lung.raw <- subset(cellecta.lung.raw, subset = percent.mt <20 & percent.mt > 2)
cellecta.lung.raw$src <- "Cellecta_Lung"
cellecta.lung.raw$cond <- "Lung"
cellecta.lung.raw$rep <- "Rep1"

#Zymo - Replicate 2
zymo.cx.raw <- tenXLoadQC(path10x = "R:/RESRoberts/Bioinformatics/scRNAOuts/S0016-zymo/filtered_feature_bc_matrix/", spec = "mixHuman")
zymo.cx.raw <- subset(zymo.cx.raw, subset = nFeature_RNA >3500 & nCount_RNA <50000 & percent.mt <15)
zymo.cx.raw$src <- "Zymo_Cx"
zymo.cx.raw$cond <- "Plate"
zymo.cx.raw$rep <- "Rep2"

#Load orthotopic bone data
zymo.tib.raw <- tenXLoadQC("R:/RESRoberts/Bioinformatics/scRNAOuts/S0018xS0028/filtered_feature_bc_matrix/", spec = "mixHuman")
zymo.tib.raw <- subset(zymo.tib.raw, subset = nFeature_RNA >3000 & nCount_RNA <60000 & percent.mt <18)
zymo.tib.raw$src <- "Zymo_Tib"
zymo.tib.raw$cond <- "Bone"
zymo.tib.raw$rep <- "Rep2"

#Load orthotopic lung data
zymo.lung.raw<- tenXLoadQC(path10x = "R:/RESRoberts/Bioinformatics/scRNAOuts/S0024xS0029/filtered_feature_bc_matrix/", spec = "mixHuman")
zymo.lung.raw <- subset(zymo.lung.raw, subset = nFeature_RNA >1250 & nCount_RNA <60000 & percent.mt <25)
zymo.lung.raw$src <- "Zymo_Lung"
zymo.lung.raw$cond <- "Lung"
zymo.lung.raw$rep <- "Rep2"

#Subset to least number of cells across all conditions (Cell culture and orthotopic lung)
#Subset all to #1900 cells

cellecta.cx.raw <- subset(cellecta.cx.raw, cells = sample(Cells(cellecta.cx.raw), 1900))
cellecta.lung.raw <- subset(cellecta.lung.raw, cells = sample(Cells(cellecta.lung.raw), 1900))
zymo.cx.raw <- subset(zymo.cx.raw, cells = sample(Cells(zymo.cx.raw), 1900))
zymo.lung.raw <- subset(zymo.lung.raw, cells = sample(Cells(zymo.lung.raw), 1900))

# Merge raw count matrices of these four datsets into a single Seurat object
Rep1_H <- merge(cellecta.cx.raw, y = c(cellecta.lung.raw, zymo.cx.raw, zymo.lung.raw),
              add.cell.ids = c("Cellecta_Cx", "Cellecta_Lung", "Zymo_Cx", "Zymo_Lung"),
              project = "Heterogeneity")

# Dimensionality reduction and clustering
Rep1_H <- NormalizeData(Rep1_H) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = Rep1_H@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.2) ##Note change in resolution

#Determine contribution of cell cycle related genes to variaiton in the datasets
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
Rep1_H <- CellCycleScoring(object = Rep1_H, s.features = s.genes,
                                g2m.features = g2m.genes, set.ident = TRUE)
DimPlot(Rep1_H, reduction = "umap", pt.size = 1, label = F, group.by = "Phase") + 
  coord_fixed() + 
  ggtitle("Reps by Source") + 
  scale_color_npg(alpha = 0.7)

# Mitigate effects of cell cycle heterogeneity by regressing out canonical G2/M- and S-Phase genes 
Rep1_H <- ScaleData(object = Rep1_H, vars.to.regress = c("S.Score", "G2M.Score"),
                  features = rownames(x = Rep1_H))

# Re-do dimensionality reduction and clustering
Rep1_H_v1 <- Rep1_H %>%
    RunPCA(pc.genes = Rep1_H_v1@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.2) ##Note change in resolution

#Visualize the data
DimPlot(Rep1_H_v1, reduction = "umap", pt.size = 1, label = F, group.by = "Phase") + 
  coord_fixed() + 
  ggtitle("Reps by Source") + 
  scale_color_npg(alpha = 0.7)

#Run Harmony
Rep1_H_v2 <- Rep1_H %>%
  RunPCA(pc.genes = Rep1_H_v2@var.genes, npcs = 20) %>%
  RunHarmony(group.by.vars = "src", plot_convergence = T) %>%
  RunUMAP(reduction = "harmony", dims = 1:20) %>%
  FindNeighbors(reduction = "harmony", dims = 1:20) %>%
  FindClusters(resolution = 0.2) ##Note change in resolution

#Visualize the data
DimPlot(Rep1_H_v2, reduction = "umap", pt.size = 1, label = F, group.by = "src") + 
  coord_fixed() + 
  scale_color_npg(alpha = 0.7)

#Run FastMNN
Rep1_H_v3 <- NormalizeData(Rep1_H) %>%
  FindVariableFeatures(selection.method = "vst")
  
Rep1_H_v3 <- RunFastMNN(object.list = SplitObject(Rep1_H_v3, split.by = "cond"))
Rep1_H_v3 <- RunUMAP(Rep1_H_v3, reduction = "mnn", dims = 1:20)
Rep1_H_v3 <- FindNeighbors(Rep1_H_v3, reduction = "mnn", dims = 1:20)
Rep1_H_v3 <- FindClusters(Rep1_H_v3, resolution = 0.2)

DimPlot(Rep1_H_v3, reduction = "umap", pt.size = 1, label = F, group.by = "rep") + 
  coord_fixed() + 
  scale_color_npg(alpha = 0.7)

# DimPlot(Rep1_H_v3, reduction = "umap", pt.size = 1, label = F, group.by = c("src", "Phase"), ncol = 2) + 
#   coord_fixed() + 
#   scale_color_npg(alpha = 0.7)
```

# BatchEffectCorrection_SCTransform

```{r BatchEffectCorrection_SCTransform, cache.lazy=F, eval = F}
#BatchEffectCorrectionSCTransform - all
source("/gpfs0/home/gdrobertslab/rssxr002/Analysis/scSeurat.R")
library(dplyr)
library(ggplot2)
library(harmony)
library(cowplot)
library(Seurat)
set.seed(100)

load("OS.combined.postCCR.RData")
#Integrate using Seurats SCTransform Integration workflow
OSrep.list <- SplitObject(OS, split.by = "src")

for (i in 1:length(OSrep.list)) {
  OSrep.list[[i]] <- SCTransform(OSrep.list[[i]], verbose = FALSE)
}

OSrep.features <- SelectIntegrationFeatures(object.list = OSrep.list, nfeatures = 3000)
OSrep.list <- PrepSCTIntegration(object.list = OSrep.list, anchor.features = OSrep.features, 
                                 verbose = FALSE)
OSrep.anchors <- FindIntegrationAnchors(object.list = OSrep.list, normalization.method = "SCT", 
                                        anchor.features = OSrep.features, verbose = FALSE)
OSrep.integrated <- IntegrateData(anchorset = OSrep.anchors, normalization.method = "SCT", 
                                  verbose = FALSE)

OSrep.integrated <- RunPCA(OSrep.integrated, verbose = FALSE)
OSrep.integrated <- RunUMAP(OSrep.integrated, dims = 1:30)

saveRDS(OSrep.integrated, file = "OSrep.integrated.rds")

pdf("OSrep.integrated.Phase.pdf")
DimPlot(OS, reduction = "umap", group.by = "Phase", pt.size = 1, label = F) + 
  coord_fixed() + 
  ggtitle("OS by Source") + 
  scale_color_npg(alpha = 1)
dev.off()

pdf("OSrep.integrated.pdf")
DimPlot(OS, reduction = "umap", pt.size = 1, label = T) +
  coord_fixed() + 
  ggtitle("OS Clusters") 
scale_color_npg(alpha = 1) 
dev.off()
```