---
title: "OSHetero2021"
author: "Sanjana Rajan, Emily Franz, Matthew Cannon"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 5
    number_sections: false
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  echo = TRUE,
  cache = TRUE,
  collapse = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  cache.lazy = FALSE
)
```

# Figure 1

**Cell line and PDX models of osteosarcoma display transcriptional heterogeneity.**

```{r Fig1_lib, cache=FALSE}
source("scSeurat.R")
source("Downstream.v2.R")
set.seed(888)
# loading libraries
library(Seurat)
library(future)
library(ggplot2)
library(msigdbr)
library(clusterProfiler)
library(SingleR)
library(dplyr)
library(pheatmap)
library(RColorBrewer)
library(viridis) # inferno color palette
library(grid)
library(ggsci)
library(data.table)
library(ggvenn)
library(tibble)
library(tidyr)
library(tidyverse)
library(stringr)
library(reshape2)
library(pzfx)
library(ggridges)
library(effsize)
library(perm)
```

## Figure 1A

*DONE*

A) UMAP analysis of OS-17 cells (n = 3178) grown in cell culture. Cell cycle distribution of cells (n = 3178) is visualized as pie charts.

```{r Fig1a_OS17}
load("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/Data/os17.cx_CCR.RData")

# Culture plot
figure1_a1 <- DimPlot(os17.cx,
                      reduction = "umap",
                      pt.size = 1,
                      label = T) +
  coord_fixed() +
  ggtitle("OS17 in Culture") +
  NoAxes() +
  scale_color_npg(alpha = 1)

# View Culture plot in Rmd
figure1_a1

# Cell cycle distribution of clusters
cid <- sort(unique(os17.cx@active.ident))

vplayout <- function(x, y) viewport(layout.pos.row = x,
                                    layout.pos.col = y)


png("figures/figure_1/figure_1a_os17_pie.png",
    units = "in",
    width = 15,
    height = 15,
    res = 300)

grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 4)))

for(i in 1:3){
  temp <- WhichCells(os17.cx,
                     idents = cid[[i]])
  LT.cells <- subset(os17.cx,
                     cells = temp)
  df <- table(LT.cells$Phase)
  bp <- ggplot(as.data.frame(df),
               aes(x = "",
                   y = Freq,
                   fill = Var1)) +
    geom_bar(width = 1,
             stat = "identity")
  pie <- bp +
    coord_polar("y",
                start = 0) +
    scale_fill_brewer(palette ="Blues") +
    theme_minimal() +
    NoAxes() +
    ggtitle(paste("Cluster",
                  cid[[i]]))
  print(pie,
        vp = vplayout(ceiling(i/4),
                      i))
}
dev.off()
#Total number of cells
# table(Idents(os17.cx)) %>%
# sum()
# [1] 3178
```

## Figure 1B

*DONE*

B) UMAP analysis of NCH-OS-7 cells (n = 1998) grown as subcutaneous flank tumor. Cell cycle distribution of cells is visualized as pie charts.

```{r Fig1b_NCHOS7}
load(file ="/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/Data/OS7.cx_CCR.RData")

# Flank plots

figure1_a1 <- DimPlot(OS7.cx,
                      reduction = "umap",
                      pt.size = 1,
                      label = T) + 
  coord_fixed() + 
  ggtitle("NCHOS7 in Flank") +
  NoAxes() +
  scale_color_npg(alpha = 1)

# View in Rmd
figure1_a1

# Cell cycle distribution of clusters
cid <- sort(unique(OS7.cx@active.ident))

vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)

png("figures/figure_1/figure_1a_nchos7_pie.png",
    units = "in",
    width = 15,
    height = 15,
    res = 300)

grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 4)))

for(i in 1:4){
  temp <- WhichCells(OS7.cx,
                     idents = cid[[i]]) 
  LT.cells <- subset(OS7.cx,
                     cells = temp)
  df <- table(LT.cells$Phase)
  bp <- ggplot(as.data.frame(df),
               aes(x="",
                   y=Freq,
                   fill=Var1)) +
    geom_bar(width = 1,
             stat = "identity")
  pie <- bp +
    coord_polar("y",
                start=0) +
    scale_fill_brewer(palette="Blues") +
    theme_minimal() +
    NoAxes() +
    NoLegend() +
    RestoreLegend(position = "right")
  ggtitle(paste("Cluster",
                cid[[i]]))
  print(pie,
        vp=vplayout(ceiling(i/4),
                    i))
}
dev.off()

#Total number of cells
# table(Idents(OS7.cx)) %>%
# sum()
# [1] 1998
```

## Figure 1C

C) Pathway enrichment analysis for hallmark gene sets associated with genes upregulated in distinct clusters identified in the two osteosarcoma models. P values were adjusted for multiple comparisons. Boxes in grey identify non-significant enrichments, whereas boxes in red identify statistically significant enrichments. MTORC1: mechanistic target of rapamycin (mTOR) complex 1. PI3K: Phosphoinositide 3-kinase. AKT: Protein kinase B. TGF: Transforming growth-factor. TNF: tumor necrosis factor. NFKB: Nuclear Factor kappa-light-chain-enhancer of activated B cells.

*Need correct input for this figure heatmap.*

```{r Fig1c_dgea, dependson='Fig1b_NCHOS7'}
# Pathway enrichment analysis
# Display adjust p-values
B.list <- list(OS17 = os17.cx,
               NCHOS7 = OS7.cx)

em.hm.list <- list()

for (i in 1:length(B.list)) {
  em.hm.list[[i]] <- DGEA(B.list[[i]])
}

for(i in 1:(length(em.hm.list))) {
  em.hm.list[[i]] <- setDT(em.hm.list[[i]],
                           keep.rownames = TRUE)[]
}

temp1 <- em.hm.list[[1]]
temp2 <- em.hm.list[[2]]
cx.pd <- full_join(temp1,
                   temp2,
                   by = "rn")

cx.pd <- column_to_rownames(cx.pd,
                            var = "rn")
cx.pd[is.na(cx.pd)] <- 1

#remove "HALLMARK_"
c <- rownames(cx.pd)
c <- gsub("HALLMARK_",
          "",
          c)

#remove "_" by removing special characters
c <- gsub("_",
          " ",
          c)
rownames(cx.pd) <- c

# Take the -log10
cx.pd.log <- -log10(cx.pd)
colnames(cx.pd.log) <- c("A0",
                         "A1",
                         "A2",
                         "B0",
                         "B1",
                         "B2",
                         "B3")

# To prevent values that show up as -0.00, we take the absolute values
cx.pd.log <- abs(cx.pd.log)

cx.pd.log %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  pivot_longer(c(-Sample),
               names_to = "Pathway",
               values_to = "pval",
               names_repair = "minimal") %>%
  mutate(Signif = pval >= (-1 * log10(0.05))) %>%
  ggplot(.,
         aes(x = Sample,
             y = Pathway,
             fill = Signif)) +
  geom_tile() +
  geom_text(aes(label = sprintf("%0.2f",
                                pval))) + 
  scale_fill_manual(values = c("gray",
                               "red")) +
  scale_y_discrete(limits = rev) +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks.x = element_blank()) +
  ylab("")
```

```{r Fig1_heatmaptake2}
#ComplexHeatmap
# Fig1.txv from above code (previously called cxpd.tsv in Sanjana's files)
cx.pd <- read.delim("Fig1.tsv", row.names = "rn")
#remove "HALLMARK_"
c <- rownames(cx.pd)
c <- gsub("HALLMARK_", "", c)
#remove "_" by removing special characters
c <- gsub("_", " ", c)
rownames(cx.pd) <- c
cx.pd.log <- -log10(cx.pd[,-1]) #log transform

cx.pd_transpose <- as.data.frame(t(cx.pd.log))
df <- as.matrix(cx.pd_transpose)
subj <- c(paste0("A", 0:3),
          paste0("B", 0:3),
          paste0("C", 0:2),
          paste0("D", 0:3))
#rownames(df) <- subj
aka2 = data.frame(ID = factor(c(rep("OS17", 4),
                                rep("t143B", 4),
                                rep("OS2", 3),
                                rep("OS7", 4)
)))
rownames(aka2)<-subj
aka3 = list(ID = c(OS17 = "#E64B35FF",
                   t143B = "#4DBBD5FF",
                   OS2 = "#00A087FF",
                   OS7 = "#3C5488FF"))
#Include percentage of cells in lung for each cluster
aka4 <- data.frame (Lpercent = c(46.577747,
                                 10.17901,
                                 39.69814,
                                 3.545104,
                                 2.2204908,
                                 92.306194,
                                 3.3502143,
                                 2.1231009,
                                 20.32767,
                                 71.480583,
                                 8.191748,
                                 71.01349,
                                 1.383604,
                                 1.176064,
                                 26.426842))
rownames(aka4) <- subj
aka5 = (c(rep("OS17", 4),
          rep("t143B", 4),
          rep("OS2", 3),
          rep("OS7", 4)))
# Plot heatmap
Fig1 <- read_delim("Fig1.txt", 
                   delim = "\t", 
                   col_names = TRUE)
colnames(Fig1) <- c("Pathway",
                    "A0",
                    "A1",
                    "A2",
                    "A3",
                    "B0",
                    "B1",
                    "B2",
                    "B3")
Fig1 %>%
  column_to_rownames(var = "Pathway") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  left_join(aka4 %>% rownames_to_column(var = "Sample")) %>%
  pivot_longer(c(-Sample, -Lpercent),
               names_to = "Pathway",
               values_to = "pval",
               names_repair = "minimal") %>%
  mutate(Signif = pval >= (-1 * log10(0.05))) %>%
  ggplot(., aes(x = Sample, y = Pathway, fill = Signif)) + 
  geom_tile() +
  geom_text(aes(label = sprintf("%0.2f", pval))) +
  scale_fill_manual(values = c("gray", "red")) +
  scale_y_discrete(limits = rev) +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks.x = element_blank()) +
  ylab("") 
```

# Figure 2

Figure 2. Osteosarcoma cells adopt distinct transcriptional profiles as they colonize tibia and lung microenvironments. A) Schematic of study design depicting generation of orthotopic primary and metastatic tumors. Tumors were harvested for scRNA-seq when mice reached endpoint.

*Seurat object loading and processing code can be found in "OSHetero2021_Load_Process.Rmd". Object outputs from that code are loaded in the code below.*

## Figure 2B

B) Venn diagrams showing overlap of differentially expressed genes that are up- or down- regulated in primary or metastatic tumors relative to corresponding starting population of cells (cell line or PDX flank tumors). Region outlined in red identifies differentially expressed genes shared across at least three models.

```{r Fig2b}
load("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/Data/OS_merged_postCCR.RData")

OS$cell_line <- stringr::str_remove(OS$src, "_.+") %>%
    stringr::str_replace("t143b", "143B") %>%
    stringr::str_replace("NCHOS2", "NCH-OS-2") %>%
    stringr::str_replace("NCHOS7", "NCH-OS-7")
OS$tissue <- stringr::str_remove(OS$src, ".+_") %>%
    stringr::str_replace("Flank", "Culture")

OS <- subset(OS, subset = cell_line != "OB")

Idents(OS) <- OS$tissue

de_out <- list()
for (tissue_name in c("Lung", "Tibia")) {
    for (cl_name in unique(OS$cell_line)) {
        message(tissue_name, " ", cl_name)
        temp_seurat <- subset(OS, subset = (tissue == tissue_name |
                                    tissue == "Culture") &
                                    cell_line == cl_name)
        temp_de_out <-
            FindMarkers(temp_seurat,
                        ident.1 = tissue_name,
                        ident.2 = "Culture",
                        min.pct = 0.1) %>%
            filter(p_val_adj <= 0.05)

        de_out[[tissue_name]][["up"]][[cl_name]] <-
            temp_de_out %>%
            filter(avg_log2FC > 0) %>%
            rownames()

        de_out[[tissue_name]][["down"]][[cl_name]] <-
            temp_de_out %>%
            filter(avg_log2FC < 0) %>%
            rownames()

        rm(temp_seurat, temp_de_out)
    }
}

figure_2b <- list()
for (tissue_name in c("Lung", "Tibia")) {
    for (direction in c("up", "down")) {
        figure_2b[[paste0(tissue_name,
                          " - ",
                          direction)]] <-
            ggvenn::ggvenn(de_out[[tissue_name]][[direction]],
                           fill_color = c("#0073C2FF",
                                          "#EFC000FF",
                                          "#868686FF",
                                          "#CD534CFF"),
                           stroke_size = 0.9,
                           set_name_size = 5,
                           text_size = 2.5,
                           show_percentage = TRUE) +
            ggtitle(paste0(tissue_name,
                           " - ",
                           direction)) +
            theme(plot.title = element_text(hjust = 0.5,
                                            vjust = -15,
                                            size = 20,
                                            face = "bold"),
                  plot.margin = margin(-3, 0, -3, 0, "cm"))

        # Fix the position of the bottom two set names
        figure_2b[[paste0(tissue_name,
                          " - ",
                          direction)]]$layers[[3]]$data$x <-
            c(-1, -0.8, 0.8, 1)
    }
}

# Put all four plots into one image
figure_2b_merged <-
    gridExtra::grid.arrange(grobs = figure_2b,
                            vp = viewport(width = 1, height = 0.9))
# Save plot
ggsave("figures/figure_2/figure_2b.png",
       plot = figure_2b_merged,
       width = 10,
       height = 10)

# Save RData obj
save(figure_2b, file = "figures/figure_2/figure_2b.RData")
```

```{r Fig2b_all, dependson="Fig2b", eval=F}
### Commented out in previous code, but possible combination plot of all Venn diagrams
# ggVennDiagram(x[1:4], label_alpha = 0.7, label = 'count', category.names = c('OS17', '143B',
# 'NCHOS2', 'NCHOS7')) + scale_fill_gradient(low = '#F4FAFE', high = '#4981BF')+
# theme(legend.title = element_text(color = 'black'), legend.position = 'right')
### Extract genes that are shared in these datasets
genes <- Lung.down
A = genes[[1]]
B = genes[[2]]
C = genes[[3]]
D = genes[[4]]
# Preparing clusterProfiler to perform hypergeometric test on msigdb signatures m_t2g.c2 <-
# msigdbr(species = 'Homo sapiens', category = 'C2') %>% dplyr::select(gs_name,
# human_gene_symbol) m_t2g.c6 <- msigdbr(species = 'Homo sapiens', category = 'C6') %>%
# dplyr::select(gs_name, human_gene_symbol)
m_t2g.h <- msigdbr(species = "Homo sapiens", category = "H") %>%
    dplyr::select(gs_name, human_gene_symbol)
m_t2n.h <- msigdbr(species = "Homo sapiens", category = "H") %>%
    dplyr::select(gs_id, gs_name)
# m_t2g=rbind(m_t2g.c2,m_t2g.c6)
# msigdb signature to use
msig.gene.set = m_t2g.h
msig.name = m_t2n.h
intersect <- c((intersect(intersect(intersect(A, B), C), D)), (intersect(intersect(A, B), C)), (intersect(intersect(A,
    B), D)), (intersect(intersect(A, D), C)), (intersect(intersect(B, D), C)))
tmp <- enricher(intersect, TERM2GENE = msig.gene.set, TERM2NAME = msig.name)
em = tmp@result[, c("ID", "p.adjust")]  #pvalue/p.adjust
rownames(em) <- NULL
em <- em[1:10, ]
em[, 2] <- -log10(em[, 2])
# barplot(em)
pathways <- list(NULL)
pathways[[4]] <- em
# write.xlsx(pathways, file
# ='R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/pathways.xlsx', col.names = TRUE,
# row.names = TRUE, append = FALSE) save.image(file
# ='R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/AllpseudoBulk_Venn.RData')
# write.xlsx(pathways, file ='pathways.xlsx', col.names = TRUE, row.names = TRUE, append =
# FALSE) save.image(file ='AllpseudoBulk_Venn.RData')
#################### Intersection of deferentially regulated genes between Bone colonization
#################### and Lung colonization total <- list( Tibia.up = Tibia.up, Tibia.down =
#################### Tibia.down, Lung.up = Lung.up, Lung.down = Lung.down ) all.genes <-
#################### list(NULL) for (i in 1:4) { genes <- total[[i]] A = genes[[1]] B =
#################### genes[[2]] C = genes[[3]] D = genes[[4]] intersect <-
#################### intersect(intersect(intersect(A,B),C),D) all.genes[[i]] <- intersect } x
#################### <- list( A = all.genes[[1]], B = all.genes[[2]], C = all.genes[[3]], D =
#################### all.genes[[4]] ) pdf('TibiaUpdown_Lungupdown.pdf', width = 5, height = 5)
#################### plot <- ggvenn( x, fill_color = c('#0073C2FF', '#EFC000FF', '#868686FF',
#################### '#CD534CFF'), stroke_size = 0.9, set_name_size = 4, show_percentage =
#################### TRUE ) plot dev.off()
```

C) Pathway enrichment analysis with adjusted p values for hallmark gene sets associated with these shared genes

```{r Fig2c, dependson = "Fig2b", fig.height=6, fig.width=10}
### Extract genes that are shared in these datasets
#genes <- Lung.down
#A = genes[[1]]
#B = genes[[2]]
#C = genes[[3]]
#D = genes[[4]]
# Preparing clusterProfiler to perform hypergeometric test on msigdb signatures m_t2g.c2 <-
# msigdbr(species = 'Homo sapiens', category = 'C2') %>% dplyr::select(gs_name,
# human_gene_symbol) m_t2g.c6 <- msigdbr(species = 'Homo sapiens', category = 'C6') %>%
# dplyr::select(gs_name, human_gene_symbol)
#m_t2g.h <- msigdbr(species = "Homo sapiens", category = "H") %>%
#    dplyr::select(gs_name, human_gene_symbol)
#m_t2n.h <- msigdbr(species = "Homo sapiens", category = "H") %>%
#    dplyr::select(gs_id, gs_name)
# m_t2g=rbind(m_t2g.c2,m_t2g.c6)
# msigdb signature to use
#msig.gene.set = m_t2g.h
#msig.name = m_t2n.h
#intersect <- c((intersect(intersect(intersect(A, B), C), D)),
#               (intersect(intersect(A, B), C)),
#               (intersect(intersect(A, B), D)),
#               (intersect(intersect(A, D), C)),
#               (intersect(intersect(B, D), C)))
#tmp <- enricher(intersect,
#                TERM2GENE = msig.gene.set,
#                TERM2NAME = msig.name)
#pvalue/p.adjust
#em = tmp@result[, c("ID", "p.adjust")]
#rownames(em) <- NULL
#em <- em[1:10, ]
#em[, 2] <- -log10(em[, 2])

#pathways <- list(NULL)
#pathways[[4]] <- em

################ Two sided Barplot
SanjanaPlots <- read.delim("SanjanaPlots.txt") %>%
    mutate(Tissue = stringr::str_replace(Tissue,
                                "Tibia",
                                "Tibia Gene Sets") %>%
                stringr::str_replace("Lung", "Lung Gene Sets"),
            Pathway = stringr::str_wrap(Pathway, width = 25))

figure_2c <-
    ggplot() +
    geom_bar(data = SanjanaPlots,
            aes(x = -1 * Order,
                y = p.adjust,
                fill = p.adjust > 0),
            stat = "identity",
            alpha = 0.8) +
    coord_flip() +
    facet_wrap(~Tissue,
                ncol = 1,
                scales = "free") +
    geom_hline(yintercept = c(-1.5, 1.5),
               color = "gray",
               linetype = 2,
               size = 1) +
    geom_hline(yintercept = 0,
               color = "black",
               size = 1) +
    geom_text(data = SanjanaPlots,
              aes(x = -1 * Order,
                  y = Label_y * 3,
                  label = Pathway)) +
    geom_text(data = tibble(y = c(-2.5, 2.5, -2.5, 2.5),
                            x = c(0.5, 0.5, 0.5, 0.5),
                            label = c("Downregulated\nduring tibia colonization",
                                      "Upregulated\nduring tibia colonization",
                                      "Downregulated\nduring lung colonization",
                                      "Upregulated\nduring lung colonization"),
                            Tissue = c("Tibia Gene Sets",
                                       "Tibia Gene Sets",
                                       "Lung Gene Sets",
                                       "Lung Gene Sets")),
            aes(x = x,
                y = y,
                label = label),
            fontface = "bold") +
    scale_fill_manual(values = c("#377eb8", "#e41a1c")) +
    theme_bw() +
    theme(strip.background = element_rect(color = "white", fill = "white"),
          strip.text.x = element_text(size = 20, face = "bold"),
          legend.position = "none",
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_blank(),
          axis.line.x = element_line(color = "black"),
          axis.text.y = element_blank(),
          axis.ticks.y = element_blank()) +
    labs(title = "Tibia Gene Sets",
         y = "",
         x = "") +
    ylim(-5, 5) +
    xlim(-5.5, 1)

ggsave("figures/figure_2/figure_2c.png",
       plot = figure_2c,
       width = 6,
       height = 9)

save(figure_2c, file = "figures/figure_2/figure_2c.RData")
```

```{r assembleFig2}
#patchwork stuff
```

# Figure 3

Figure 3. Osteosarcoma cells retain phenotypic heterogeneity despite adaptive changes in response to changing microenvironments. A) Schematic outlining scRNA-seq bioinformatics workflow. In B-D, samples subset to equal number of cells to allow inter-model comparison (n=1500 per condition).

```{r Fig3_load}
load("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/Data/OS.listv3.CCR.RData")
```

## Figure 3B

*statistical method subject to change*

B) Osteosarcoma cells exhibited higher ITH scores than that of osteoblasts grown in cell culture (####p <0.001). Within model comparison, identified increase in ITH scores as cells colonize tibia and lung microenvironments, with the exception of NCH-OS-7 (****p <0.001). P values were adjusted using Šidák multiple comparisons test.

```{r Fig3b, dependson='Fig3_load', eval=F}
ITH.score <- list()
for (i in 1:length(OS.list)) {
  z <- (Embeddings(OS.list[[i]], reduction = "pca")[, 1:20])
  mat <- dist(z, diag = TRUE, upper = FALSE)
  mat2 <- as.matrix(mat)
  mat2[upper.tri(mat2, diag = TRUE)] <- NA
  ITH.score[[i]] <- colMeans(mat2, na.rm = TRUE)
}

library(xlsx)
for (i in c(1:12)){
  write.xlsx(ITH.score[i],
             file=paste(i,"test.xlsx"))
}

raw_ith <- read_pzfx("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/ITH_Scores_1500.pzfx",
                     table = 1)

ith <- melt(raw_ith)
colnames(ith) <- c("Sample", "ITHScore")
ith$logITH <- log10(ith$ITHScore)
ridge_cols <- c("#B8B8B8FF",
                "#D43F3AFF", "#D43F3AFF", "#D43F3AFF",
                "#EEA236FF", "#EEA236FF", "#EEA236FF",
                "#357EBDFF", "#357EBDFF", "#357EBDFF",
                "#5CB85CFF", "#5CB85CFF", "#5CB85CFF"
)
rplot <- ggplot(ith, aes(x = logITH, y = Sample, fill = Sample)) +
  geom_density_ridges(scale = 2) +
  scale_y_discrete(expand = c(0, 0), limits = rev) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_fill_manual(values = ridge_cols) +
  theme_ridges() +
  xlab("log[ITH Score]") +
  theme(legend.position = "none",
        axis.title.x = element_text(hjust = 0.5),
        axis.title.y = element_blank())
# Perform a Mann-Whitney U for each group
comps <- tribble(
  ~ctl, ~comp,
  1, 2,
  1, 5,
  1, 8,
  1, 11,
  2, 3,
  2, 4,
  5, 6,
  5, 7,
  8, 9,
  8, 10,
  11, 12,
  11, 13
)
mwu <- matrix(nrow = 13, ncol = 13)
for(i in seq_along(comps$ctl)) {
  s <- wilcox.test(raw_ith[,comps$ctl[i]], raw_ith[,comps$comp[i]])
  mwu[comps$ctl[i], comps$comp[i]] <- s$p.value
}
efs <- matrix(nrow = 13, ncol = 13)
for(i in seq_along(comps$ctl)) {
  s <- cohen.d(raw_ith[,comps$ctl[i]], raw_ith[,comps$comp[i]])
  efs[comps$ctl[i], comps$comp[i]] <- s$estimate
}
per <- matrix(nrow = 13, ncol = 13)
for(i in seq_along(comps$ctl)) {
  s <- permTS(raw_ith[,comps$ctl[i]], raw_ith[,comps$comp[i]], alternative = "two.sided")
  per[comps$ctl[i], comps$comp[i]] <- s$p.value
}
```

## Figure 3C

C) UMAP analysis for merged primary and metastatic samples in each of the four models. Cells in grey represent remaining cells in merged sample. Cluster enrichment analysis shows distribution of cells in each cluster in the two microenvironment conditions (tibia, lung). While some cells in the primary and metastatic lesions adopted shared phenotypes, others adopted distinct phenotypes.

```{r Fig3c, dependson='Fig3_load', eval=F}
# Isolate tibia and lung Seurat objects from the OS.list of Seurat objects
OS17_Tibia <- OS.list$OS17.Tibia
OS17_Lung <- OS.list$OS17.Lung

t143B_Tibia <- OS.list$t143B.Tibia
t143B_Lung <- OS.list$t143B.Lung

OS2_Tibia <- OS.list$OS2.Tibia
OS2_Lung <- OS.list$OS2.Lung

OS7_Tibia <- OS.list$OS7.Tibia
OS7_Lung  <- OS.list$OS7.Lung

# Merge tibia and lung objects into a single Seurat object
OS17_TL <- merge(OS17_Tibia,
                 y = c(OS17_Lung),
                 add.cell.ids = c("OS17_Tibia",
                                  "OS17_Lung"),
                 project = "Heterogeneity")
t143b_TL <- merge(t143B_Tibia,
               y = c(t143B_Lung),
               add.cell.ids = c("t143b_Tibia",
                                "t143b_Lung"),
               project = "Heterogeneity")
OS2_TL <- merge(OS2_Tibia,
             y = c(OS2_Lung),
             add.cell.ids = c("OS2_Tibia",
                              "OS2_Lung"),
             project = "Heterogeneity")
OS7_TL <- merge(OS7_Tibia,
             y = c(OS7_Lung),
             add.cell.ids = c("OS7_Tibia",
                              "OS7_Lung"),
             project = "Heterogeneity")

# Process and cluster

#OS17
OS17_TL <- RunPCA(OS17_TL,
                  pc.genes = OS17_TL@var.genes,
                  npcs = 20) %>%
  RunUMAP(reduction = "pca",
          dims = 1:20) %>%
  FindNeighbors(reduction = "pca",
                dims = 1:20) %>%
  FindClusters(resolution = 0.3)

#t143B
t143b_TL <- RunPCA(t143b_TL,
                pc.genes = t143b_TL@var.genes,
                npcs = 20) %>%
  RunUMAP(reduction = "pca",
          dims = 1:20) %>%
  FindNeighbors(reduction = "pca",
                dims = 1:20) %>%
  FindClusters(resolution = 0.3)

#OS2
OS2_TL <- RunPCA(OS2_TL,
              pc.genes = OS2_TL@var.genes,
              npcs = 20) %>%
  RunUMAP(reduction = "pca",
          dims = 1:20) %>%
  FindNeighbors(reduction = "pca",
                dims = 1:20) %>%
  FindClusters(resolution = 0.3)

#OS7
OS7_TL <- RunPCA(OS7_TL,
              pc.genes = OS7_TL@var.genes,
              npcs = 20) %>%
  RunUMAP(reduction = "pca",
          dims = 1:20) %>%
  FindNeighbors(reduction = "pca",
                dims = 1:20) %>%
  FindClusters(resolution = 0.3)

# Cluster distribution in each sample
# OS17
table(Idents(OS17_TL),
      OS17_TL$src)
prop <- prop.table(table(Idents(OS17_TL),
                         OS17_TL$src),
                   margin = 2)*100
colSums(prop)
prop
#     Tibia       Lung
# 0   50.263250   46.577747
# 1   43.453843   10.179010
# 2   1.053001   39.698140
# 3   5.229905   3.545104
#t143b
table(Idents(t143b_TL),
      t143b_TL$src)
prop <- prop.table(table(Idents(t143b_TL),
                         t143b_TL$src),
                   margin = 2)*100
colSums(prop)
prop
#     Tibia        Lung
# 0   97.8963771    2.2204908
# 1   1.1492014   92.3061940
# 2   0.4479938   3.3502143
# 3   0.5064277   2.1231009
# OS2
table(Idents(OS2_TL),
      OS2_TL$src)
prop <- prop.table(table(Idents(OS2_TL),
                         OS2_TL$src),
                   margin = 2)*100
colSums(prop)
prop
#     Tibia       Lung
# 0   82.281553   20.327670
# 1   5.339806    71.480583
# 2   12.378641   8.191748
# OS7
table(Idents(OS7_TL),
      OS7_TL$src)
prop <- prop.table(table(Idents(OS7_TL),
                         OS7_TL$src),
                   margin = 2)*100
colSums(prop)
prop
#     Tibia       Lung
# 0    0.000000   71.013490
# 1   59.425804    1.383604
# 2   40.574196    1.176064
# 3    0.000000   26.426842

#Modify data to account for optimal number of clusters using nRes and pSil functions
OS17_TL.nRes <- nRes(OS17_TL, 
                     res = seq(from = 0.1,
                               to = 0.2,
                               by = 0.05))
OS17_TL.nRes
plot <- pSil(OS17_TL.nRes,
             0.15)
OS17_TL <- FindClusters(OS17_TL,
                        resolution = 0.20)
DimPlot(OS17_TL,
        reduction = "umap",
        pt.size = 1,
        label = T,
        split.by = "src") +
  coord_fixed() +
  NoLegend() +
  NoAxes()

#Modify data to account for optimal number of clusters using nRes and pSil functions
t143b_TL.nRes <- nRes(t143b_TL, 
                     res = seq(from = 0.1,
                               to = 0.4,
                               by = 0.05))
t143b_TL.nRes
plot <- pSil(t143b_TL.nRes,
             0.15)
t143b_TL <- FindClusters(t143b_TL,
                      resolution = 0.15)
DimPlot(t143b_TL,
        reduction = "umap",
        pt.size = 1,
        label = T,
        split.by = "src") +
  coord_fixed()

#Modify data to account for optimal number of clusters using nRes and pSil functions
OS2_TL.nRes <- nRes(OS2_TL, 
                     res = seq(from = 0.1,
                               to = 0.4,
                               by = 0.05))
plot <- pSil(OS2_TL.nRes, 0.15)
OS2_TL <- FindClusters(OS2_TL, resolution = 0.15)
DimPlot(OS2_TL,
        reduction = "umap",
        pt.size = 1,
        label = T,
        split.by = "src") +
  coord_fixed() +
  NoLegend() +
  NoAxes()

#Modify data to account for optimal number of clusters using nRes and pSil functions
OS7_TL.nRes <- nRes(OS7_TL,
                 res = seq(from = 0.1,
                           to = 0.2,
                           by = 0.05))
OS7_TL.nRes
plot <- pSil(OS7_TL.nRes,
             0.1)
OS7_TL <- FindClusters(OS7_TL,
                    resolution = 0.1)
DimPlot(OS7_TL,
        reduction = "umap",
        pt.size = 1,
        label = T,
        split.by = "src") +
  coord_fixed() +
  NoLegend() +
  NoAxes()

OS17_TL <- OS17_TL %>% FindClusters(resolution = 0.2)
t143b_TL <- t143b_TL %>% FindClusters(resolution = 0.15)
OS2_TL <- OS2_TL %>% FindClusters(resolution = 0.15)
OS7_TL <- OS7_TL %>% FindClusters(resolution = 0.2)

# Generate PDF files for the Figure 3 plots
# Previous code used a subset of 3000 (contrary to description of 1500 cells for Figure 3 in paper)
pdf("figures/figure_3/OS17_TL_fig3.pdf", width = 5, height = 5)
OS17_TL %>% FindClusters(resolution = 0.2) %>% 
  DimPlot(reduction = "umap", pt.size = 1, label = T, split.by = "src") + 
  coord_fixed() + NoLegend() + NoAxes()+
  scale_color_npg(alpha = 1)
dev.off()

pdf("figures/figure_3/t143b_fig3.pdf", width = 5, height = 5)
t143b_TL %>% FindClusters(resolution = 0.15) %>% 
  DimPlot(reduction = "umap", pt.size = 1, label = T, split.by = "src") + 
  coord_fixed() + NoLegend() + NoAxes()+
  scale_color_npg(alpha = 1)
dev.off()

pdf("figures/figure_3/OS2_fig3.pdf", width = 5, height = 5)
OS2_TL %>% FindClusters(resolution = 0.15) %>% 
  DimPlot(reduction = "umap", pt.size = 1, label = T, split.by = "src") + 
  coord_fixed() + NoLegend() + NoAxes()+
  scale_color_npg(alpha = 1)
dev.off()

pdf("figures/figure_3/OS7_fig3.pdf", width = 10, height = 10)
OS7_TL %>% FindClusters(resolution = 0.2) %>% 
  DimPlot(reduction = "umap", pt.size = 1, label = T, split.by = "src") +
  ggtitle("OS7")+
  coord_fixed() + NoLegend() + NoAxes()+
  scale_color_npg(alpha = 1)
dev.off()
```

## Figure 3D

D) Metabolic heterogeneity in glycolysis activation. We used a pathway enrichment analysis for hallmark gene sets using genes differentially upregulated in each cluster relative to every other cluster within the same model. P values were adjusted for multiple comparisons. Boxes in grey identify non-significant pathway enrichments, whereas boxes in red identify statistically significant enrichments. Bar plot shows percentage of cells identified per cluster in lung metastases. In B-D, samples subset to equal number of cells to allow inter-model comparison (n=1500 per condition).

### Processing

```{r Fig3d_process, dependson='Fig3c', eval=F}
library(data.table)
library(pheatmap)
#Create list of objects used in Figure 3 heatmap 
B.list <- list(OS17 = OS17_TL,
               t143b = t143b_TL,
               OS2 = OS2_TL,
               OS7 = OS7_TL)
# Use DGEA function to find hallmark pathway expression
em.hm.list <- list()
for (i in 1:length(B.list)) {
  
  em.hm.list[[i]] <- DGEA(B.list[[i]])
  
}
for(i in 1:(length(em.hm.list))) {
  em.hm.list[[i]] <- setDT(em.hm.list[[i]],
                           keep.rownames = TRUE)[]
  
}
temp1 <- em.hm.list[[1]]
temp2 <- em.hm.list[[2]]
temp3 <- em.hm.list[[3]]
temp4 <- em.hm.list[[4]]
cx <- inner_join(temp1, temp2, by = "rn")
pd <- inner_join(temp3, temp4, by = "rn")
cx.pd <- inner_join(cx, pd, by = "rn")
head(cx.pd)
##Write table to edit rownames to easily convert to dataframe
write.table(cx.pd, file=('Figure 3/cxpd.tsv'), quote=FALSE, sep='\t')
cx.pd <- read.table(file = "Figure 3/cxpd.tsv", header=T)
# Make the "rn" column the rownames -Emily
cx.pd <- remove_rownames(cx.pd) %>%
  column_to_rownames(var = "rn")
#remove "HALLMARK_"
c <- rownames(cx.pd)
c <- gsub("HALLMARK_", "", c)
#remove "_" by removing special characters
c <- gsub("_", " ", c)
rownames(cx.pd) <- c
#log transform
cx.pd.log <- -log10(cx.pd)
cx.pd_transpose <- as.data.frame(t(cx.pd.log))
df <- as.matrix(cx.pd_transpose)
#Write table to print supplemental information
write.table(df, file=('Figure 3/cxpd_plot.tsv'), quote=FALSE, sep='\t')
subj<-c(paste0("A", 0:3),
        paste0("B", 0:3),
        paste0("C", 0:2),
        paste0("D", 0:3))
rownames(df)<-subj
aka2 = data.frame(ID = factor(c(rep("OS17", 4),
                                rep("t143B", 4),
                                rep("OS2", 3),
                                rep("OS7", 4)
)))
rownames(aka2)<-subj
aka3 = list(ID = c(OS17 = "#E64B35FF",
                   t143B = "#4DBBD5FF",
                   OS2 = "#00A087FF",
                   OS7 = "#3C5488FF"))
library(ComplexHeatmap)
color = colorRampPalette(rev(brewer.pal(n = 7, name =
                                          "RdYlBu")))(100)
#calculate percentage of cells in lung for each cluster
aka4 <- data.frame (Lpercent = c(46.577747,
                                 10.17901,
                                 39.69814,
                                 3.545104,
                                 2.2204908,
                                 92.306194,
                                 3.3502143,
                                 2.1231009,
                                 20.32767,
                                 71.480583,
                                 8.191748,
                                 71.01349,
                                 1.383604,
                                 1.176064,
                                 26.426842))
rownames(aka4) <- subj
aka5 = (c(rep("OS17", 4),
          rep("t143B", 4),
          rep("OS2", 3),
          rep("OS7", 4)))

# cx.pd <- cx.pd[-(23:30),] #remove rows with names NA
cx.pd.log <- -log10(cx.pd) #log transform
library(pheatmap)   
cx.pd_transpose <- as.data.frame(t(cx.pd.log))
df <- as.matrix(cx.pd_transpose)
subj <- c(paste0("A", 0:3),
        paste0("B", 0:3),
        paste0("C", 0:2),
        paste0("D", 0:3))
rownames(df) <- subj
aka2 = data.frame(ID = factor(c(rep("OS17", 4),
                                rep("t143B", 4),
                                rep("OS2", 3),
                                rep("OS7", 4)
)))
rownames(aka2)<-subj
aka3 = list(ID = c(OS17 = "#E64B35FF",
                   t143B = "#4DBBD5FF",
                   OS2 = "#00A087FF",
                   OS7 = "#3C5488FF"))
# df[] <- lapply(df, gsub, pattern='HALLMARK_', replacement='')
# df
```

### Heatmap

```{r Fig3d_heatmap, dependson='Fig3d_process', eval=F}
# Pathway enrichment analysis 
P1 <- df %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  full_join(aka4 %>%
              rownames_to_column(var = "Sample")) %>%
  pivot_longer(c(-Sample,
                 -Lpercent),
               names_to = "Pathway",
               values_to = "pval",
               names_repair = "minimal") %>%
  mutate(Signif = pval >= (-1 * log10(0.05)),
         sample_order = str_remove(Sample,
                                   "[0-9]") %>%
           rank() * 1000 - Lpercent,
         Sample = reorder(Sample,
                          sample_order)) %>%
  ggplot(.,
         aes(x = Sample,
             y = Pathway,
             fill = Signif)) +
  geom_tile() +
  geom_text(aes(label = sprintf("%0.2f",
                                pval))) +
  scale_fill_manual(values = c("gray",
                               "#CC3333")) +
  scale_y_discrete(limits = rev) +
  theme(legend.position = "none") +
  ylab("")
P2 <- aka4 %>%
  rownames_to_column(var = "Sample") %>%
  full_join(aka2 %>%
              rownames_to_column(var = "Sample")) %>%
  mutate(sample_order = str_remove(Sample,
                                   "[0-9]") %>%
           rank() * 1000 - Lpercent,
         Sample = reorder(Sample,
                          sample_order),
         ID = reorder(ID, sample_order)) %>%
  ggplot(.,
         aes(x = Sample,
             y = Lpercent,
             fill = ID)) +
  geom_bar(stat = "identity") +
  ylab("Percent\nin lung") +
  xlab("") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  theme(legend.position = "top", ) +
  labs(fill = "")
# Plot combined bar plot (percentage of cells identified per cluster in lung metastases) and heatmap of hallmark pathway expressions
pdf("figures/figure_3/figure_3d.pdf", height = 12, width = 12)
cowplot::plot_grid(P2,
                   P1,
                   ncol = 1,
                   align = "v",
                   rel_heights = c(2, 10))
dev.off()
```

# Figure 4 - Images

Figure 4. Heterogeneity in Glycolysis activation identified within metastatic osteosarcoma lesions.

Figure 4 of this paper is composed of microscopy images and so is not part of the code for this paper. 

# Figure 5

Figure 5. Lung colonization bottleneck enriches for clonal populations with a common transcriptional phenotype. A) Schematic of experimental set-up. B) UMAP analysis of lineage tagged OS-17 cells in culture, implanted into tibia, or inoculated to generate lung metastases. These lineage tags are heritable allowing us to track the transcriptional profiles of daughter cells originating from the same ancestor. Metastatic and orthotopic samples share many cells with similar phenotypes, though cultured cells are dramatically different (n = 2800 per condition) C) Frequency distribution of clones identified in each of the conditions. A high level of clonal diversity is maintained in the orthotopic tumors (78% unique clones in primary tumor compared to 86% unique clones in cell culture); though dominant clones emerge in the lung (only 33% unique clones). D) Overlay of cells sharing a common ancestor onto the dimensional reduction plots in the top four enriched clonal families. E) UMAP visualization of cells belonging to top ten enriched clonal families highlighted in red. F) Pathway enrichment analysis for hallmark gene sets comparing enriched clones relative to the remaining metastatic tumor cells identifies Hypoxia, Glycolysis, EMT and MTORC1 signaling related genes to be significantly upregulated

### Load

```{r Fig5load}
source("scSeurat.R")

# Load raw objects for lineage tag analysis
load("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/Data/Raw/os17.cx.raw.RData")
load("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/Data/Raw/os17.tib.raw.RData")
load("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/Data/Raw/os17.lung.raw.RData")

# Load merged and cell cycle regressed OS17 object
load("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/Data/os17.RData")
```

## Figure 5B

B) UMAP analysis of lineage tagged OS-17 cells in culture, implanted into tibia, or inoculated to generate lung metastases. These lineage tags are heritable allowing us to track the transcriptional profiles of daughter cells originating from the same ancestor. Metastatic and orthotopic samples share many cells with similar phenotypes, though cultured cells are dramatically different (n = 2800 per condition)

```{r Fig5b, dependson="Fig5load"}
# Set cell identities to proper cluster resolution 
Idents(os17) <- os17$RNA_snn_res.0.3

# Plot os17
set.seed(100)
cell.ids <- sample(colnames(os17))
figure_5b <- DimPlot(os17,
                     reduction = "umap",
                     group.by = "src",
                     pt.size = 1,
                     label = F,
                     order = cell.ids) +
  coord_fixed() +
  ggtitle("OS17 by Source") +
  scale_color_npg()

figure_5b
```

## Figure 5C

C) Frequency distribution of clones identified in each of the conditions. A high level of clonal diversity is maintained in the orthotopic tumors (78% unique clones in primary tumor compared to 86% unique clones in cell culture); though dominant clones emerge in the lung (only 33% unique clones). 

```{r Fig5c, dependson="Fig5load"}
# Plot histograms of lineage tracing tags in the Seurat objects
figure_5c1 <- processLTBC(os17.cx.raw,
                          lt.loc = "/gpfs0/home/gdrobertslab/lab/Counts/S0016/outs/lt.fq",
                          cid.loc = "/gpfs0/home/gdrobertslab/lab/Counts/S0016/outs/cid.fq",
                          histogram = T,
                          plot.only = T,
                          title = "Culture, Top 40 Clones",
                          ymax = 1.75,
                          col.fill = "#E64B35FF",
                          relative = T)

figure_5c2 <- processLTBC(os17.tib.raw,
                          lt.loc = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/lt.fq",
                          cid.loc = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/cid.fq",
                          histogram = T,
                          plot.only = T,
                          title = "Tibia, Top 40 Clones",
                          ymax = 1.75,
                          col.fill = "#00A087FF",
                          relative = T)

figure_5c3 <- processLTBC(os17.lung.raw,
                          lt.loc = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/lt.fq",
                          cid.loc = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/cid.fq",
                          histogram = T,
                          plot.only = T,
                          title = "Lung, Top 40 Clones",
                          ymax = 1.75,
                          relative = T)
```

Determine number of unique clones for each tissue.

```{r uniquelt, dependson='Fig5b'}
# Determine number of unique LTs
length((os17.cx.raw$lt))
length(unique(os17.cx.raw$lt))
(length(unique(os17.cx.raw$lt))/length((os17.cx.raw$lt)))*1000

length((os17.tib.raw$lt))
length(unique(os17.tib.raw$lt))
(length(unique(os17.tib.raw$lt))/length((os17.tib.raw$lt)))*1000

length((os17.lung.raw$lt))
length(unique(os17.lung.raw$lt))
(length(unique(os17.lung.raw$lt))/length((os17.lung.raw$lt)))*1000
```

## Figure 5D

D) Overlay of cells sharing a common ancestor onto the dimensional reduction plots in the top four enriched clonal families.

### Processing

```{r Fig5d_processing, dependson="Fig5load"}
# Extract the barcode frequency lists
cx.lt.list <- processLTBC(os17.cx.raw,
                          lt.loc = "/gpfs0/home/gdrobertslab/rsexf001/OSHetero2021/S0016-zymo/lt.fq",
                          cid.loc = "/gpfs0/home/gdrobertslab/rsexf001/OSHetero2021/S0016-zymo/cid.fq",
                          ret.list = TRUE)

tib.lt.list <- processLTBC(os17.tib.raw,
                           lt.loc = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/lt.fq",
                           cid.loc = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/cid.fq",
                           ret.list = TRUE)

lung.lt.list <- processLTBC(os17.lung.raw,
                            lt.loc = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/lt.fq",
                            cid.loc = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/cid.fq",
                            ret.list = TRUE)

# Subset the three samples, but retain the current umap data
cx.sub <- subset(os17, subset = src == "Culture")
tib.sub <- subset(os17, subset = src == "Tibia")
lung.sub <- subset(os17, subset = src == "Lung")

# Subset to cells that have a lineage tag
# Culture
Culture.LTcells <- vector(mode = "character")
for (i in 1:nrow(cx.lt.list)) {
  temp <- WhichCells(cx.sub,
                     expression = lt == cx.lt.list[[i, 1]])
  Culture.LTcells <- append(Culture.LTcells,
                            temp)
}
cx.tagged <- subset(cx.sub,
                    cells = Culture.LTcells)

# Tibia
# Remove outlier cells (names may vary in different runs of the code)
tib.sub <- subset(tib.sub,
                  idents = c("0",
                             "1",
                             "3",
                             "6")) 

tib.LTcells <- vector(mode = "character")
for (i in 1:nrow(tib.lt.list)) {
  temp <- WhichCells(tib.sub,
                     expression = lt == tib.lt.list[[i, 1]])
  tib.LTcells <- append(tib.LTcells, temp)
}
tib.tagged <- subset(tib.sub,
                     cells = tib.LTcells)

# Lung
# Remove outlier cells (names may vary in different runs of the code)
lung.sub <- subset(lung.sub,
                   idents = c("0",
                              "1",
                              "3",
                              "6"))

lung.LTcells <- vector(mode = "character")
for (i in 1:nrow(lung.lt.list)) {
  temp <- WhichCells(lung.sub,
                     expression = lt == lung.lt.list[[i, 1]])
  lung.LTcells <- append(lung.LTcells,
                         temp)
}
lung.tagged <- subset(lung.sub,
                      cells = lung.LTcells)
```


### Culture Panel

```{r Fig5d_cx, dependson='Fig5d_processing', fig.height=10, fig.width=10}
# Culture
vplayout <- function(x,y) viewport(layout.pos.row = x,
                                   layout.pos.col = y)

png("figures/figure_5/figure_5d1_grid_culture.png",
    units = "in",
    width = 10,
    height = 6,
    res = 300)

grid.newpage()

pushViewport(viewport(layout = grid.layout(1,4)))

for (i in 1:4) {
  p <- DimPlot(cx.tagged,
               reduction = "umap",
               pt.size = 1,
               cells.highlight = WhichCells(cx.sub,
                                            expression = lt == cx.lt.list[[i, 1]]),
               cols.highlight = "#E64B35FF",
               sizes.highlight = 3) +
    coord_fixed() +
    theme(legend.position = "none") +
    ggtitle(paste("Clone",
                  cx.lt.list[[i,1]])) +
    NoLegend() +
    NoAxes()

  print(p, vp = vplayout(ceiling(i/4), i))
}
dev.off()
```

### Tibia Panel

```{r Fig5d_tibia, dependson='Fig5d_processing'}
# Tibia
vplayout <- function(x,y) viewport(layout.pos.row = x,
                                   layout.pos.col = y)

png("figures/figure_5/figure_5d2_grid_tibia.png",
    units = "in",
    width = 10,
    height = 6,
    res = 300)

grid.newpage()

pushViewport(viewport(layout = grid.layout(1,4)))

for (i in 1:4) {
  p <- DimPlot(tib.tagged,
               reduction = "umap",
               pt.size = 1,
               cells.highlight = WhichCells(tib.sub,
                                            expression = lt == tib.lt.list[[i, 1]]),
               cols.highlight = "#00A087FF",
               sizes.highlight = 3) +
    coord_fixed() +
    theme(legend.position = "none") +
    ggtitle(paste("Clone",
                  tib.lt.list[[i,1]])) +
    NoLegend() +
    NoAxes()
  
  print(p, vp = vplayout(ceiling(i/4), i))
}
dev.off()
```

### Lung Panel

```{r Fig5d_lung, dependson='Fig5d_processing'}
# Lung
vplayout <- function(x,y) viewport(layout.pos.row = x,
                                   layout.pos.col = y)

png("figures/figure_5/figure_5d3_grid_lung.png",
    units = "in",
    width = 10,
    height = 6,
    res = 300)

grid.newpage()

pushViewport(viewport(layout = grid.layout(1,4)))

for (i in 1:4) {
  p <- DimPlot(lung.tagged,
               reduction = "umap",
               pt.size = 1,
               cells.highlight = WhichCells(lung.sub,
                                            expression = lt == lung.lt.list[[i, 1]]),
               cols.highlight = "#4DBBD5FF",
               sizes.highlight = 3) +
    coord_fixed() +
    theme(legend.position = "none") +
    ggtitle(paste("Clone",
                  lung.lt.list[[i,1]])) +
    NoLegend() +
    NoAxes()
  
  print(p, vp = vplayout(ceiling(i/4), i))
}
dev.off()
```

## Figure 5E

E) UMAP visualization of cells belonging to top ten enriched clonal families highlighted in red.

```{r Fig5e, dependson='Fig5d_processing'}
# Compare Enriched vs non-enriched 
# Group first 10 against the tagged Lung
LT <- vector(mode = "character")

for (i in 1:10) {
  temp <- WhichCells(lung.tagged,
                     expression = lt == lung.lt.list[[i, 1]])
  LT <- append(LT, temp)
}

Idents(lung.tagged, cells = LT) <- "Enriched clones"

cell.ids <- sample(colnames(lung.tagged))

# Culture plots
figure_5e <- DimPlot(lung.sub,
                     reduction = "umap",
                     pt.size = 1,
                     cells.highlight = WhichCells(lung.tagged,
                                                  idents = "Enriched clones"),
                     cols.highlight = "#E64B35FF",
                     sizes.highlight = 2,
                     order = cell.ids) +
  coord_fixed() + 
  scale_color_discrete(name = "Clones",
                       type = c("#C3C3C3",
                                "#E64B35FF"),
                       labels = c("Unselected" = "Non-enriched clones",
                                  "Group_1" = "Enriched clones"))
figure_5e
```

## Figure 5F

F) Pathway enrichment analysis for hallmark gene sets comparing enriched clones relative to the remaining metastatic tumor cells identifies Hypoxia, Glycolysis, EMT and MTORC1 signaling related genes to be significantly upregulated

```{r Fig5f, dependson='Fig5e'}
# Pathways associated with Enriched clonal families
source("Downstream.v2.R")

# Rename identities from the total lung tumor to include enriched clones
lung.sub.enr <- SetIdent(lung.sub,
                         cells = WhichCells(lung.tagged,
                                            idents = "Enriched clones"),
                         value = "Enriched clones")

# Analyze enriched clones against the remaining tumor cells in the lung
lung.sub.enr <- RenameIdents(lung.sub.enr,
                             "0" = "Remaining Tumor Cells",
                             "1" = "Remaining Tumor Cells",
                             "3" = "Remaining Tumor Cells",
                             "6" = "Remaining Tumor Cells")

B.list <- list(OS17 = lung.sub.enr)

# Perform differential gene expression analysis for hallmark pathways
em.hm.list <- list()
for (i in 1:length(B.list)) {
  em.hm.list[[i]] <- DGEA(B.list[[i]])
}

# Achieve -log10 of the p-value
temp1 <- em.hm.list[[1]]
temp1 <- -log10(temp1)

# Remove unnecessary characters
c <- rownames(temp1)
c <- gsub("_", " ", c)
c <- gsub("HALLMARK ", "", c) 
rownames(temp1) <- c

# Remove negative 0 values due to number of decimal points displayed
temp1 <- abs(temp1)
temp1 <- temp1["Enriched clones"]

# Subset to only pathways with significant pvalue (-1 * log10(0.05)) = 1.30103)
temp1["Enriched clones"] >= 1.30103 -> temp1$logical
colnames(temp1) <- c("pvalue",
                     "significant")
logical <- temp1$significant
temp1 <- temp1[logical, ]
temp1 <- rownames_to_column(temp1,
                            var = "Pathway")

# Order by -log10pvalue for plotting
temp1$Pathway <- factor(temp1$Pathway,
                        levels = temp1$Pathway[order(temp1$pvalue)])

# Plot 
figure_5f <- ggplot(temp1,
                    aes(x = pvalue,
                        y = Pathway)) +
  geom_bar(stat = "identity") +
  scale_fill_brewer(type = "qual",
                    palette = "Set1") +
  theme_bw() +
  ylab("")

figure_5f
```

```{r session}
sessionInfo()
```
