---
title: "OSHetero2021"
author: "Sanjana Rajan and Emily Franz"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 5
    number_sections: false
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  echo = TRUE,
  cache = TRUE,
  collapse = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  cache.lazy = FALSE
)
```

# Figure 1

**Cell line and PDX models of osteosarcoma display transcriptional heterogeneity.**

```{r Fig1_lib, cache=FALSE}
setwd("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021")
source("scSeurat.R")
source("Downstream.v2.R")

# loading libraries
library(Seurat)
library(future)
library(ggplot2)
# library(fgsea)
library(msigdbr)
library(clusterProfiler)
library(SingleR)
library(dplyr) # %<%
library(pheatmap)
library(RColorBrewer)
library(viridis) # inferno color palette
library(grid)
library(ggsci)
library(data.table)
library(ggvenn)
```

## Figure 1A

*DONE*

A) UMAP analysis of OS-17 cells (n = 3178) grown in cell culture. Cell cycle distribution of cells (n = 3178) is visualized as pie charts.

```{r Fig1a_OS17_ccr}
# From CulturePDXHetero.R

# OS17
# Create Seurat objects and perform initial QC. Label original source.
cx.raw <- tenXLoadQC("/gpfs0/home/gdrobertslab/lab/Counts/S0016/outs/filtered_feature_bc_matrix/",
                     spec = "mixHuman")
cx.raw <- subset(cx.raw,
                 subset = nFeature_RNA >3500 &
                   nCount_RNA <50000 &
                   percent.mt <15)
cx.raw$src <- "Culture"

cx.raw <- NormalizeData(cx.raw) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = cx.raw@var.genes,
         npcs = 20) %>%
  RunUMAP(reduction = "pca",
          dims = 1:20) %>%
  FindNeighbors(reduction = "pca",
                dims = 1:20) %>%
  FindClusters(resolution = 0.3)

DimPlot(cx.raw,
        reduction = "umap",
        pt.size = 1,
        label = T) +
  coord_fixed() +
  scale_color_npg(alpha = 0.7)

# Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

cx.raw <- CellCycleScoring(object = cx.raw,
                           s.features = s.genes,
                           g2m.features = g2m.genes,
                           set.ident = TRUE)

cx.raw <- ScaleData(object = cx.raw,
                    vars.to.regress = c("S.Score",
                                        "G2M.Score"),
                    features = rownames(x = cx.raw))

cx.raw <- RunPCA(cx.raw,
                 pc.genes = cx.raw@var.genes,
                 npcs = 20) %>%
  RunUMAP(reduction = "pca",
          dims = 1:20) %>%
  FindNeighbors(reduction = "pca",
                dims = 1:20) %>%
  FindClusters(resolution = 0.3)

# Find optimal clustering resolution
cx.raw.nC <- nRes(cx.raw, 
                  res = seq(from = 0.1,
                            to = 0.2,
                            by = 0.05))

plot <- pSil(cx.raw.nC,
             0.15)
plot

# With res = 0.15, cells in cluster 3 do not have any significantly different genes that are deferentially regulated
# Therefore, we decided to proceed with res = 0.1
cx.raw <- FindClusters(cx.raw,
                       resolution = 0.1)
```

```{r saveOS17culture}
save(cx.raw, file ="OS17culture_cx.raw_CCR_v1.RData")
```

```{r Fig1a_OS17}
# Culture plot
pdf("Figure 1/OS17.Culture.pdf", width = 5, height = 5)

pcx <- DimPlot(cx.raw,
               reduction = "umap",
               pt.size = 1,
               label = T) +
  coord_fixed() +
  ggtitle("OS17 in Culture") +
  NoAxes() 
pcx + scale_color_npg(alpha = 1)

dev.off()

# View Culture plot in Rmd
pcx + scale_color_npg(alpha = 1)


# Cell cycle distribution of clusters
cid <- sort(unique(cx.raw@active.ident))

vplayout <- function(x, y) viewport(layout.pos.row = x,
                                    layout.pos.col = y)

pdf("Figure 1/OS17_pie.pdf", width = 15, height = 15)
grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 4)))

for(i in 1:3){
  temp <- WhichCells(cx.raw,
                     idents = cid[[i]])
  LT.cells <- subset(cx.raw,
                     cells = temp)
  df <- table(LT.cells$Phase)
  bp <- ggplot(as.data.frame(df),
               aes(x = "",
                   y = Freq,
                   fill = Var1)) +
    geom_bar(width = 1,
             stat = "identity")
  pie <- bp +
    coord_polar("y",
                start = 0) +
    scale_fill_brewer(palette ="Blues") +
    theme_minimal() +
    NoAxes() +
    ggtitle(paste("Cluster",
                  cid[[i]]))
  print(pie,
        vp = vplayout(ceiling(i/4),
                      i))
}

dev.off()

#Total number of cells
# table(Idents(cx.raw)) %>%
# sum()
# [1] 3178
```

## Figure 1B

*DONE*

B) UMAP analysis of NCH-OS-7 cells (n = 1998) grown as subcutaneous flank tumor. Cell cycle distribution of cells is visualized as pie charts.

```{r Fig1b_NCHOS7_ccr}
# From CulturePDXHetero.R
# NCHOS7 (flank)
# Create Seurat objects and perform initial QC. Label original source.

S0043 <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0043/outs/filtered_feature_bc_matrix/",
                    spec = "mixHuman")
S0043 <- subset(S0043,
                subset = nCount_RNA <50000 &
                  percent.mt <25)
S0043$src <- "Flank"

VlnPlot(S0043,
        features = c("nFeature_RNA",
                     "nCount_RNA",
                     "percent.mt"),
        ncol = 3)

S0043 <- NormalizeData(S0043) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = S0043@var.genes,
         npcs = 20) %>%
  RunUMAP(reduction = "pca",
          dims = 1:20) %>%
  FindNeighbors(reduction = "pca",
                dims = 1:20) %>%
  FindClusters(resolution = 0.3)

DimPlot(S0043,
        reduction = "umap",
        pt.size = 1,
        label = T) +
  coord_fixed() +
  scale_color_npg(alpha = 0.7)

# Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

S0043.CCR <- CellCycleScoring(object = S0043,
                              s.features = s.genes,
                              g2m.features = g2m.genes,
                              set.ident = TRUE)

S0043.CCR <- ScaleData(object = S0043.CCR,
                       vars.to.regress = c("S.Score",
                                           "G2M.Score"),
                       features = rownames(x = S0043.CCR))

S0043.CCR <- RunPCA(S0043.CCR,
                    pc.genes = S0043.CCR@var.genes,
                    npcs = 20) %>%
  RunUMAP(reduction = "pca",
          dims = 1:20) %>%
  FindNeighbors(reduction = "pca",
                dims = 1:20) %>%
  FindClusters(resolution = 0.3)

#Find optimal clustering resolution
S0043.CCR.nC <- nRes(S0043.CCR, 
                     res = seq(from = 0.1,
                               to = 0.15,
                               by = 0.01))

plot <- pSil(S0043.CCR.nC,
             0.15)
plot

S0043.CCR <- FindClusters(S0043.CCR,
                          resolution = 0.15)

# save(S0043.CCR, file ="NCHOS7-S0043_CCR_v1.RData")
```

```{r saveNCHOS7flank}
save(S0043.CCR, file ="NCHOS7-S0043_CCR_v1.RData")
```

```{r Fig1b_NCHOS7}
# Culture plots
pdf("NCHOS7.Flank.pdf", width = 5, height = 5)

pcx <- DimPlot(S0043.CCR,
               reduction = "umap",
               pt.size = 1,
               label = T) + 
  coord_fixed() + 
  ggtitle("NCHOS7 in Flank") +
  NoAxes()

pcx +  scale_color_npg(alpha = 1)

dev.off()

# View in Rmd
pcx +  scale_color_npg(alpha = 1)

#cell cycle distribution of clusters

cid <- sort(unique(S0043.CCR@active.ident))

vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)

pdf("NCHOS7_pie.pdf", width = 15, height = 15)
grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 4)))

for(i in 1:4){
  temp <- WhichCells(S0043.CCR,
                     idents = cid[[i]]) 
  LT.cells <- subset(S0043.CCR,
                     cells = temp)
  df <- table(LT.cells$Phase)
  bp <- ggplot(as.data.frame(df),
               aes(x="",
                   y=Freq,
                   fill=Var1)) +
    geom_bar(width = 1,
             stat = "identity")
  pie <- bp +
    coord_polar("y",
                start=0) +
    scale_fill_brewer(palette="Blues") +
    theme_minimal() +
    NoAxes() +
    NoLegend() +
    RestoreLegend(position = "right")
  ggtitle(paste("Cluster",
                cid[[i]]))
  print(pie,
        vp=vplayout(ceiling(i/4),
                    i))
}

dev.off()

#Total number of cells
# table(Idents(S0043.CCR)) %>%
# sum()
# [1] 1998
```

## Figure 1C

C) Pathway enrichment analysis for hallmark gene sets associated with genes upregulated in distinct clusters identified in the two osteosarcoma models. P values were adjusted for multiple comparisons. Boxes in grey identify non-significant enrichments, whereas boxes in red identify statistically significant enrichments. MTORC1: mechanistic target of rapamycin (mTOR) complex 1. PI3K: Phosphoinositide 3-kinase. AKT: Protein kinase B. TGF: Transforming growth-factor. TNF: tumor necrosis factor. NFKB: Nuclear Factor kappa-light-chain-enhancer of activated B cells.

*Need correct input for this figure heatmap.*

```{r Fig1c_dgea}
# Pathway enrichment analysis
# Display adjust p-values
B.list <- list(OS17 = cx.raw,
               NCHOS7 = S0043.CCR)

em.hm.list <- list()

for (i in 1:length(B.list)) {
  em.hm.list[[i]] <- DGEA(B.list[[i]])
}

for(i in 1:(length(em.hm.list))) {
  em.hm.list[[i]] <- setDT(em.hm.list[[i]],
                           keep.rownames = TRUE)[]
}

temp1 <- em.hm.list[[1]]
temp2 <- em.hm.list[[2]]

cx.pd <- full_join(temp1,
                   temp2,
                   by = "rn")

# Write table to save for supplemental data and for generating Heatmap
# write.table(cx.pd, file=('Fig1.tsv'), quote=FALSE, sep='\t')
# save.image(file ="CulturePDXHetero.RData")

cx.pd <- column_to_rownames(cx.pd,
                            var = "rn")
cx.pd[is.na(cx.pd)] <- 1

#remove "HALLMARK_"
c <- rownames(cx.pd)
c <- gsub("HALLMARK_",
          "",
          c)

#remove "_" by removing special characters
c <- gsub("_",
          " ",
          c)
rownames(cx.pd) <- c

# Take the -log10
cx.pd.log <- -log10(cx.pd)

colnames(cx.pd.log) <- c("A0",
                         "A1",
                         "A2",
                         "B0",
                         "B1",
                         "B2",
                         "B3")

# Need to find out where aka4 comes from

# To prevent values that show up as -0.00, we take the absolute values
cx.pd.log <- abs(cx.pd.log)

cx.pd.log %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  left_join(aka4 %>% rownames_to_column(var = "Sample")) %>%
  pivot_longer(c(-Sample,
                 -Lpercent),
               names_to = "Pathway",
               values_to = "pval",
               names_repair = "minimal") %>%
  mutate(Signif = pval >= (-1 * log10(0.05))) %>%
  ggplot(.,
         aes(x = Sample,
             y = Pathway,
             fill = Signif)) +
  geom_tile() +
  geom_text(aes(label = sprintf("%0.2f",
                                pval))) + 
  scale_fill_manual(values = c("gray",
                               "red")) +
  scale_y_discrete(limits = rev) +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks.x = element_blank()) +
  ylab("")

```

```{r Fig1_heatmaptake2}
#ComplexHeatmap
# Fig1.txv from above code (previously called cxpd.tsv in Sanjana's files)
cx.pd <- read.delim("Fig1.tsv", row.names = "rn")

#remove "HALLMARK_"
c <- rownames(cx.pd)
c <- gsub("HALLMARK_", "", c)
#remove "_" by removing special characters
c <- gsub("_", " ", c)
rownames(cx.pd) <- c
cx.pd.log <- -log10(cx.pd[,-1]) #log transform
  
cx.pd_transpose <- as.data.frame(t(cx.pd.log))
df <- as.matrix(cx.pd_transpose)
subj <- c(paste0("A", 0:3),
          paste0("B", 0:3),
          paste0("C", 0:2),
          paste0("D", 0:3))

#rownames(df) <- subj
aka2 = data.frame(ID = factor(c(rep("OS17", 4),
                                rep("t143B", 4),
                                rep("OS2", 3),
                                rep("OS7", 4)
)))
rownames(aka2)<-subj
aka3 = list(ID = c(OS17 = "#E64B35FF",
                   t143B = "#4DBBD5FF",
                   OS2 = "#00A087FF",
                   OS7 = "#3C5488FF"))

#Include percentage of cells in lung for each cluster
aka4 <- data.frame (Lpercent = c(46.577747,
                                 10.17901,
                                 39.69814,
                                 3.545104,
                                 2.2204908,
                                 92.306194,
                                 3.3502143,
                                 2.1231009,
                                 20.32767,
                                 71.480583,
                                 8.191748,
                                 71.01349,
                                 1.383604,
                                 1.176064,
                                 26.426842))
rownames(aka4) <- subj

aka5 = (c(rep("OS17", 4),
          rep("t143B", 4),
          rep("OS2", 3),
          rep("OS7", 4)))

# Plot heatmap
Fig1 <- read_delim("Fig1.txt", 
                   delim = "\t", 
                   col_names = TRUE)

colnames(Fig1) <- c("Pathway",
                    "A0",
                    "A1",
                    "A2",
                    "A3",
                    "B0",
                    "B1",
                    "B2",
                    "B3")

Fig1 %>%
  column_to_rownames(var = "Pathway") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  left_join(aka4 %>% rownames_to_column(var = "Sample")) %>%
  pivot_longer(c(-Sample, -Lpercent),
               names_to = "Pathway",
               values_to = "pval",
               names_repair = "minimal") %>%
  mutate(Signif = pval >= (-1 * log10(0.05))) %>%
  ggplot(., aes(x = Sample, y = Pathway, fill = Signif)) + 
  geom_tile() +
  geom_text(aes(label = sprintf("%0.2f", pval))) +
  scale_fill_manual(values = c("gray", "red")) +
  scale_y_discrete(limits = rev) +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks.x = element_blank()) +
  ylab("") 

```

## Figure 1 Supplemental Cluster Heatmaps {.tabset}

### OS17 DT

```{r Fig1_os17c_supplheat_dt}
OS17markers <- FindAllMarkers(cx.raw,
                          min.pct = 0.25,
                          logfc.threshold = 0.25)

# Print top 10 in each cluster and organize for best viewing
OS17markers %>%
  group_by(cluster) %>%
  top_n(n = 10, wt = avg_log2FC) -> t

t <- as.data.frame(t)
t$cluster <- as.numeric(as.character(t$cluster))
t <- t[order(t$cluster),]
t <- t[,c(7, 1:6)]
t <- t[,c(7, 1:6)]
#could also use arrange() to sort by column

DT::datatable(t, 
              rownames = FALSE,
              extensions = c('FixedColumns',
                             'Buttons'),
              options = list(
                pageLength = 10,
                scrollX = TRUE,
                scrollCollapse = TRUE,
                dom = 'Bfrtip',
                buttons = c('copy',
                            'csv',
                            'excel')
              ))
```

### OS17 Heatmap

```{r Fig1_os17c_supplheat}
# Visualize colors that match those in the dimplot for group.colors
#library("scales")
#show_col(pal_npg("nrc")(10))

DoHeatmap(cx.raw,
          features = t$gene,
          group.by = 'ident',
          group.bar = T,
          label = T,
          group.bar.height = 0.08,
          draw.lines = T,
          lines.width = 15,
          group.colors = c("#E64B35FF",
                           "#4DBBD5FF",
                           "#00A087FF",
                           "#3C5488FF")) +
  scale_fill_gradientn(colors = c("blue",
                                  "light grey",
                                  "red"))
```

### NCH-OS-7 DT

```{r Fig1_os7f_supplheat_dt}
NCHOS7markers <- FindAllMarkers(S0043.CCR,
                                 min.pct = 0.25,
                                 logfc.threshold = 0.25)

# Print top 10 in each cluster and organize for best viewing
NCHOS7markers %>%
  group_by(cluster) %>%
  top_n(n = 10, wt = avg_log2FC) -> t

t <- as.data.frame(t)

t <- as.data.frame(t)
t$cluster <- as.numeric(as.character(t$cluster))
t <- t[order(t$cluster),]
t <- t[,c(7, 1:6)]
t <- t[,c(7, 1:6)]
#could also use arrange() to sort by column

DT::datatable(t, 
              rownames = FALSE,
              extensions = c('FixedColumns',
                             'Buttons'),
              options = list(
                pageLength = 10,
                scrollX = TRUE,
                scrollCollapse = TRUE,
                dom = 'Bfrtip',
                buttons = c('copy',
                            'csv',
                            'excel')
              ))
```

### NCH-OS-7 Heatmap

```{r Fig1_os7f_supplheat}
# Visualize colors that match those in the dimplot for group.colors
#library("scales")
#show_col(pal_npg("nrc")(10))

DoHeatmap(S0043.CCR,
          features = t$gene,
          group.by = 'ident',
          group.bar = T,
          label = T,
          group.bar.height = 0.08,
          draw.lines = T,
          lines.width = 15,
          group.colors = c("#E64B35FF",
                           "#4DBBD5FF",
                           "#00A087FF",
                           "#3C5488FF")) +
  scale_fill_gradientn(colors = c("blue",
                                  "light grey",
                                  "red")) 
```

# Figure 2

Figure 2. Osteosarcoma cells adopt distinct transcriptional profiles as they colonize tibia and lung microenvironments. A) Schematic of study design depicting generation of orthotopic primary and metastatic tumors. Tumors were harvested for scRNA-seq when mice reached endpoint.

### Load Seurat objects 

Load Seurat objects here so only have to complete once (commented out elsewhere). 

```{r loadallobj}
setwd("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021")
source("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/scSeurat.R")
source("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/Downstream.v2.R")

# Create Seurat objects and perform initial QC.  Label original source.
OB <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0031/outs/filtered_feature_bc_matrix/",
                 spec = "human")
OB <- subset(OB, subset = nCount_RNA <20000 & percent.mt <10)
OB$src <- "OB"
OB$cond <- "OB"

#OS17
os17.cx.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0016/outs/filtered_feature_bc_matrix/",
                          spec = "mixHuman")
os17.cx.raw <- subset(os17.cx.raw, subset = nFeature_RNA >3500 & nCount_RNA <50000 & percent.mt <15)
os17.cx.raw$src <- "OS17_Culture"
os17.cx.raw$cond <- "OS17"

os17.tib.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/filtered_feature_bc_matrix/",
                           spec = "mixHuman")
os17.tib.raw <- subset(os17.tib.raw, subset = nFeature_RNA >3000 & nCount_RNA <60000 & percent.mt <18)
os17.tib.raw$src <- "OS17_Tibia"
os17.tib.raw$cond <- "OS17"

## double check usage throughout rest of code - Emily (fixed issue where coming from wrong file)
os17.lung.raw <- tenXLoadQC(path10x = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/filtered_feature_bc_matrix",
                            spec = "mixHuman")
os17.lung.raw <- subset(os17.lung.raw, subset = nFeature_RNA >1250 & nCount_RNA <60000 & percent.mt <25)
os17.lung.raw$src <- "OS17_Lung"
os17.lung.raw$cond <- "OS17"

#t143b
t143b.cx.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0017/outs/filtered_feature_bc_matrix/",
                           spec = "mixHuman")
t143b.cx.raw <- subset(t143b.cx.raw, subset = nFeature_RNA >4000 & nCount_RNA <74000 & percent.mt <17)
VlnPlot(t143b.cx.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
t143b.cx.raw$src <- "t143b_Culture"
t143b.cx.raw$cond <- "t143b"

t143b.tib.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0052/outs/filtered_feature_bc_matrix/",
                            spec = "mixHuman")
t143b.tib.raw <- subset(t143b.tib.raw, subset = nFeature_RNA >1000 & nCount_RNA <30000 & percent.mt <22)
VlnPlot(t143b.tib.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
t143b.tib.raw$src <- "t143b_Tibia"
t143b.tib.raw$cond <- "t143b"

t143b.lung.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0019-143B-lung/outs/filtered_feature_bc_matrix/",
                             spec = "mixHuman")
t143b.lung.raw <- subset(t143b.lung.raw, subset = nFeature_RNA >1000 & nCount_RNA <30000 & percent.mt <22)
VlnPlot(t143b.lung.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
t143b.lung.raw$src <- "t143b_Lung"
t143b.lung.raw$cond <- "t143b"

#NCHOS2
OS2.cx.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0076/outs/filtered_feature_bc_matrix/",
                         spec = "mixHuman")
OS2.cx.raw <- subset(OS2.cx.raw, subset = nCount_RNA <36000 & percent.mt <50)
VlnPlot(OS2.cx.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS2.cx.raw$src <- "NCHOS2_Flank"
OS2.cx.raw$cond <- "NCHOS2"

OS2.tib.raw<- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0042/outs/filtered_feature_bc_matrix/",
                         spec = "mixHuman")
OS2.tib.raw <- subset(OS2.tib.raw, subset = nFeature_RNA >2500 & nCount_RNA <40000 & percent.mt <20)
OS2.tib.raw$src <- "NCHOS2_Tibia"
VlnPlot(OS2.tib.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS2.tib.raw$cond <- "NCHOS2"

OS2.lung.raw<- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0041/outs/filtered_feature_bc_matrix/",
                          spec = "mixHuman")
OS2.lung.raw <- subset(OS2.lung.raw, subset = nFeature_RNA >2500 & nCount_RNA <60000 & percent.mt <30)
OS2.lung.raw$src <- "NCHOS2_Lung"
VlnPlot(OS2.lung.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS2.lung.raw$cond <- "NCHOS2"

#NCHOS7
OS7.cx.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0043/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
OS7.cx.raw <- subset(OS7.cx.raw, subset = nCount_RNA <50000 & percent.mt <25)
OS7.cx.raw$src <- "NCHOS7_Flank"
VlnPlot(OS7.cx.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS7.cx.raw$cond <- "NCHOS7"

OS7.tib.raw <- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0034/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
OS7.tib.raw <- subset(OS7.tib.raw, subset = nFeature_RNA >2000 & nCount_RNA <35000 & percent.mt <14)
OS7.tib.raw$src <- "NCHOS7_Tibia"
VlnPlot(OS7.tib.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS7.tib.raw$cond <- "NCHOS7"

OS7.lung.raw<- tenXLoadQC(path10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0055/outs/filtered_feature_bc_matrix/", spec = "mixHuman")
OS7.lung.raw <- subset(OS7.lung.raw, subset = nCount_RNA <42000 & percent.mt <20)
OS7.lung.raw$src <- "NCHOS7_Lung"
VlnPlot(OS7.lung.raw, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
OS7.lung.raw$cond <- "NCHOS7"
```

```{r mergeallobj, dependson='loadallobj'}
# Merge into a single Seurat object
OS <- merge(OB, y = c(os17.cx.raw,
                      os17.tib.raw,
                      os17.lung.raw,
                      t143b.cx.raw,
                      t143b.tib.raw,
                      t143b.lung.raw,
                      OS2.cx.raw,
                      OS2.tib.raw,
                      OS2.lung.raw,
                      OS7.cx.raw,
                      OS7.tib.raw,
                      OS7.lung.raw),
            add.cell.ids = c("OB",
                             "OS17_Culture",
                             "OS17_Tibia",
                             "OS17_Lung",
                             "t143b_Culture",
                             "t143b_Tibia",
                             "t143b_Lung",
                             "NCHOS2_Flank",
                             "NCHOS2_Tibia",
                             "NCHOS2_Lung",
                             "NCHOS7_Flank",
                             "NCHOS7_Tibia",
                             "NCHOS7_Lung"),
            project = "Heterogeneity")

# Process and cluster
OS <- NormalizeData(OS) %>%
    FindVariableFeatures(selection.method = "vst") %>%
    ScaleData() %>%
    RunPCA(pc.genes = os.17@var.genes, npcs = 20) %>%
    RunUMAP(reduction = "pca", dims = 1:20) %>%
    FindNeighbors(reduction = "pca", dims = 1:20) %>%
    FindClusters(resolution = 0.3)

# rm(OB, os17.cx.raw, os17.tib.raw, os17.lung.raw, t143b.cx.raw, t143b.tib.raw,
# t143b.lung.raw, OS2.cx.raw, OS2.tib.raw, OS2.lung.raw, OS7.cx.raw, OS7.tib.raw,
# OS7.lung.raw)

# CCR Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
OS <- CellCycleScoring(object = OS,
                       s.features = s.genes,
                       g2m.features = g2m.genes,
                       set.ident = TRUE)

OS <- ScaleData(object = OS,
                vars.to.regress = c("S.Score",
                                    "G2M.Score"),
                features = rownames(x = OS))

OS <- RunPCA(OS, pc.genes = OS@var.genes, npcs = 20) %>%
             RunUMAP(reduction = "pca", dims = 1:20) %>%
             FindNeighbors(reduction = "pca", dims = 1:20) %>%
             FindClusters(resolution = 0.3)

save(OS, file = "OS.combined.OBpostCCR.RData")

OS$cond <- as.factor(OS$cond)
OS$cond <- factor(as.factor(OS$cond),
                  levels = c("OS17",
                             "t143b",
                             "NCHOS2",
                             "NCHOS7",
                             "OB"))

# Plot the data
DimPlot(OS,
        reduction = "umap",
        group.by = "cond",
        pt.size = 1,
        label = T) +
  coord_fixed() +
  ggtitle("OS by Condition") +
    scale_color_npg(alpha = 1)

DimPlot(OS,
        reduction = "umap",
        group.by = "src",
        pt.size = 1,
        label = T) +
  coord_fixed() +
  ggtitle("OS by Source")

DimPlot(OS,
        reduction = "umap",
        pt.size = 1,
        label = T) +
  coord_fixed() +
  ggtitle("OS by Clusters") +
    scale_color_npg(alpha = 0.7)
```

## Figure 2B

B) Venn diagrams showing overlap of differentially expressed genes that are up- or down- regulated in primary or metastatic tumors relative to corresponding starting population of cells (cell line or PDX flank tumors). Region outlined in red identifies differentially expressed genes shared across at least three models.

```{r Fig2b, dependson='mergeallobj'}

Idents(OS) <- OS$src

OS <- RenameIdents(OS,
                   OS17_Culture = "Culture",
                   OS17_Lung = "Lung",
                   OS17_Tibia = "Tibia",
                   t143b_Culture = "Culture",
                   t143b_Lung = "Lung",
                   t143b_Tibia = "Tibia",
                   NCHOS2_Flank = "Culture",
                   NCHOS2_Lung = "Lung",
                   NCHOS2_Tibia = "Tibia",
                   NCHOS7_Flank = "Culture",
                   NCHOS7_Lung = "Lung",
                   NCHOS7_Tibia = "Tibia")

OS$tissue <- Idents(OS)

os17 <- subset(OS, cells = WhichCells(OS, expression = cond == "OS17"))
t143B <- subset(OS, cells = WhichCells(OS, expression = cond == "t143b"))
NCHOS2 <- subset(OS, cells = WhichCells(OS, expression = cond == "NCHOS2"))
NCHOS7 <- subset(OS, cells = WhichCells(OS, expression = cond == "NCHOS7"))

# Create list of objects to use for Find markers analysis
data <- list(os17 = os17,
             t143B = t143B,
             NCHOS2 = NCHOS2,
             NCHOS7 = NCHOS7)

# Lung marker - UP
Lung.up <- list(NULL)
for (i in 1:length(data)) {
  tmp <- data[[i]]
  x <- FindMarkers(tmp,
                   ident.1 = "Lung",
                   ident.2 = "Culture",
                   only.pos = TRUE,
                   min.pct = 0.1)
  Lung.up[[i]] <- rownames(x)
}

# Tibia markers - UP
Tibia.up <- list(NULL)
for (i in 1:length(data)) {
  tmp <- data[[i]]
  x <- FindMarkers(tmp,
                   ident.1 = "Tibia",
                   ident.2 = "Culture",
                   only.pos = TRUE,
                   min.pct = 0.1)
  Tibia.up[[i]] <- rownames(x)
}

# Lung markers - DOWN
Lung.down <- list(NULL)
for (i in 1:length(data)) {
  tmp <- data[[i]]
  x <- FindMarkers(tmp,
                   ident.1 = "Lung",
                   ident.2 = "Culture",
                   only.pos = FALSE,
                   min.pct = 0.1)
  x <- x[x$avg_log2FC < 0, ]
  Lung.down[[i]] <- rownames(x)
}

# Tibia markers - DOWN
Tibia.down <- list(NULL)
for (i in 1:length(data)) {
  tmp <- data[[i]]
  x <- FindMarkers(tmp,
                   ident.1 = "Tibia",
                   ident.2 = "Culture",
                   only.pos = FALSE,
                   min.pct = 0.1)
  x <- x[x$avg_log2FC < 0, ]
  Tibia.down[[i]] <- rownames(x)
}

# Create Venn Diagram for Figure 2
library(ggvenn)
genes <- Lung.down
x <- list(A = genes[[1]],
          B = genes[[2]],
          C = genes[[3]],
          D = genes[[4]])

# PDF of Venn Diagram
pdf("Lung.down.pdf", width = 7, height = 7)
plot <- ggvenn(x,
               fill_color = c("#0073C2FF",
                              "#EFC000FF",
                              "#868686FF",
                              "#CD534CFF"),
               stroke_size = 0.9,
               set_name_size = 10,
               show_percentage = TRUE)
plot
dev.off()
```

```{r Fig2b_all, dependson="Fig2b", eval=F}
### Commented out in previous code, but possible combination plot of all Venn diagrams

# ggVennDiagram(x[1:4], label_alpha = 0.7, label = 'count', category.names = c('OS17', '143B',
# 'NCHOS2', 'NCHOS7')) + scale_fill_gradient(low = '#F4FAFE', high = '#4981BF')+
# theme(legend.title = element_text(color = 'black'), legend.position = 'right')

### Extract genes that are shared in these datasets
genes <- Lung.down
A = genes[[1]]
B = genes[[2]]
C = genes[[3]]
D = genes[[4]]

# Preparing clusterProfiler to perform hypergeometric test on msigdb signatures m_t2g.c2 <-
# msigdbr(species = 'Homo sapiens', category = 'C2') %>% dplyr::select(gs_name,
# human_gene_symbol) m_t2g.c6 <- msigdbr(species = 'Homo sapiens', category = 'C6') %>%
# dplyr::select(gs_name, human_gene_symbol)
m_t2g.h <- msigdbr(species = "Homo sapiens", category = "H") %>%
    dplyr::select(gs_name, human_gene_symbol)
m_t2n.h <- msigdbr(species = "Homo sapiens", category = "H") %>%
    dplyr::select(gs_id, gs_name)
# m_t2g=rbind(m_t2g.c2,m_t2g.c6)

# msigdb signature to use
msig.gene.set = m_t2g.h
msig.name = m_t2n.h

intersect <- c((intersect(intersect(intersect(A, B), C), D)), (intersect(intersect(A, B), C)), (intersect(intersect(A,
    B), D)), (intersect(intersect(A, D), C)), (intersect(intersect(B, D), C)))

tmp <- enricher(intersect, TERM2GENE = msig.gene.set, TERM2NAME = msig.name)
em = tmp@result[, c("ID", "p.adjust")]  #pvalue/p.adjust
rownames(em) <- NULL
em <- em[1:10, ]
em[, 2] <- -log10(em[, 2])
# barplot(em)

pathways <- list(NULL)
pathways[[4]] <- em

# write.xlsx(pathways, file
# ='R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/pathways.xlsx', col.names = TRUE,
# row.names = TRUE, append = FALSE) save.image(file
# ='R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/AllpseudoBulk_Venn.RData')
# write.xlsx(pathways, file ='pathways.xlsx', col.names = TRUE, row.names = TRUE, append =
# FALSE) save.image(file ='AllpseudoBulk_Venn.RData')

#################### Intersection of deferentially regulated genes between Bone colonization
#################### and Lung colonization total <- list( Tibia.up = Tibia.up, Tibia.down =
#################### Tibia.down, Lung.up = Lung.up, Lung.down = Lung.down ) all.genes <-
#################### list(NULL) for (i in 1:4) { genes <- total[[i]] A = genes[[1]] B =
#################### genes[[2]] C = genes[[3]] D = genes[[4]] intersect <-
#################### intersect(intersect(intersect(A,B),C),D) all.genes[[i]] <- intersect } x
#################### <- list( A = all.genes[[1]], B = all.genes[[2]], C = all.genes[[3]], D =
#################### all.genes[[4]] ) pdf('TibiaUpdown_Lungupdown.pdf', width = 5, height = 5)
#################### plot <- ggvenn( x, fill_color = c('#0073C2FF', '#EFC000FF', '#868686FF',
#################### '#CD534CFF'), stroke_size = 0.9, set_name_size = 4, show_percentage =
#################### TRUE ) plot dev.off()
```

C) Pathway enrichment analysis with adjusted p values for hallmark gene sets associated with these shared genes

```{r Fig2c, dependson = Fig2b, fig.height=6, fig.width=10}
### Extract genes that are shared in these datasets
genes <- Lung.down
A = genes[[1]]
B = genes[[2]]
C = genes[[3]]
D = genes[[4]]

# Preparing clusterProfiler to perform hypergeometric test on msigdb signatures m_t2g.c2 <-
# msigdbr(species = 'Homo sapiens', category = 'C2') %>% dplyr::select(gs_name,
# human_gene_symbol) m_t2g.c6 <- msigdbr(species = 'Homo sapiens', category = 'C6') %>%
# dplyr::select(gs_name, human_gene_symbol)
m_t2g.h <- msigdbr(species = "Homo sapiens", category = "H") %>%
    dplyr::select(gs_name, human_gene_symbol)
m_t2n.h <- msigdbr(species = "Homo sapiens", category = "H") %>%
    dplyr::select(gs_id, gs_name)
# m_t2g=rbind(m_t2g.c2,m_t2g.c6)

# msigdb signature to use
msig.gene.set = m_t2g.h
msig.name = m_t2n.h

intersect <- c((intersect(intersect(intersect(A, B), C), D)),
               (intersect(intersect(A, B), C)),
               (intersect(intersect(A, B), D)),
               (intersect(intersect(A, D), C)),
               (intersect(intersect(B, D), C)))

tmp <- enricher(intersect,
                TERM2GENE = msig.gene.set,
                TERM2NAME = msig.name)

#pvalue/p.adjust
em = tmp@result[, c("ID", "p.adjust")]
rownames(em) <- NULL
em <- em[1:10, ]
em[, 2] <- -log10(em[, 2])
# barplot(em)

pathways <- list(NULL)
pathways[[4]] <- em

# write.xlsx(pathways, file
# ='R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/pathways.xlsx', col.names = TRUE,
# row.names = TRUE, append = FALSE) save.image(file
# ='R:/RESRoberts/Bioinformatics/Analysis/Sanjana/2020/AllpseudoBulk_Venn.RData')
# write.xlsx(pathways, file ='pathways.xlsx', col.names = TRUE, row.names = TRUE, append =
# FALSE) save.image(file ='AllpseudoBulk_Venn.RData')

################ Two sided Barplot
SanjanaPlots <- read.delim("SanjanaPlots.txt")

ggplot(SanjanaPlots, aes(x = -1 * Order,
                         y = p.adjust,
                         fill = p.adjust > 0)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  facet_wrap(~Tissue) +
  scale_fill_brewer(type = "qual",
                    palette = "Set1") +
  geom_text(aes(y = Label_y *
                  3, label = Pathway)) +
  theme_bw() +
  ylab("") +
  ylim(-5, 5)
```


# Figure 3

Figure 3. Osteosarcoma cells retain phenotypic heterogeneity despite adaptive changes in response to changing microenvironments. A) Schematic outlining scRNA-seq bioinformatics workflow.

## Figure 3B

B) Osteosarcoma cells exhibited higher ITH scores than that of osteoblasts grown in cell culture (####p <0.001). Within model comparison, identified increase in ITH scores as cells colonize tibia and lung microenvironments, with the exception of NCH-OS-7 (****p <0.001). P values were adjusted using Šidák multiple comparisons test.

```{r Fig3b}

#Create List to compute ITH scores
OS.list <- list(OB = OB,
                OS17.cx = os17.cx.raw,
                OS17.Tibia = os17.tib.raw, 
                OS17.Lung = os17.lung.raw, 
                t143B.cx = t143b.cx.raw,
                t143B.Tibia = t143b.tib.raw, 
                t143B.Lung = t143b.lung.raw,
                OS2.Flank = OS2.cx.raw,
                OS2.Tibia = OS2.tib.raw, 
                OS2.Lung = OS2.lung.raw, 
                OS7.Flank= OS7.cx.raw,
                OS7.Tibia = OS7.tib.raw, 
                OS7.Lung = OS7.lung.raw) 

#load(file = "/gpfs0/home/gdrobertslab/rssxr002/Analysis/2020-OS/OS.list.RData")
for (i in 1: length(OS.list)) {
  OS.list[[i]] <- NormalizeData(OS.list[[i]]) %>%
    FindVariableFeatures(selection.method = "vst") %>%
    ScaleData() %>%
    RunPCA(pc.genes = OS.list[[i]]@var.genes, npcs = 20) %>%
    RunUMAP(reduction = "pca", dims = 1:20) %>%
    FindNeighbors(reduction = "pca", dims = 1:20) %>%
    FindClusters(resolution = 0.3) %>%
    subset(cells = sample(Cells(OS.list[[i]]), 1500))
}

# CCR
# Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

for (i in 1: length(OS.list)) {
  OS.list[[i]] <- CellCycleScoring(object = OS.list[[i]], s.features = s.genes,
                                   g2m.features = g2m.genes, set.ident = TRUE)
  OS.list[[i]] <- ScaleData(object = OS.list[[i]],
                            vars.to.regress = c("S.Score", "G2M.Score"),
                            features = rownames(x = OS.list[[i]]))
  OS.list[[i]] <- RunPCA(OS.list[[i]], pc.genes = OS.list[[i]]@var.genes, npcs = 20) %>%
    RunUMAP(reduction = "pca", dims = 1:20) %>%
    FindNeighbors(reduction = "pca", dims = 1:20) %>%
    FindClusters(resolution = 0.3)
}

save(OS.list, file = "OS.listv3.CCR.RData")
```

## Figure 3C

C) UMAP analysis for merged primary and metastatic samples in each of the four models. Cells in grey represent remaining cells in merged sample. Cluster enrichment analysis shows distribution of cells in each cluster in the two microenvironment conditions (tibia, lung). While some cells in the primary and metastatic lesions adopted shared phenotypes, others adopted distinct phenotypes.

```{r Fig3c}
# Merge by condition

# OS17
# os17.tib.raw <- tenXLoadQC('R:/RESRoberts/Bioinformatics/scRNAOuts/S0018xS0028/filtered_feature_bc_matrix/',
# spec = 'mixHuman')
# os17.tib.raw <- subset(os17.tib.raw,
#                        subset = nFeature_RNA > 3000 &
#                          nCount_RNA < 60000 &
#                          percent.mt < 18)
os17.tib.raw$src <- "OS17_Tibia"
os17.tib.raw$cond <- "OS17"


# os17.lung.raw <-
# tenXLoadQC('R:/RESRoberts/Bioinformatics/scRNAOuts/S0024xS0029/filtered_feature_bc_matrix/',
# spec = 'mixHuman')
# os17.lung.raw <- subset(os17.lung.raw, subset = nFeature_RNA >1250 &
# nCount_RNA <60000 & percent.mt <25)
os17.lung.raw$src <- "OS17_Lung"
os17.lung.raw$cond <- "OS17"

# Subset number of cells in each Seurat object to the least in the group
a <- table(os17.tib.raw$orig.ident)
b <- table(os17.lung.raw$orig.ident)
list <- c(a, b)
min <- min(list)

os17.tib.raw <- subset(os17.tib.raw,
                       cells = sample(Cells(os17.tib.raw),
                                      min))
os17.lung.raw <- subset(os17.lung.raw,
                        cells = sample(Cells(os17.lung.raw),
                                       min))

# t143b
# t143b.cx.raw <- tenXLoadQC(path10x =
# 'R:/RESRoberts/Bioinformatics/scRNAOuts/S0017-143b/filtered_feature_bc_matrix/', spec =
# 'mixHuman') t143b.cx.raw <- subset(t143b.cx.raw, subset = nFeature_RNA >4000 & nCount_RNA
# <74000 & percent.mt <17) VlnPlot(t143b.cx.raw, features = c('nFeature_RNA', 'nCount_RNA',
# 'percent.mt'), ncol = 3)
t143b.cx.raw$src <- "t143b_Cx"
t143b.cx.raw$cond <- "t143b"

# t143b.tib.raw <- tenXLoadQC(path10x =
# 'R:/RESRoberts/Bioinformatics/scRNAOuts/S0052-143B-Tibia/filtered_feature_bc_matrix/', spec
# = 'mixHuman') t143b.tib.raw <- subset(t143b.tib.raw, subset = nFeature_RNA >1000 &
# nCount_RNA <30000 & percent.mt <22) VlnPlot(t143b.tib.raw, features = c('nFeature_RNA',
# 'nCount_RNA', 'percent.mt'), ncol = 3)
t143b.tib.raw$src <- "t143b_Tibia"
t143b.tib.raw$cond <- "t143b"

# t143b.lung.raw <- tenXLoadQC(path10x =
# 'R:/RESRoberts/Bioinformatics/scRNAOuts/S0019-143B-lung/filtered_feature_bc_matrix/', spec =
# 'mixHuman') t143b.lung.raw <- subset(t143b.lung.raw, subset = nFeature_RNA >1000 &
# nCount_RNA <30000 & percent.mt <22) VlnPlot(t143b.lung.raw, features = c('nFeature_RNA',
# 'nCount_RNA', 'percent.mt'), ncol = 3)
t143b.lung.raw$src <- "t143b_Lung"
t143b.lung.raw$cond <- "t143b"

# Subset number of cells in each Seurat object to the least in the group
a <- table(t143b.tib.raw$orig.ident)
b <- table(t143b.lung.raw$orig.ident)
list <- c(a, b)
min <- min(list)

t143b.tib.raw <- subset(t143b.tib.raw,
                        cells = sample(Cells(t143b.tib.raw),
                                       min))
t143b.lung.raw <- subset(t143b.lung.raw,
                         cells = sample(Cells(t143b.lung.raw),
                                        min))

# NCHOS2
#OS2.tib.raw<- tenXLoadQC(path10x =
# 'R:/RESRoberts/Bioinformatics/scRNAOuts/S0042-NCHOS2-tibia/filtered_feature_bc_matrix/',
# spec = 'mixHuman') OS2.tib.raw <- subset(OS2.tib.raw, subset = nFeature_RNA >2500 &
# nCount_RNA <40000 & percent.mt <20)
OS2.tib.raw$src <- "NCHOS2_Tibia"
OS2.tib.raw$cond <- "NCHOS2"

# OS2.lung.raw <- tenXLoadQC(path10x =
# 'R:/RESRoberts/Bioinformatics/scRNAOuts/S0041-NCHOS2-lung/filtered_feature_bc_matrix/', spec
# = 'mixHuman') OS2.lung.raw <- subset(OS2.lung.raw, subset = nFeature_RNA >2500 & nCount_RNA
# <60000 & percent.mt <30)
OS2.lung.raw$src <- "NCHOS2_Lung"
OS2.lung.raw$cond <- "NCHOS2"

# Subset number of cells in each Seurat object to the least in the group
a <- table(OS2.tib.raw$orig.ident)
b <- table(OS2.lung.raw$orig.ident)
list <- c(a, b)
min <- min(list)

OS2.tib.raw <- subset(OS2.tib.raw,
                      cells = sample(Cells(OS2.tib.raw),
                                     min))
OS2.lung.raw <- subset(OS2.lung.raw,
                       cells = sample(Cells(OS2.lung.raw),
                                      min))

# NCHOS7 
# OS7.tib.raw<- tenXLoadQC(path10x =
# 'R:/RESRoberts/Bioinformatics/scRNAOuts/S0034-NCHOS7-tib/filtered_feature_bc_matrix/', spec
# = 'mixHuman') OS7.tib.raw <- subset(OS7.tib.raw, subset = nFeature_RNA >2000 & nCount_RNA
# <35000 & percent.mt <14)
OS7.tib.raw$src <- "NCHOS7_Tibia"
OS7.tib.raw$cond <- "NCHOS7"

# OS7.lung.raw<- tenXLoadQC(path10x =
# 'R:/RESRoberts/Bioinformatics/scRNAOuts/S0055-NCHOS7-lung/filtered_feature_bc_matrix/', spec
# = 'mixHuman') OS7.lung.raw <- subset(OS7.lung.raw, subset = nCount_RNA <42000 & percent.mt
# <20)
OS7.lung.raw$src <- "NCHOS7_Lung"
OS7.lung.raw$cond <- "NCHOS7"

# Subset number of cells in each Seurat object to the least in the group
a <- table(OS7.tib.raw$orig.ident)
b <- table(OS7.lung.raw$orig.ident)
list <- c(a, b)
min <- min(list)

OS7.tib.raw <- subset(OS7.tib.raw,
                      cells = sample(Cells(OS7.tib.raw),
                                     min))
OS7.lung.raw <- subset(OS7.lung.raw,
                       cells = sample(Cells(OS7.lung.raw),
                                      min))

# Merge tibia and lung objects into a single Seurat object
OS17.TL <- merge(os17.tib.raw,
                 y = c(os17.lung.raw),
                 add.cell.ids = c("OS17_Tibia",
                                  "OS17_Lung"),
                 project = "Heterogeneity")

t143b <- merge(t143b.tib.raw,
               y = c(t143b.lung.raw),
               add.cell.ids = c("t143b_Tibia",
                                "t143b_Lung"),
               project = "Heterogeneity")

OS2 <- merge(OS2.tib.raw,
             y = c(OS2.lung.raw),
             add.cell.ids = c("OS2_Tibia",
                              "OS2_Lung"),
             project = "Heterogeneity")

OS7 <- merge(OS7.tib.raw,
             y = c(OS7.lung.raw),
             add.cell.ids = c("OS7_Tibia",
                              "OS7_Lung"),
             project = "Heterogeneity")

# Process and cluster
OS17.TL <- NormalizeData(OS17.TL) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = OS17.TL@var.genes,
         npcs = 20) %>%
  RunUMAP(reduction = "pca",
          dims = 1:20) %>%
  FindNeighbors(reduction = "pca",
                dims = 1:20) %>%
  FindClusters(resolution = 0.3)

t143b <- NormalizeData(t143b) %>%
    FindVariableFeatures(selection.method = "vst") %>%
    ScaleData() %>%
    RunPCA(pc.genes = t143b@var.genes,
           npcs = 20) %>%
    RunUMAP(reduction = "pca",
            dims = 1:20) %>%
    FindNeighbors(reduction = "pca",
                  dims = 1:20) %>%
    FindClusters(resolution = 0.3)

OS2 <- NormalizeData(OS2) %>%
    FindVariableFeatures(selection.method = "vst") %>%
    ScaleData() %>%
    RunPCA(pc.genes = OS2@var.genes,
           npcs = 20) %>%
    RunUMAP(reduction = "pca",
            dims = 1:20) %>%
    FindNeighbors(reduction = "pca",
                  dims = 1:20) %>%
    FindClusters(resolution = 0.3)

OS7 <- NormalizeData(OS7) %>%
    FindVariableFeatures(selection.method = "vst") %>%
    ScaleData() %>%
    RunPCA(pc.genes = OS7@var.genes,
           npcs = 20) %>%
    RunUMAP(reduction = "pca",
            dims = 1:20) %>%
    FindNeighbors(reduction = "pca",
                  dims = 1:20) %>%
    FindClusters(resolution = 0.3)

# CCR
# Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

#OS17
OS17.TL <- CellCycleScoring(object = OS17.TL,
                            s.features = s.genes,
                            g2m.features = g2m.genes,
                            set.ident = TRUE)

OS17.TL <- ScaleData(object = OS17.TL,
                     vars.to.regress = c("S.Score",
                                         "G2M.Score"),
                     features = rownames(x = OS17.TL))

OS17.TL <- RunPCA(OS17.TL,
                  pc.genes = OS17.TL@var.genes,
                  npcs = 20) %>%
  RunUMAP(reduction = "pca",
          dims = 1:20) %>%
  FindNeighbors(reduction = "pca",
                dims = 1:20) %>%
  FindClusters(resolution = 0.3)

#t143b
t143b <- CellCycleScoring(object = t143b,
                          s.features = s.genes,
                            g2m.features = g2m.genes,
                          set.ident = TRUE)

t143b <- ScaleData(object = t143b,
                   vars.to.regress = c("S.Score",
                                       "G2M.Score"),
                     features = rownames(x = t143b))

t143b <- RunPCA(t143b,
                pc.genes = t143b@var.genes,
                npcs = 20) %>%
  RunUMAP(reduction = "pca",
          dims = 1:20) %>%
  FindNeighbors(reduction = "pca",
                dims = 1:20) %>%
  FindClusters(resolution = 0.3)

#OS2
OS2 <- CellCycleScoring(object = OS2,
                        s.features = s.genes,
                       g2m.features = g2m.genes,
                       set.ident = TRUE)

OS2 <- ScaleData(object = OS2,
                 vars.to.regress = c("S.Score",
                                     "G2M.Score"),
                 features = rownames(x = OS2))

OS2 <- RunPCA(OS2,
              pc.genes = OS2@var.genes,
              npcs = 20) %>%
  RunUMAP(reduction = "pca",
          dims = 1:20) %>%
  FindNeighbors(reduction = "pca",
                dims = 1:20) %>%
  FindClusters(resolution = 0.3)

#OS7
OS7 <- CellCycleScoring(object = OS7,
                        s.features = s.genes,
                        g2m.features = g2m.genes,
                        set.ident = TRUE)

OS7 <- ScaleData(object = OS7,
                 vars.to.regress = c("S.Score",
                                     "G2M.Score"),
                features = rownames(x = OS7))

OS7 <- RunPCA(OS7,
              pc.genes = OS7@var.genes,
              npcs = 20) %>%
  RunUMAP(reduction = "pca",
          dims = 1:20) %>%
  FindNeighbors(reduction = "pca",
                dims = 1:20) %>%
  FindClusters(resolution = 0.3)

# Plot the data
## OS17
DimPlot(OS17.TL,
        reduction = "umap",
        group.by = "src",
        pt.size = 1,
        label = T) +
  coord_fixed() +
  ggtitle("OS17 by Source") +
  scale_color_npg(alpha = 0.7)

DimPlot(OS17.TL,
        reduction = "umap",
        pt.size = 1,
        label = T,
        split.by = "src") +
  coord_fixed() +
  ggtitle("OS17 Clusters") +
  scale_color_npg(alpha = 0.7)

## t143b
DimPlot(t143b,
        reduction = "umap",
        group.by = "src",
        pt.size = 1,
        label = T) +
  coord_fixed() +
  ggtitle("t143b by Source") +
  scale_color_npg(alpha = 0.7)

DimPlot(t143b,
        reduction = "umap",
        pt.size = 1,
        label = T,
        split.by = "src") +
  coord_fixed() +
  ggtitle("t143b Clusters") +
  scale_color_npg(alpha = 1)

## OS2
DimPlot(OS2,
        reduction = "umap",
        group.by = "src",
        pt.size = 1,
        label = T) +
  coord_fixed() +
  ggtitle("OS2 by Source") +
  scale_color_npg(alpha = 0.7)

DimPlot(OS2,
        reduction = "umap",
        pt.size = 1,
        label = T,
        split.by = "src") + 
  coord_fixed() + 
  ggtitle("OS2 Clusters") + 
  scale_color_npg(alpha = 0.7)

## OS7
DimPlot(OS7,
        reduction = "umap",
        group.by = "src",
        pt.size = 1,
        label = T) +
  coord_fixed() +
  ggtitle("OS7 by Source") +
  scale_color_npg(alpha = 0.7)

DimPlot(OS7,
        reduction = "umap",
        pt.size = 1,
        label = T,
        split.by = "src") +
  coord_fixed() +
  ggtitle("OS7 Clusters") +
  scale_color_npg(alpha = 0.7)

#Plot the cluster distribution in each sample
#OS17
table(Idents(OS17.TL),
      OS17.TL$src)

prop <- prop.table(table(Idents(OS17.TL),
                         OS17.TL$src),
                   margin = 2)*100
colSums(prop)
prop
#     Tibia       Lung
# 0   50.263250   46.577747
# 1   43.453843   10.179010
# 2   1.053001   39.698140
# 3   5.229905   3.545104

#t143b
table(Idents(t143b),
      t143b$src)
prop <- prop.table(table(Idents(t143b),
                         t143b$src),
                   margin = 2)*100
colSums(prop)
prop
#     Tibia        Lung
# 0   97.8963771    2.2204908
# 1   1.1492014   92.3061940
# 2   0.4479938   3.3502143
# 3   0.5064277   2.1231009

# OS2
table(Idents(OS2),
      OS2$src)
prop <- prop.table(table(Idents(OS2),
                         OS2$src),
                   margin = 2)*100
colSums(prop)
prop
#     Tibia       Lung
# 0   82.281553   20.327670
# 1   5.339806    71.480583
# 2   12.378641   8.191748

# OS7
table(Idents(OS7),
      OS7$src)
prop <- prop.table(table(Idents(OS7),
                         OS7$src),
                   margin = 2)*100
colSums(prop)
prop
#     Tibia       Lung
# 0    0.000000   71.013490
# 1   59.425804    1.383604
# 2   40.574196    1.176064
# 3    0.000000   26.426842

#save(list = c(cx.sub, lung.sub, tib.sub, os17), 
#           file = "TL_OS17_143B_OS2_OS7_2/OS17CTL.sub.RData")
#save(list = c(os17.tib.raw, os17.lung.raw, OS17.TL), 
#     file = "TL_OS17_143B_OS2_OS7_2/Fig3_OS17TL.RData")
#save(list = c(t143b.cx.raw, t143b.lung.raw, t143b.tib.raw, t143b), 
#     file = "TL_OS17_143B_OS2_OS7_2/Fig3_t143bTL.RData")
#save(list = c(OS2.tib.raw, OS2.lung.raw, OS2), 
#     file = "TL_OS17_143B_OS2_OS7_2/Fig3_OS2TL.RData")
#save(list = c(OS7.tib.raw, OS7.lung.raw, OS7), 
#     file = "TL_OS17_143B_OS2_OS7_2/Fig3_OS7TL.RData")

#remove these objects and then run CCR on OS and save that file separately; create OS.list once cleared and re-started
#save(list = c(OS, OS.CCR), 
#     file = "TL_OS17_143B_OS2_OS7_2/Fig3_OS7TL.RData")

#Modify data to account for optimal number of clusters using nRes and pSil functions
OS17.TL.nRes <- nRes(OS17.TL, 
                     res = seq(from = 0.1,
                               to = 0.2,
                               by = 0.05))
OS17.TL.nRes

plot <- pSil(OS17.TL.nRes,
             0.15)

OS17.TL <- FindClusters(OS17.TL,
                        resolution = 0.20)

DimPlot(OS17.TL,
        reduction = "umap",
        pt.size = 1,
        label = T,
        split.by = "src") +
  coord_fixed() +
  NoLegend() +
  NoAxes()

#Modify data to account for optimal number of clusters using nRes and pSil functions
t143b.nRes <- nRes(t143b, 
                     res = seq(from = 0.1,
                               to = 0.4,
                               by = 0.05))
t143b.nRes

plot <- pSil(t143b.nRes,
             0.15)

t143b <- FindClusters(t143b,
                      resolution = 0.15)

DimPlot(t143b,
        reduction = "umap",
        pt.size = 1,
        label = T,
        split.by = "src") +
  coord_fixed()

#Modify data to account for optimal number of clusters using nRes and pSil functions
OS2.nRes <- nRes(OS2, 
                     res = seq(from = 0.1,
                               to = 0.4,
                               by = 0.05))

plot <- pSil(OS2.nRes, 0.15)

OS2 <- FindClusters(OS2, resolution = 0.15)

DimPlot(OS2,
        reduction = "umap",
        pt.size = 1,
        label = T,
        split.by = "src") +
  coord_fixed() +
  NoLegend() +
  NoAxes()

#Modify data to account for optimal number of clusters using nRes and pSil functions
OS7.nRes <- nRes(OS7,
                 res = seq(from = 0.1,
                           to = 0.2,
                           by = 0.05))
OS7.nRes

plot <- pSil(OS7.nRes,
             0.1)

OS7 <- FindClusters(OS7,
                    resolution = 0.1)

DimPlot(OS7,
        reduction = "umap",
        pt.size = 1,
        label = T,
        split.by = "src") +
  coord_fixed() +
  NoLegend() +
  NoAxes()

library(plyr)
OS17.TL$src <- as.factor(OS17.TL$src)
OS17.TL$src <- factor(as.factor(OS17.TL$src),
                      levels = c("OS17_Tibia",
                                 "OS17_Lung"))
OS17.TL$src <- revalue(OS17.TL$src,
                       c("OS17_Tibia"="Tibia",
                         "OS17_Lung"="Lung"))

OS2$src <- as.factor(OS2$src)
OS2$src <- factor(as.factor(OS2$src),
                  levels = c("NCHOS2_Tibia",
                             "NCHOS2_Lung"))
OS2$src <- revalue(OS2$src, c("NCHOS2_Tibia"="Tibia",
                              "NCHOS2_Lung"="Lung"))

OS7$src <- as.factor(OS7$src)
OS7$src <- factor(as.factor(OS7$src),
                  levels = c("NCHOS7_Tibia",
                             "NCHOS7_Lung"))
OS7$src <- revalue(OS7$src, c("NCHOS7_Tibia"="Tibia",
                              "NCHOS7_Lung"="Lung"))

OS17.TL <- OS17.TL %>% FindClusters(resolution = 0.2)
t143b <- t143b %>% FindClusters(resolution = 0.15)
OS2 <- OS2 %>% FindClusters(resolution = 0.15)
OS7 <- OS7 %>% FindClusters(resolution = 0.2)

# Generate PDF files for the Figure 3 plots
pdf("TL_OS17_143B_OS2_OS7_2/OS17.TL_fig3.pdf", width = 5, height = 5)
OS17.TL %>% FindClusters(resolution = 0.2) %>% 
  subset(cells = sample(Cells(OS17.TL), 3000)) %>% 
  DimPlot(reduction = "umap", pt.size = 1, label = T, split.by = "src") + 
  coord_fixed() + NoLegend() + NoAxes()+
  scale_color_npg(alpha = 1)
dev.off()

pdf("TL_OS17_143B_OS2_OS7_2/t143b_fig3.pdf", width = 5, height = 5)
t143b %>% FindClusters(resolution = 0.15) %>% 
  subset(cells = sample(Cells(t143b), 3000)) %>% 
  DimPlot(reduction = "umap", pt.size = 1, label = T, split.by = "src") + 
  coord_fixed() + NoLegend() + NoAxes()+
  scale_color_npg(alpha = 1)
dev.off()

pdf("TL_OS17_143B_OS2_OS7_2/OS2_fig3.pdf", width = 5, height = 5)
OS2 %>% FindClusters(resolution = 0.15) %>% 
  subset(cells = sample(Cells(OS2), 3000)) %>% 
  DimPlot(reduction = "umap", pt.size = 1, label = T, split.by = "src") + 
  coord_fixed() + NoLegend() + NoAxes()+
  scale_color_npg(alpha = 1)
dev.off()

pdf("TL_OS17_143B_OS2_OS7_2/OS7_fig3.pdf", width = 10, height = 10)
OS7 %>% FindClusters(resolution = 0.2) %>% 
  subset(cells = sample(Cells(OS7), 3000)) %>% 
  DimPlot(reduction = "umap", pt.size = 1, label = T, split.by = "src") +
  ggtitle("OS7")+
  coord_fixed() + NoLegend() + NoAxes()+
  scale_color_npg(alpha = 1)
dev.off()
```

## Figure 3D

D) Metabolic heterogeneity in glycolysis activation. We used a pathway enrichment analysis for hallmark gene sets using genes differentially upregulated in each cluster relative to every other cluster within the same model. P values were adjusted for multiple comparisons. Boxes in grey identify non-significant pathway enrichments, whereas boxes in red identify statistically significant enrichments. Bar plot shows percentage of cells identified per cluster in lung metastases. In B-D, samples subset to equal number of cells to allow inter-model comparison (n=1500 per condition).

### Processing

```{r Fig3d_1}
library(data.table)
library(pheatmap)
#Create list of objects used in Figure 3 heatmap 
B.list <- list(OS17 = OS17.TL,
               t143b = t143b,
               OS2 = OS2,
               OS7 = OS7)

# Use DGEA function to find hallmark pathway expression
em.hm.list <- list()

for (i in 1:length(B.list)) {
  
  em.hm.list[[i]] <- DGEA(B.list[[i]])
  
}


for(i in 1:(length(em.hm.list))) {

  em.hm.list[[i]] <- setDT(em.hm.list[[i]],
                           keep.rownames = TRUE)[]
  
}

temp1 <- em.hm.list[[1]]
temp2 <- em.hm.list[[2]]
temp3 <- em.hm.list[[3]]
temp4 <- em.hm.list[[4]]

cx <- inner_join(temp1, temp2, by = "rn")
pd <- inner_join(temp3, temp4, by = "rn")
cx.pd <- inner_join(cx, pd, by = "rn")

head(cx.pd)

##Write table to edit rownames to easily convert to dataframe
write.table(cx.pd, file=('Figure 3/cxpd.tsv'), quote=FALSE, sep='\t')
cx.pd <- read.table(file = "Figure 3/cxpd.tsv", header=T)

# Make the "rn" column the rownames -Emily
cx.pd <- remove_rownames(cx.pd) %>%
        column_to_rownames(var = "rn")

#remove "HALLMARK_"
c <- rownames(cx.pd)
c <- gsub("HALLMARK_", "", c)

#remove "_" by removing special characters
c <- gsub("_", " ", c)
rownames(cx.pd) <- c

#log transform
cx.pd.log <- -log10(cx.pd)

cx.pd_transpose <- as.data.frame(t(cx.pd.log))
df <- as.matrix(cx.pd_transpose)

#Write table to print supplemental information
write.table(df, file=('Figure 3/cxpd_plot.tsv'), quote=FALSE, sep='\t')

subj<-c(paste0("A", 0:3),
        paste0("B", 0:3),
        paste0("C", 0:2),
        paste0("D", 0:3))

rownames(df)<-subj

aka2 = data.frame(ID = factor(c(rep("OS17", 4),
                                rep("t143B", 4),
                                rep("OS2", 3),
                                rep("OS7", 4)
                                )))

rownames(aka2)<-subj

aka3 = list(ID = c(OS17 = "#E64B35FF",
                   t143B = "#4DBBD5FF",
                   OS2 = "#00A087FF",
                   OS7 = "#3C5488FF"))

library(ComplexHeatmap)
color = colorRampPalette(rev(brewer.pal(n = 7, name =
                                          "RdYlBu")))(100)

#calculate percentage of cells in lung for each cluster
aka4 <- data.frame (Lpercent = c(46.577747,
                                 10.17901,
                                 39.69814,
                                 3.545104,
                                 2.2204908,
                                 92.306194,
                                 3.3502143,
                                 2.1231009,
                                 20.32767,
                                 71.480583,
                                 8.191748,
                                 71.01349,
                                 1.383604,
                                 1.176064,
                                 26.426842))
rownames(aka4) <- subj

aka5 = (c(rep("OS17", 4),
          rep("t143B", 4),
          rep("OS2", 3),
          rep("OS7", 4)))

column_ha = HeatmapAnnotation(model = aka5,
                              bar1 = anno_barplot(aka4))
#pdf("TL_OS17_143B_OS2_OS7_2/temp.pdf", height = 7, width = 8)
#Heatmap(t(df), 
#        col = col, 
#        name = "-log10(p)", 
##        cluster_columns = FALSE,
#       top_annotation = column_ha)+ geom_text(aes(label = pvalue))
#dev.off()

col.breaks=seq(-log10(1),min(max(-log10(cx.pd))+1,18),by=0.5)
col=inferno(length(col.breaks)) # library(viridis)
col=c("white",colorRampPalette(brewer.pal(n = 7, name ="Reds"))(50))

pheatmap(-log10(cx.pd),cluster_rows = TRUE,cluster_cols = TRUE,
         cellwidth = 5,cellheight = 7,treeheight_row = 0,treeheight_col=0,
         color = col,scale='none',breaks=col.breaks,fontsize = 8)
```

```{r Fig3d_2}
library(dichromat)

#IPA analysis
# Decided to proceed with A2, B1, (B2, B3), C1, D0 and D3
# Find differentially expressed features selected cluster and remaining cells

#OS17
A0.markers <- FindMarkers(OS17.TL, ident.1 = "0", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
A1.markers <- FindMarkers(OS17.TL, ident.1 = "1", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
A2.markers <- FindMarkers(OS17.TL, ident.1 = "2", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
A3.markers <- FindMarkers(OS17.TL, ident.1 = "3", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)

#remove unnecessary columns
myList <- list(A0.markers, A1.markers, A2.markers, A3.markers)
myList <- lapply(myList, function(x) { x["p_val"] <- NULL; x[, 2:3] <- NULL; x })
A0.markers <- myList[[1]]
A1.markers <- myList[[2]]
A2.markers <- myList[[3]]
A3.markers <- myList[[4]]
  
# view results
# head(A2.markers)
# FeaturePlot(object = OS17.TL, features = 'NFKBIA')
B0.markers <- FindMarkers(t143b, ident.1 = "0", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
B1.markers <- FindMarkers(t143b, ident.1 = "1", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
B2.markers <- FindMarkers(t143b, ident.1 = "2", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
B3.markers <- FindMarkers(t143b, ident.1 = "3", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)

myList <- list(B0.markers, B1.markers, B2.markers, B3.markers)
myList <- lapply(myList, function(x) { x["p_val"] <- NULL; x[, 2:3] <- NULL; x })
B0.markers <- myList[[1]]
B1.markers <- myList[[2]]
B2.markers <- myList[[3]]
B3.markers <- myList[[4]]

C0.markers <- FindMarkers(OS2, ident.1 = "0", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
C1.markers <- FindMarkers(OS2, ident.1 = "1", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
C2.markers <- FindMarkers(OS2, ident.1 = "2", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)

myList <- list(C0.markers, C1.markers, C2.markers)
myList <- lapply(myList, function(x) { x["p_val"] <- NULL; x[, 2:3] <- NULL; x })
C0.markers <- myList[[1]]
C1.markers <- myList[[2]]
C2.markers <- myList[[3]]

D0.markers <- FindMarkers(OS7, ident.1 = "0", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
D1.markers <- FindMarkers(OS7, ident.1 = "1", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
D2.markers <- FindMarkers(OS7, ident.1 = "2", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)
D3.markers <- FindMarkers(OS7, ident.1 = "3", ident.2 = NULL, only.pos = FALSE, min.pct = 0.25)

myList <- list(D0.markers, D1.markers, D2.markers, D3.markers)
myList <- lapply(myList, function(x) { x["p_val"] <- NULL; x[, 2:3] <- NULL; x })
D0.markers <- myList[[1]]
D1.markers <- myList[[2]]
D2.markers <- myList[[3]]
D3.markers <- myList[[4]]

#library(xlsx)
#write.xlsx(A0.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/A0.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(A1.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/A1.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(A2.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/A2.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(A3.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/A3.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)


#write.xlsx(B0.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/B0.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(B1.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/B1.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(B2.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/B2.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(B3.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/B3.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)

#write.xlsx(C0.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/C0.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(C1.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/C1.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(C2.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/C2.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)

#write.xlsx(D0.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/D0.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(D1.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/D1.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(D2.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/D2.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)
#write.xlsx(D3.markers, file = "TL_OS17_143B_OS2_OS7_2/markers/D3.markers.xlsx", 
#           col.names = TRUE, row.names = TRUE, append = FALSE)

#Uploaded individual datasets on IPA; ran Expression analysis with adj p value cutoff at 0.01
#Emily commented out below
#df <- read.delim("TL_OS17_143B_OS2_OS7_2/IPA_updown.txt", header = TRUE)
#rownames(df) <- df[,1]
#df[,1] <- NULL
```

```{r Fig3d_3}
# cx.pd <- cx.pd[-(23:30),] #remove rows with names NA - figure out why we have NAs!!
cx.pd.log <- -log10(cx.pd) #log transform
library(pheatmap)   

cx.pd_transpose <- as.data.frame(t(cx.pd.log))
df <- as.matrix(cx.pd_transpose)

subj<-c(paste0("A", 0:3),
        paste0("B", 0:3),
        paste0("C", 0:2),
        paste0("D", 0:3))
rownames(df)<-subj

aka2 = data.frame(ID = factor(c(rep("OS17", 4),
                                rep("t143B", 4),
                                rep("OS2", 3),
                                rep("OS7", 4)
)))
rownames(aka2)<-subj

aka3 = list(ID = c(OS17 = "#E64B35FF",
                   t143B = "#4DBBD5FF",
                   OS2 = "#00A087FF",
                   OS7 = "#3C5488FF"))

# df[] <- lapply(df, gsub, pattern='HALLMARK_', replacement='')
# df

cols <- colorRampPalette(c("snow1", "red"), # distances 3 to max(distmat) colored from green to black
                             100)

# not the heatmap code used in final version -Emily commented out
#pheatmap(t(scale(df)),
 #        color = cols, 
 #        annotation_col = aka2, 
 #        annotation_colors = aka3,
 #        annotation_legend = TRUE,
         #gaps_col =  4,
 #        show_colnames = T, show_rownames = T, cluster_rows = T, 
 #        cluster_cols = T, legend = TRUE, 
#         clustering_distance_rows = "euclidean", border_color = FALSE)

#Heatmap(t(scale(df)), 
#        col = col, 
#        name = "-log10(p)", 
#        top_annotation = column_ha)

#not used elsewhere here -Emily
#color = colorRampPalette(rev(brewer.pal(n = 7, name =
#                                          "RdYlBu")))(100)

# Apply glycolysis, Hypoxia, tnfa, emt modules and compare between clusters
data <- OS17.TL

for(i in 1:length(mod_names)) {
  data <- AddModuleScore(data, gl[i], name = names(gl[i]))
}
RidgePlot(data, features = str_c(names(gl[6]), "1"))
# VlnPlot(data, features = str_c(names(gl[3]), "1"))
DotPlot(data, features = TNFA, cluster.idents = TRUE)

#noted a clear differnece in the TNFA_VIA_NFkB module
#Identify genes that are most differntially expressed in this module

data <- OS17.TL
pdf("TL_OS17_143B_OS2_OS7_2/Dotplot_OS17.pdf", height = 30, width = 7)
DotPlot(data, features = TNFA, cluster.idents = TRUE)+coord_flip()
dev.off()

data <- t143b
pdf("TL_OS17_143B_OS2_OS7_2/Dotplot_143B.pdf", height = 30, width = 7)
DotPlot(data, features = TNFA, cluster.idents = TRUE)+coord_flip()
dev.off()

data <- OS2
pdf("TL_OS17_143B_OS2_OS7_2/Dotplot_OS2.pdf", height = 30, width = 7)
DotPlot(data, features = TNFA, cluster.idents = TRUE)+coord_flip()
dev.off()

data <- OS7
pdf("TL_OS17_143B_OS2_OS7_2/Dotplot_OS7.pdf", height = 30, width = 7)
DotPlot(data, features = TNFA, cluster.idents = TRUE)+coord_flip()
dev.off()
```

### Heatmap

```{r Fig3d_heatmap}
# Pathway enrichment analysis 

P1 <- df %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  full_join(aka4 %>%
              rownames_to_column(var = "Sample")) %>%
  pivot_longer(c(-Sample,
                 -Lpercent),
               names_to = "Pathway",
               values_to = "pval",
               names_repair = "minimal") %>%
  mutate(Signif = pval >= (-1 * log10(0.05)),
         sample_order = str_remove(Sample,
                                   "[0-9]") %>%
           rank() * 1000 - Lpercent,
         Sample = reorder(Sample,
                          sample_order)) %>%
  ggplot(.,
         aes(x = Sample,
             y = Pathway,
             fill = Signif)) +
  geom_tile() +
  geom_text(aes(label = sprintf("%0.2f",
                                pval))) +
  scale_fill_manual(values = c("gray",
                               "#CC3333")) +
  scale_y_discrete(limits = rev) +
  theme(legend.position = "none") +
  ylab("")

P2 <- aka4 %>%
  rownames_to_column(var = "Sample") %>%
  full_join(aka2 %>%
              rownames_to_column(var = "Sample")) %>%
  mutate(sample_order = str_remove(Sample,
                                   "[0-9]") %>%
           rank() * 1000 - Lpercent,
         Sample = reorder(Sample,
                          sample_order),
         ID = reorder(ID, sample_order)) %>%
  ggplot(.,
         aes(x = Sample,
             y = Lpercent,
             fill = ID)) +
  geom_bar(stat = "identity") +
  ylab("Percent\nin lung") +
  xlab("") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  theme(legend.position = "top", ) +
  labs(fill = "")

# Plot combined bar plot (percentage of cells identified per cluster in lung metastases) and heatmap of hallmark pathway expressions
pdf("Figure 3/Pvalueplot.pdf", height = 12, width = 12)
cowplot::plot_grid(P2,
                   P1,
                   ncol = 1,
                   align = "v",
                   rel_heights = c(2, 10))
dev.off()

```

# Figure 4 - Images

Figure 4. Heterogeneity in Glycolysis activation identified within metastatic osteosarcoma lesions.

Figure 4 of this paper is composed of microscopy images and so is not part of the code for this paper. 

# Figure 5

Figure 5. Lung colonization bottleneck enriches for clonal populations with a common transcriptional phenotype. A) Schematic of experimental set-up. B) UMAP analysis of lineage tagged OS-17 cells in culture, implanted into tibia, or inoculated to generate lung metastases. These lineage tags are heritable allowing us to track the transcriptional profiles of daughter cells originating from the same ancestor. Metastatic and orthotopic samples share many cells with similar phenotypes, though cultured cells are dramatically different (n = 2800 per condition) C) Frequency distribution of clones identified in each of the conditions. A high level of clonal diversity is maintained in the orthotopic tumors (78% unique clones in primary tumor compared to 86% unique clones in cell culture); though dominant clones emerge in the lung (only 33% unique clones). D) Overlay of cells sharing a common ancestor onto the dimensional reduction plots in the top four enriched clonal families. E) UMAP visualization of cells belonging to top ten enriched clonal families highlighted in red. F) Pathway enrichment analysis for hallmark gene sets comparing enriched clones relative to the remaining metastatic tumor cells identifies Hypoxia, Glycolysis, EMT and MTORC1 signaling related genes to be significantly upregulated

### Processing

```{r Fig5processing}
source("/home/gdrobertslab/rsexf001/OSHetero2021/scSeurat_updated.R")

# Note that below is all OS-17 - Emily Create Seurat objects and perform initial QC.  Label
# original source.  below loading is not correct and does not match the paper's figure -Emily
# cx.raw <-
# tenXLoadQC('/gpfs0/home2/gdrobertslab/lab/Counts/S0016xS0027/filtered_feature_bc_matrix/',
# spec = 'mixHuman') cx.raw cx.raw <- subset(cx.raw, subset = nFeature_RNA >3500 & nCount_RNA
# <50000 & percent.mt <15) cx.raw$src <- 'Culture'

# Zymo copied from RESRoberts (orginal code referenced this) Create Seurat objects and perform
# initial QC. Label original source.
cx.raw <- tenXLoadQC("/gpfs0/home2/gdrobertslab/rsexf001/OSHetero2021/S0016-zymo/filtered_feature_bc_matrix/",
    spec = "mixHuman")

cx.raw <- subset(cx.raw, subset = nFeature_RNA > 3500 & nCount_RNA < 50000 & percent.mt < 15)
cx.raw$src <- "Culture"

tib.raw <- tenXLoadQC("/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/filtered_feature_bc_matrix/",
    spec = "mixHuman")

tib.raw <- subset(tib.raw, subset = nFeature_RNA > 3000 & nCount_RNA < 60000 & percent.mt < 18)
tib.raw$src <- "Tibia"

lung.raw <- tenXLoadQC("/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/filtered_feature_bc_matrix/",
    spec = "mixHuman")

lung.raw <- subset(lung.raw, subset = nFeature_RNA > 1250 & nCount_RNA < 60000 & percent.mt < 25)
lung.raw$src <- "Lung"
```

## Figure 5B

B) UMAP analysis of lineage tagged OS-17 cells in culture, implanted into tibia, or inoculated to generate lung metastases. These lineage tags are heritable allowing us to track the transcriptional profiles of daughter cells originating from the same ancestor. Metastatic and orthotopic samples share many cells with similar phenotypes, though cultured cells are dramatically different (n = 2800 per condition)

```{r Fig5b}
# Subset all to #2800 cells in each condition
cx.raw <- subset(cx.raw, cells = sample(Cells(cx.raw), 2800))
tib.raw <- subset(tib.raw, cells = sample(Cells(tib.raw), 2800))
lung.raw <- subset(lung.raw, cells = sample(Cells(lung.raw), 2800))

# Merge into a single Seurat object
os17 <- merge(cx.raw, y = c(tib.raw, lung.raw), add.cell.ids = c("Culture", "Tibia", "Lung"), project = "LineageTracing")

# Process and cluster
os17 <- NormalizeData(os17) %>%
    FindVariableFeatures(selection.method = "vst") %>%
    ScaleData() %>%
    RunPCA(pc.genes = os.17@var.genes, npcs = 20) %>%
    RunUMAP(reduction = "pca", dims = 1:20) %>%
    FindNeighbors(reduction = "pca", dims = 1:20) %>%
    FindClusters(resolution = 0.3)

# CCR Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
os17 <- CellCycleScoring(object = os17, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

os17 <- ScaleData(object = os17, vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(x = os17))
os17 <- RunPCA(os17, pc.genes = os.17@var.genes, npcs = 20) %>%
    RunUMAP(reduction = "pca", dims = 1:20) %>%
    FindNeighbors(reduction = "pca", dims = 1:20) %>%
    FindClusters(resolution = 0.3)

DimPlot(os17, reduction = "umap", pt.size = 1, label = T) + coord_fixed() + ggtitle("OS17 Clusters") +
    scale_color_npg(alpha = 0.7)

# Plot the data
set.seed(100)
cell.ids <- sample(colnames(os17))
DimPlot(os17, reduction = "umap", group.by = "src", pt.size = 1, label = F, order = cell.ids) +
    coord_fixed() + ggtitle("OS17 by Source") + scale_color_npg()

################## Find the optimum resolution for clustering
os17.nC <- nRes(os17, res = seq(from = 0.2, to = 0.3, by = 0.01))

save(os17, file = "/home/gdrobertslab/rsexf001/OSHetero2021/os17.RData")
```

## Figure 5C

C) Frequency distribution of clones identified in each of the conditions. A high level of clonal diversity is maintained in the orthotopic tumors (78% unique clones in primary tumor compared to 86% unique clones in cell culture); though dominant clones emerge in the lung (only 33% unique clones). 

```{r Fig5c}
# Add lineage tracing tags to the Seurat objects
cx.raw <- processLTBC(cx.raw, lt.loc = "/gpfs0/home/gdrobertslab/rsexf001/OSHetero2021/S0016-zymo/lt.fq",
    cid.loc = "/gpfs0/home/gdrobertslab/rsexf001/OSHetero2021/S0016-zymo/cid.fq", histogram = T,
    title = "Culture, Top 40 Clones", ymax = 1.75, col.fill = "#E64B35FF", relative = T)

tib.raw <- processLTBC(tib.raw, lt.loc = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/lt.fq",
    cid.loc = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/cid.fq", histogram = T, title = "Tibia, Top 40 Clones",
    ymax = 1.75, col.fill = "#00A087FF", relative = T)

lung.raw <- processLTBC(lung.raw, lt.loc = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/lt.fq",
    cid.loc = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/cid.fq", histogram = T, title = "Lineage Barcode Enrichment by Microenvironment",
    ymax = 1.75, relative = T)
```

## Figure 5D

D) Overlay of cells sharing a common ancestor onto the dimensional reduction plots in the top four enriched clonal families.

### Processing

```{r Fig5d_1}
source("/home/gdrobertslab/rsexf001/OSHetero2021/scSeurat_updated.R")
load("/home/gdrobertslab/rsexf001/OSHetero2021/os17.RData")
# Extract the barcode frequency lists
cx.lt.list <- processLTBC(cx.raw, lt.loc = "/gpfs0/home/gdrobertslab/rsexf001/OSHetero2021/S0016-zymo/lt.fq",
    cid.loc = "/gpfs0/home/gdrobertslab/rsexf001/OSHetero2021/S0016-zymo/cid.fq", ret.list = TRUE)
tib.lt.list <- processLTBC(tib.raw, lt.loc = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/lt.fq",
    cid.loc = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/cid.fq", ret.list = TRUE)
lung.lt.list <- processLTBC(lung.raw, lt.loc = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/lt.fq",
    cid.loc = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/cid.fq", ret.list = TRUE)

# Subset the three samples, but retain the current umap data
cx.sub <- subset(os17, subset = src == "Culture")
tib.sub <- subset(os17, subset = src == "Tibia")
lung.sub <- subset(os17, subset = src == "Lung")

Idents(os17) <- os17$RNA_snn_res.0.3
DimPlot(os17, label = T, split.by = "src") + coord_fixed()

# Subset the three samples, but retain the current umap data
cx.sub <- subset(os17, subset = src == "Culture")
tib.sub <- subset(os17, subset = src == "Tibia")
lung.sub <- subset(os17, subset = src == "Lung")

DimPlot(lung.sub, label = T) + coord_fixed()

#################### subset to cells that have a lineage tag memory.limit(size = 40000)
source("/home/gdrobertslab/rsexf001/OSHetero2021/Downstream.v2.R")
# library(xlsx) Culture
Culture.LTcells <- vector(mode = "character")
for (i in 1:nrow(cx.lt.list)) {
    temp <- WhichCells(cx.sub, expression = lt == cx.lt.list[[i, 1]])
    Culture.LTcells <- append(Culture.LTcells, temp)
}
cx.tagged <- subset(cx.sub, cells = Culture.LTcells)

# Tibia
tib.LTcells <- vector(mode = "character")
for (i in 1:nrow(tib.lt.list)) {
    temp <- WhichCells(tib.sub, expression = lt == tib.lt.list[[i, 1]])
    tib.LTcells <- append(tib.LTcells, temp)
}
tib.tagged <- subset(tib.sub, cells = tib.LTcells)

# Lung
lung.LTcells <- vector(mode = "character")
for (i in 1:nrow(lung.lt.list)) {
    # originally said 1:515, unclear why and received a 'subscript out of bounds' error, so
    # changed to match the previous code for cx and tib -Emily
    temp <- WhichCells(lung.sub, expression = lt == lung.lt.list[[i, 1]])
    lung.LTcells <- append(lung.LTcells, temp)
}

lung.tagged <- subset(lung.sub, cells = lung.LTcells)

lung.sub <- subset(lung.sub, idents = c("0", "1", "3", "6"))  ##Remove outlier cells in cluster 2 and 5 # Had to edit some of these cluster names (likely due to random seed changes in plotting) -Emily

pdf("LungOutline.pdf", height = 5, width = 5)
DimPlot(lung.sub, reduction = "umap", pt.size = 1, label = TRUE) + coord_fixed() + ggtitle("Lung-derived") +
    # xlim(0,9) +
NoAxes() + scale_color_npg(alpha = 1)
dev.off()
## X11cairo 
##        2

# Show example in Rm html file -Emily
DimPlot(lung.sub, reduction = "umap", pt.size = 1, label = TRUE) + coord_fixed() + ggtitle("Lung-derived") +
    # xlim(0,9) +
NoAxes() + scale_color_npg(alpha = 1)
```

### Lung Panel

```{r Fig5d}
### Key Figure Image for Lung -Emily
### Need to repeat for culture and tibia -Emily
# Lung
vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
pdf("grid_lung_v4.pdf", width = 10, height = 10)
grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 4)))
for (i in 1:4) {
    p <- DimPlot(lung.tagged, reduction = "umap", pt.size = 1, cells.highlight = WhichCells(lung.sub,
        expression = lt == lung.lt.list[[i, 1]]), cols.highlight = "#4DBBD5FF", sizes.highlight = 3) +
        coord_fixed() + theme(legend.position = "none") + ggtitle(paste("Clone", lung.lt.list[[i,
        1]])) + #xlim(0,9) + 1]])) + #xlim(0,9) +
    NoLegend() + NoAxes()
    print(p, vp = vplayout(ceiling(i/4), i))
}
dev.off()
```

## Figure 5E

E) UMAP visualization of cells belonging to top ten enriched clonal families highlighted in red.

```{r Fig5e}
# Compare Enriched vs non-enriched Glycolysis and Hypoxia seem to be trending together

# Group first 12 against the tagged Lung
data <- lung.tagged
LT <- vector(mode = "character")
for (i in 1:10) {
    temp <- WhichCells(lung.sub, expression = lt == lung.lt.list[[i, 1]])
    LT <- append(LT, temp)
}
Idents(data, cells = LT) <- "Enriched"

# Non-enriched cells are clones with a frequency of 1 lung.lt.list[134,] 2 lung.lt.list[134,]
# 1
LT <- vector(mode = "character")
for (i in 135:377) {
    temp <- WhichCells(lung.sub, expression = lt == lung.lt.list[[i, 1]])
    LT <- append(LT, temp)
}
Idents(data, cells = LT) <- "Non-enriched"

cell.ids <- sample(colnames(data))

# Culture plots pdf('OS17_Enr.pdf', width = 5, height = 5)
DimPlot(data, reduction = "umap", pt.size = 1, cells.highlight = WhichCells(data, idents = "Enriched"),
    cols.highlight = "#E64B35FF", sizes.highlight = 2, order = cell.ids) + coord_fixed()
```

## Figure 5F

F) Pathway enrichment analysis for hallmark gene sets comparing enriched clones relative to the remaining metastatic tumor cells identifies Hypoxia, Glycolysis, EMT and MTORC1 signaling related genes to be significantly upregulated

```{r Fig5f}
# Pathways associated with Enriched clonal families
library(tibble)
library(tidyr)
library(ggplot2)
library(tidyverse)
library(stringr)

B.list <- list(OS17 = data)

em.hm.list <- list()
for (i in 1:length(B.list)) {
    em.hm.list[[i]] <- DGEA(B.list[[i]])
}

temp1 <- em.hm.list[[1]]
temp1 <- -log10(temp1)
# remove '_' by removing special characters
c <- rownames(temp1)
c <- gsub("HALLMARK ", "", c)
c <- gsub("_", " ", c)
rownames(temp1) <- c
# remove negative 0 values due to number of decimal points displayed
temp1 <- abs(temp1)
temp1 <- temp1["Enriched"]

# Subset to only pathways with siginificant pvalue (-1 * log10(0.05)) = 1.30103
temp1["Enriched"] >= 1.30103 -> temp1$logical
colnames(temp1) <- c("pvalue", "significant")
logical <- temp1$significant
temp1 <- temp1[logical, ]
temp1 <- rownames_to_column(temp1, var = "Pathway")
position <- order(temp1$pvalue, decreasing = TRUE)
temp1 <- temp1[position, ]
plot <- ggplot(temp1, aes(x = pvalue, y = Pathway)) + geom_bar(stat = "identity") + scale_fill_brewer(type = "qual",
    palette = "Set1") + theme_bw() + ylab("")

pdf("Enr_Pathway.pdf", height = 5, width = 10)
plot(plot)
dev.off()

# for vewing in html output -Emily
plot
```

